// HDLtool/debug/debug.js
"use strict";

const vscode  = require("vscode");
const child   = require("child_process");


class toolRegister {
    constructor (opeParam) {
        this.opeParam = opeParam;
        this.setting  = vscode.workspace.getConfiguration();
        this.getConfig();
        vscode.workspace.onDidChangeConfiguration(function () {
            this.getConfig();
        });
    }
    getConfig() {
        this.Parity    = this.setting.get('TOOL.SerialPortMonitor.Parity');
        this.BaudRate  = this.setting.get('TOOL.SerialPortMonitor.BaudRate');
        this.DataBits  = this.setting.get('TOOL.SerialPortMonitor.DataBits');
        this.StopBits  = this.setting.get('TOOL.SerialPortMonitor.StopBits');
    }
    serialPortTerminal(serialPortName,command) {
        // var serialPort = new serialport.SerialPort(com_num, {
        //     parity:   this.Parity,    //校验位
        //     baudrate: this.BaudRate,  //波特率设置
        //     databits: this.DataBits,  //数据位
        //     stopbits: this.StopBits   //停止位
        // });
        if (this.terminal.ensureTerminalExists(`${serialPortName}`)) {
            vscode.window.showWarningMessage('This serial port number is in use!');
        } else {
            let serialPort = vscode.window.createTerminal({ name: `${serialPortName}` });
            serialPort.show(true);
            serialPort.sendText(command);
        }
    }
    runSerialPort(command,) {
        child.exec(command,function (error, stdout, stderr) {
            let content = stdout.replace(/\s*/g,'');
            let SerialPortList = content.split("-");
            let porteries = `${BaudRate} ${DataBits} ${StopBits} ${Parity}`;
            if (SerialPortList[0] == "none") {
                vscode.window.showWarningMessage("Not found any serial port !");
            }
            if (SerialPortList[0] == "one") {
                porteries = SerialPortList[1] + " " + porteries;
                let command = `python ${this.opeParam.rootPath}/../resources/tool/.Script/Serial_Port.py runthread ${porteries}`;
                this.serialPortTerminal(SerialPortList[1],command);
            }
            if (SerialPortList[0] == "multiple") {
                SerialPortList.splice(0,1);
                vscode.window.showQuickPick(SerialPortList).then(selection => {
                    if (!selection) {
                        return;
                    }
                    porteries = selection + " " + porteries;
                    let command = `python ${this.opeParam.rootPath}/../resources/tool/.Script/Serial_Port.py runthread ${porteries}`;
                    this.serialPortTerminal(selection,command);
                });
            }
            if (error !== null) {
                vscode.window.showErrorMessage(error);
            }
        });
    }
    genBootLoadFile() {
        
    }
    clean() {
        
    }
    serialPort() {
        let command = `python ${this.opeParam.rootPath}/../resources/tool/.Script/Serial_Port.py getCurrentPort`;
        runSerialPort(command,this.opeParam.rootPath);
    }
}
exports.toolRegister = toolRegister;
