// HDLparser/src/vhdl/vhdlListener.js
// Generated from src/vhdl/vhdl.g4 by ANTLR 4.7.2
// jshint ignore: start
var antlr4 = require('antlr4/index');

// This class defines a complete listener for a parse tree produced by vhdlParser.
function vhdlListener() {
	antlr4.tree.ParseTreeListener.call(this);
	return this;
}

vhdlListener.prototype = Object.create(antlr4.tree.ParseTreeListener.prototype);
vhdlListener.prototype.constructor = vhdlListener;

// Enter a parse tree produced by vhdlParser#abstract_literal.
vhdlListener.prototype.enterAbstract_literal = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#abstract_literal.
vhdlListener.prototype.exitAbstract_literal = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#access_type_definition.
vhdlListener.prototype.enterAccess_type_definition = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#access_type_definition.
vhdlListener.prototype.exitAccess_type_definition = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#across_aspect.
vhdlListener.prototype.enterAcross_aspect = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#across_aspect.
vhdlListener.prototype.exitAcross_aspect = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#actual_designator.
vhdlListener.prototype.enterActual_designator = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#actual_designator.
vhdlListener.prototype.exitActual_designator = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#actual_parameter_part.
vhdlListener.prototype.enterActual_parameter_part = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#actual_parameter_part.
vhdlListener.prototype.exitActual_parameter_part = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#actual_part.
vhdlListener.prototype.enterActual_part = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#actual_part.
vhdlListener.prototype.exitActual_part = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#adding_operator.
vhdlListener.prototype.enterAdding_operator = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#adding_operator.
vhdlListener.prototype.exitAdding_operator = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#aggregate.
vhdlListener.prototype.enterAggregate = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#aggregate.
vhdlListener.prototype.exitAggregate = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#alias_declaration.
vhdlListener.prototype.enterAlias_declaration = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#alias_declaration.
vhdlListener.prototype.exitAlias_declaration = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#alias_designator.
vhdlListener.prototype.enterAlias_designator = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#alias_designator.
vhdlListener.prototype.exitAlias_designator = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#alias_indication.
vhdlListener.prototype.enterAlias_indication = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#alias_indication.
vhdlListener.prototype.exitAlias_indication = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#allocator.
vhdlListener.prototype.enterAllocator = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#allocator.
vhdlListener.prototype.exitAllocator = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#architecture_body.
vhdlListener.prototype.enterArchitecture_body = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#architecture_body.
vhdlListener.prototype.exitArchitecture_body = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#architecture_declarative_part.
vhdlListener.prototype.enterArchitecture_declarative_part = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#architecture_declarative_part.
vhdlListener.prototype.exitArchitecture_declarative_part = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#architecture_statement.
vhdlListener.prototype.enterArchitecture_statement = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#architecture_statement.
vhdlListener.prototype.exitArchitecture_statement = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#architecture_statement_part.
vhdlListener.prototype.enterArchitecture_statement_part = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#architecture_statement_part.
vhdlListener.prototype.exitArchitecture_statement_part = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#array_nature_definition.
vhdlListener.prototype.enterArray_nature_definition = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#array_nature_definition.
vhdlListener.prototype.exitArray_nature_definition = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#array_type_definition.
vhdlListener.prototype.enterArray_type_definition = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#array_type_definition.
vhdlListener.prototype.exitArray_type_definition = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#assertion.
vhdlListener.prototype.enterAssertion = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#assertion.
vhdlListener.prototype.exitAssertion = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#assertion_statement.
vhdlListener.prototype.enterAssertion_statement = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#assertion_statement.
vhdlListener.prototype.exitAssertion_statement = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#association_element.
vhdlListener.prototype.enterAssociation_element = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#association_element.
vhdlListener.prototype.exitAssociation_element = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#association_list.
vhdlListener.prototype.enterAssociation_list = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#association_list.
vhdlListener.prototype.exitAssociation_list = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#attribute_declaration.
vhdlListener.prototype.enterAttribute_declaration = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#attribute_declaration.
vhdlListener.prototype.exitAttribute_declaration = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#attribute_designator.
vhdlListener.prototype.enterAttribute_designator = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#attribute_designator.
vhdlListener.prototype.exitAttribute_designator = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#attribute_specification.
vhdlListener.prototype.enterAttribute_specification = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#attribute_specification.
vhdlListener.prototype.exitAttribute_specification = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#base_unit_declaration.
vhdlListener.prototype.enterBase_unit_declaration = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#base_unit_declaration.
vhdlListener.prototype.exitBase_unit_declaration = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#binding_indication.
vhdlListener.prototype.enterBinding_indication = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#binding_indication.
vhdlListener.prototype.exitBinding_indication = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#block_configuration.
vhdlListener.prototype.enterBlock_configuration = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#block_configuration.
vhdlListener.prototype.exitBlock_configuration = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#block_declarative_item.
vhdlListener.prototype.enterBlock_declarative_item = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#block_declarative_item.
vhdlListener.prototype.exitBlock_declarative_item = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#block_declarative_part.
vhdlListener.prototype.enterBlock_declarative_part = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#block_declarative_part.
vhdlListener.prototype.exitBlock_declarative_part = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#block_header.
vhdlListener.prototype.enterBlock_header = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#block_header.
vhdlListener.prototype.exitBlock_header = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#block_specification.
vhdlListener.prototype.enterBlock_specification = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#block_specification.
vhdlListener.prototype.exitBlock_specification = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#block_statement.
vhdlListener.prototype.enterBlock_statement = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#block_statement.
vhdlListener.prototype.exitBlock_statement = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#block_statement_part.
vhdlListener.prototype.enterBlock_statement_part = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#block_statement_part.
vhdlListener.prototype.exitBlock_statement_part = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#branch_quantity_declaration.
vhdlListener.prototype.enterBranch_quantity_declaration = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#branch_quantity_declaration.
vhdlListener.prototype.exitBranch_quantity_declaration = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#break_element.
vhdlListener.prototype.enterBreak_element = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#break_element.
vhdlListener.prototype.exitBreak_element = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#break_list.
vhdlListener.prototype.enterBreak_list = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#break_list.
vhdlListener.prototype.exitBreak_list = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#break_selector_clause.
vhdlListener.prototype.enterBreak_selector_clause = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#break_selector_clause.
vhdlListener.prototype.exitBreak_selector_clause = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#break_statement.
vhdlListener.prototype.enterBreak_statement = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#break_statement.
vhdlListener.prototype.exitBreak_statement = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#case_statement.
vhdlListener.prototype.enterCase_statement = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#case_statement.
vhdlListener.prototype.exitCase_statement = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#case_statement_alternative.
vhdlListener.prototype.enterCase_statement_alternative = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#case_statement_alternative.
vhdlListener.prototype.exitCase_statement_alternative = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#choice.
vhdlListener.prototype.enterChoice = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#choice.
vhdlListener.prototype.exitChoice = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#choices.
vhdlListener.prototype.enterChoices = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#choices.
vhdlListener.prototype.exitChoices = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#component_configuration.
vhdlListener.prototype.enterComponent_configuration = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#component_configuration.
vhdlListener.prototype.exitComponent_configuration = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#component_declaration.
vhdlListener.prototype.enterComponent_declaration = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#component_declaration.
vhdlListener.prototype.exitComponent_declaration = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#component_instantiation_statement.
vhdlListener.prototype.enterComponent_instantiation_statement = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#component_instantiation_statement.
vhdlListener.prototype.exitComponent_instantiation_statement = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#component_specification.
vhdlListener.prototype.enterComponent_specification = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#component_specification.
vhdlListener.prototype.exitComponent_specification = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#composite_nature_definition.
vhdlListener.prototype.enterComposite_nature_definition = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#composite_nature_definition.
vhdlListener.prototype.exitComposite_nature_definition = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#composite_type_definition.
vhdlListener.prototype.enterComposite_type_definition = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#composite_type_definition.
vhdlListener.prototype.exitComposite_type_definition = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#concurrent_assertion_statement.
vhdlListener.prototype.enterConcurrent_assertion_statement = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#concurrent_assertion_statement.
vhdlListener.prototype.exitConcurrent_assertion_statement = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#concurrent_break_statement.
vhdlListener.prototype.enterConcurrent_break_statement = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#concurrent_break_statement.
vhdlListener.prototype.exitConcurrent_break_statement = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#concurrent_procedure_call_statement.
vhdlListener.prototype.enterConcurrent_procedure_call_statement = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#concurrent_procedure_call_statement.
vhdlListener.prototype.exitConcurrent_procedure_call_statement = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#concurrent_signal_assignment_statement.
vhdlListener.prototype.enterConcurrent_signal_assignment_statement = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#concurrent_signal_assignment_statement.
vhdlListener.prototype.exitConcurrent_signal_assignment_statement = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#condition.
vhdlListener.prototype.enterCondition = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#condition.
vhdlListener.prototype.exitCondition = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#condition_clause.
vhdlListener.prototype.enterCondition_clause = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#condition_clause.
vhdlListener.prototype.exitCondition_clause = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#conditional_signal_assignment.
vhdlListener.prototype.enterConditional_signal_assignment = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#conditional_signal_assignment.
vhdlListener.prototype.exitConditional_signal_assignment = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#conditional_waveforms.
vhdlListener.prototype.enterConditional_waveforms = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#conditional_waveforms.
vhdlListener.prototype.exitConditional_waveforms = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#configuration_declaration.
vhdlListener.prototype.enterConfiguration_declaration = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#configuration_declaration.
vhdlListener.prototype.exitConfiguration_declaration = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#configuration_declarative_item.
vhdlListener.prototype.enterConfiguration_declarative_item = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#configuration_declarative_item.
vhdlListener.prototype.exitConfiguration_declarative_item = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#configuration_declarative_part.
vhdlListener.prototype.enterConfiguration_declarative_part = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#configuration_declarative_part.
vhdlListener.prototype.exitConfiguration_declarative_part = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#configuration_item.
vhdlListener.prototype.enterConfiguration_item = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#configuration_item.
vhdlListener.prototype.exitConfiguration_item = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#configuration_specification.
vhdlListener.prototype.enterConfiguration_specification = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#configuration_specification.
vhdlListener.prototype.exitConfiguration_specification = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#constant_declaration.
vhdlListener.prototype.enterConstant_declaration = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#constant_declaration.
vhdlListener.prototype.exitConstant_declaration = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#constrained_array_definition.
vhdlListener.prototype.enterConstrained_array_definition = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#constrained_array_definition.
vhdlListener.prototype.exitConstrained_array_definition = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#constrained_nature_definition.
vhdlListener.prototype.enterConstrained_nature_definition = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#constrained_nature_definition.
vhdlListener.prototype.exitConstrained_nature_definition = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#constraint.
vhdlListener.prototype.enterConstraint = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#constraint.
vhdlListener.prototype.exitConstraint = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#context_clause.
vhdlListener.prototype.enterContext_clause = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#context_clause.
vhdlListener.prototype.exitContext_clause = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#context_item.
vhdlListener.prototype.enterContext_item = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#context_item.
vhdlListener.prototype.exitContext_item = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#delay_mechanism.
vhdlListener.prototype.enterDelay_mechanism = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#delay_mechanism.
vhdlListener.prototype.exitDelay_mechanism = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#parse.
vhdlListener.prototype.enterParse = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#parse.
vhdlListener.prototype.exitParse = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#design_unit.
vhdlListener.prototype.enterDesign_unit = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#design_unit.
vhdlListener.prototype.exitDesign_unit = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#designator.
vhdlListener.prototype.enterDesignator = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#designator.
vhdlListener.prototype.exitDesignator = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#direction.
vhdlListener.prototype.enterDirection = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#direction.
vhdlListener.prototype.exitDirection = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#disconnection_specification.
vhdlListener.prototype.enterDisconnection_specification = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#disconnection_specification.
vhdlListener.prototype.exitDisconnection_specification = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#discrete_range.
vhdlListener.prototype.enterDiscrete_range = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#discrete_range.
vhdlListener.prototype.exitDiscrete_range = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#element_association.
vhdlListener.prototype.enterElement_association = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#element_association.
vhdlListener.prototype.exitElement_association = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#element_declaration.
vhdlListener.prototype.enterElement_declaration = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#element_declaration.
vhdlListener.prototype.exitElement_declaration = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#element_subnature_definition.
vhdlListener.prototype.enterElement_subnature_definition = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#element_subnature_definition.
vhdlListener.prototype.exitElement_subnature_definition = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#element_subtype_definition.
vhdlListener.prototype.enterElement_subtype_definition = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#element_subtype_definition.
vhdlListener.prototype.exitElement_subtype_definition = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#entity_aspect.
vhdlListener.prototype.enterEntity_aspect = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#entity_aspect.
vhdlListener.prototype.exitEntity_aspect = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#entity_class.
vhdlListener.prototype.enterEntity_class = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#entity_class.
vhdlListener.prototype.exitEntity_class = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#entity_class_entry.
vhdlListener.prototype.enterEntity_class_entry = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#entity_class_entry.
vhdlListener.prototype.exitEntity_class_entry = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#entity_class_entry_list.
vhdlListener.prototype.enterEntity_class_entry_list = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#entity_class_entry_list.
vhdlListener.prototype.exitEntity_class_entry_list = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#entity_declaration.
vhdlListener.prototype.enterEntity_declaration = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#entity_declaration.
vhdlListener.prototype.exitEntity_declaration = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#entity_declarative_item.
vhdlListener.prototype.enterEntity_declarative_item = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#entity_declarative_item.
vhdlListener.prototype.exitEntity_declarative_item = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#entity_declarative_part.
vhdlListener.prototype.enterEntity_declarative_part = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#entity_declarative_part.
vhdlListener.prototype.exitEntity_declarative_part = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#entity_designator.
vhdlListener.prototype.enterEntity_designator = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#entity_designator.
vhdlListener.prototype.exitEntity_designator = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#entity_header.
vhdlListener.prototype.enterEntity_header = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#entity_header.
vhdlListener.prototype.exitEntity_header = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#entity_name_list.
vhdlListener.prototype.enterEntity_name_list = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#entity_name_list.
vhdlListener.prototype.exitEntity_name_list = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#entity_specification.
vhdlListener.prototype.enterEntity_specification = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#entity_specification.
vhdlListener.prototype.exitEntity_specification = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#entity_statement.
vhdlListener.prototype.enterEntity_statement = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#entity_statement.
vhdlListener.prototype.exitEntity_statement = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#entity_statement_part.
vhdlListener.prototype.enterEntity_statement_part = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#entity_statement_part.
vhdlListener.prototype.exitEntity_statement_part = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#entity_tag.
vhdlListener.prototype.enterEntity_tag = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#entity_tag.
vhdlListener.prototype.exitEntity_tag = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#enumeration_literal.
vhdlListener.prototype.enterEnumeration_literal = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#enumeration_literal.
vhdlListener.prototype.exitEnumeration_literal = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#enumeration_type_definition.
vhdlListener.prototype.enterEnumeration_type_definition = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#enumeration_type_definition.
vhdlListener.prototype.exitEnumeration_type_definition = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#exit_statement.
vhdlListener.prototype.enterExit_statement = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#exit_statement.
vhdlListener.prototype.exitExit_statement = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#expression.
vhdlListener.prototype.enterExpression = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#expression.
vhdlListener.prototype.exitExpression = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#factor.
vhdlListener.prototype.enterFactor = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#factor.
vhdlListener.prototype.exitFactor = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#file_declaration.
vhdlListener.prototype.enterFile_declaration = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#file_declaration.
vhdlListener.prototype.exitFile_declaration = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#file_logical_name.
vhdlListener.prototype.enterFile_logical_name = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#file_logical_name.
vhdlListener.prototype.exitFile_logical_name = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#file_open_information.
vhdlListener.prototype.enterFile_open_information = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#file_open_information.
vhdlListener.prototype.exitFile_open_information = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#file_type_definition.
vhdlListener.prototype.enterFile_type_definition = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#file_type_definition.
vhdlListener.prototype.exitFile_type_definition = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#formal_parameter_list.
vhdlListener.prototype.enterFormal_parameter_list = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#formal_parameter_list.
vhdlListener.prototype.exitFormal_parameter_list = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#formal_part.
vhdlListener.prototype.enterFormal_part = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#formal_part.
vhdlListener.prototype.exitFormal_part = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#free_quantity_declaration.
vhdlListener.prototype.enterFree_quantity_declaration = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#free_quantity_declaration.
vhdlListener.prototype.exitFree_quantity_declaration = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#generate_statement.
vhdlListener.prototype.enterGenerate_statement = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#generate_statement.
vhdlListener.prototype.exitGenerate_statement = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#generation_scheme.
vhdlListener.prototype.enterGeneration_scheme = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#generation_scheme.
vhdlListener.prototype.exitGeneration_scheme = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#generic_clause.
vhdlListener.prototype.enterGeneric_clause = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#generic_clause.
vhdlListener.prototype.exitGeneric_clause = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#generic_list.
vhdlListener.prototype.enterGeneric_list = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#generic_list.
vhdlListener.prototype.exitGeneric_list = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#generic_map_aspect.
vhdlListener.prototype.enterGeneric_map_aspect = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#generic_map_aspect.
vhdlListener.prototype.exitGeneric_map_aspect = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#group_constituent.
vhdlListener.prototype.enterGroup_constituent = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#group_constituent.
vhdlListener.prototype.exitGroup_constituent = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#group_constituent_list.
vhdlListener.prototype.enterGroup_constituent_list = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#group_constituent_list.
vhdlListener.prototype.exitGroup_constituent_list = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#group_declaration.
vhdlListener.prototype.enterGroup_declaration = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#group_declaration.
vhdlListener.prototype.exitGroup_declaration = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#group_template_declaration.
vhdlListener.prototype.enterGroup_template_declaration = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#group_template_declaration.
vhdlListener.prototype.exitGroup_template_declaration = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#guarded_signal_specification.
vhdlListener.prototype.enterGuarded_signal_specification = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#guarded_signal_specification.
vhdlListener.prototype.exitGuarded_signal_specification = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#identifier.
vhdlListener.prototype.enterIdentifier = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#identifier.
vhdlListener.prototype.exitIdentifier = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#identifier_list.
vhdlListener.prototype.enterIdentifier_list = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#identifier_list.
vhdlListener.prototype.exitIdentifier_list = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#if_statement.
vhdlListener.prototype.enterIf_statement = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#if_statement.
vhdlListener.prototype.exitIf_statement = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#index_constraint.
vhdlListener.prototype.enterIndex_constraint = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#index_constraint.
vhdlListener.prototype.exitIndex_constraint = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#index_specification.
vhdlListener.prototype.enterIndex_specification = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#index_specification.
vhdlListener.prototype.exitIndex_specification = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#index_subtype_definition.
vhdlListener.prototype.enterIndex_subtype_definition = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#index_subtype_definition.
vhdlListener.prototype.exitIndex_subtype_definition = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#instantiated_unit.
vhdlListener.prototype.enterInstantiated_unit = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#instantiated_unit.
vhdlListener.prototype.exitInstantiated_unit = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#instantiation_list.
vhdlListener.prototype.enterInstantiation_list = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#instantiation_list.
vhdlListener.prototype.exitInstantiation_list = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#interface_constant_declaration.
vhdlListener.prototype.enterInterface_constant_declaration = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#interface_constant_declaration.
vhdlListener.prototype.exitInterface_constant_declaration = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#interface_declaration.
vhdlListener.prototype.enterInterface_declaration = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#interface_declaration.
vhdlListener.prototype.exitInterface_declaration = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#interface_element.
vhdlListener.prototype.enterInterface_element = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#interface_element.
vhdlListener.prototype.exitInterface_element = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#interface_file_declaration.
vhdlListener.prototype.enterInterface_file_declaration = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#interface_file_declaration.
vhdlListener.prototype.exitInterface_file_declaration = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#interface_signal_list.
vhdlListener.prototype.enterInterface_signal_list = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#interface_signal_list.
vhdlListener.prototype.exitInterface_signal_list = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#interface_port_list.
vhdlListener.prototype.enterInterface_port_list = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#interface_port_list.
vhdlListener.prototype.exitInterface_port_list = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#interface_list.
vhdlListener.prototype.enterInterface_list = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#interface_list.
vhdlListener.prototype.exitInterface_list = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#interface_quantity_declaration.
vhdlListener.prototype.enterInterface_quantity_declaration = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#interface_quantity_declaration.
vhdlListener.prototype.exitInterface_quantity_declaration = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#interface_port_declaration.
vhdlListener.prototype.enterInterface_port_declaration = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#interface_port_declaration.
vhdlListener.prototype.exitInterface_port_declaration = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#interface_signal_declaration.
vhdlListener.prototype.enterInterface_signal_declaration = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#interface_signal_declaration.
vhdlListener.prototype.exitInterface_signal_declaration = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#interface_terminal_declaration.
vhdlListener.prototype.enterInterface_terminal_declaration = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#interface_terminal_declaration.
vhdlListener.prototype.exitInterface_terminal_declaration = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#interface_variable_declaration.
vhdlListener.prototype.enterInterface_variable_declaration = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#interface_variable_declaration.
vhdlListener.prototype.exitInterface_variable_declaration = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#iteration_scheme.
vhdlListener.prototype.enterIteration_scheme = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#iteration_scheme.
vhdlListener.prototype.exitIteration_scheme = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#label_colon.
vhdlListener.prototype.enterLabel_colon = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#label_colon.
vhdlListener.prototype.exitLabel_colon = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#library_clause.
vhdlListener.prototype.enterLibrary_clause = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#library_clause.
vhdlListener.prototype.exitLibrary_clause = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#library_unit.
vhdlListener.prototype.enterLibrary_unit = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#library_unit.
vhdlListener.prototype.exitLibrary_unit = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#literal.
vhdlListener.prototype.enterLiteral = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#literal.
vhdlListener.prototype.exitLiteral = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#logical_name.
vhdlListener.prototype.enterLogical_name = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#logical_name.
vhdlListener.prototype.exitLogical_name = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#logical_name_list.
vhdlListener.prototype.enterLogical_name_list = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#logical_name_list.
vhdlListener.prototype.exitLogical_name_list = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#logical_operator.
vhdlListener.prototype.enterLogical_operator = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#logical_operator.
vhdlListener.prototype.exitLogical_operator = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#loop_statement.
vhdlListener.prototype.enterLoop_statement = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#loop_statement.
vhdlListener.prototype.exitLoop_statement = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#signal_mode.
vhdlListener.prototype.enterSignal_mode = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#signal_mode.
vhdlListener.prototype.exitSignal_mode = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#multiplying_operator.
vhdlListener.prototype.enterMultiplying_operator = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#multiplying_operator.
vhdlListener.prototype.exitMultiplying_operator = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#name.
vhdlListener.prototype.enterName = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#name.
vhdlListener.prototype.exitName = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#name_part.
vhdlListener.prototype.enterName_part = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#name_part.
vhdlListener.prototype.exitName_part = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#selected_name.
vhdlListener.prototype.enterSelected_name = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#selected_name.
vhdlListener.prototype.exitSelected_name = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#selected_name_part.
vhdlListener.prototype.enterSelected_name_part = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#selected_name_part.
vhdlListener.prototype.exitSelected_name_part = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#function_call_or_indexed_name_part.
vhdlListener.prototype.enterFunction_call_or_indexed_name_part = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#function_call_or_indexed_name_part.
vhdlListener.prototype.exitFunction_call_or_indexed_name_part = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#slice_name_part.
vhdlListener.prototype.enterSlice_name_part = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#slice_name_part.
vhdlListener.prototype.exitSlice_name_part = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#attribute_name_part.
vhdlListener.prototype.enterAttribute_name_part = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#attribute_name_part.
vhdlListener.prototype.exitAttribute_name_part = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#nature_declaration.
vhdlListener.prototype.enterNature_declaration = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#nature_declaration.
vhdlListener.prototype.exitNature_declaration = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#nature_definition.
vhdlListener.prototype.enterNature_definition = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#nature_definition.
vhdlListener.prototype.exitNature_definition = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#nature_element_declaration.
vhdlListener.prototype.enterNature_element_declaration = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#nature_element_declaration.
vhdlListener.prototype.exitNature_element_declaration = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#next_statement.
vhdlListener.prototype.enterNext_statement = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#next_statement.
vhdlListener.prototype.exitNext_statement = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#numeric_literal.
vhdlListener.prototype.enterNumeric_literal = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#numeric_literal.
vhdlListener.prototype.exitNumeric_literal = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#object_declaration.
vhdlListener.prototype.enterObject_declaration = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#object_declaration.
vhdlListener.prototype.exitObject_declaration = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#opts.
vhdlListener.prototype.enterOpts = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#opts.
vhdlListener.prototype.exitOpts = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#package_body.
vhdlListener.prototype.enterPackage_body = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#package_body.
vhdlListener.prototype.exitPackage_body = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#package_body_declarative_item.
vhdlListener.prototype.enterPackage_body_declarative_item = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#package_body_declarative_item.
vhdlListener.prototype.exitPackage_body_declarative_item = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#package_body_declarative_part.
vhdlListener.prototype.enterPackage_body_declarative_part = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#package_body_declarative_part.
vhdlListener.prototype.exitPackage_body_declarative_part = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#package_declaration.
vhdlListener.prototype.enterPackage_declaration = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#package_declaration.
vhdlListener.prototype.exitPackage_declaration = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#package_declarative_item.
vhdlListener.prototype.enterPackage_declarative_item = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#package_declarative_item.
vhdlListener.prototype.exitPackage_declarative_item = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#package_declarative_part.
vhdlListener.prototype.enterPackage_declarative_part = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#package_declarative_part.
vhdlListener.prototype.exitPackage_declarative_part = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#parameter_specification.
vhdlListener.prototype.enterParameter_specification = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#parameter_specification.
vhdlListener.prototype.exitParameter_specification = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#physical_literal.
vhdlListener.prototype.enterPhysical_literal = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#physical_literal.
vhdlListener.prototype.exitPhysical_literal = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#physical_type_definition.
vhdlListener.prototype.enterPhysical_type_definition = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#physical_type_definition.
vhdlListener.prototype.exitPhysical_type_definition = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#port_clause.
vhdlListener.prototype.enterPort_clause = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#port_clause.
vhdlListener.prototype.exitPort_clause = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#port_list.
vhdlListener.prototype.enterPort_list = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#port_list.
vhdlListener.prototype.exitPort_list = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#port_map_aspect.
vhdlListener.prototype.enterPort_map_aspect = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#port_map_aspect.
vhdlListener.prototype.exitPort_map_aspect = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#primary.
vhdlListener.prototype.enterPrimary = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#primary.
vhdlListener.prototype.exitPrimary = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#primary_unit.
vhdlListener.prototype.enterPrimary_unit = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#primary_unit.
vhdlListener.prototype.exitPrimary_unit = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#procedural_declarative_item.
vhdlListener.prototype.enterProcedural_declarative_item = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#procedural_declarative_item.
vhdlListener.prototype.exitProcedural_declarative_item = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#procedural_declarative_part.
vhdlListener.prototype.enterProcedural_declarative_part = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#procedural_declarative_part.
vhdlListener.prototype.exitProcedural_declarative_part = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#procedural_statement_part.
vhdlListener.prototype.enterProcedural_statement_part = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#procedural_statement_part.
vhdlListener.prototype.exitProcedural_statement_part = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#procedure_call.
vhdlListener.prototype.enterProcedure_call = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#procedure_call.
vhdlListener.prototype.exitProcedure_call = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#procedure_call_statement.
vhdlListener.prototype.enterProcedure_call_statement = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#procedure_call_statement.
vhdlListener.prototype.exitProcedure_call_statement = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#process_declarative_item.
vhdlListener.prototype.enterProcess_declarative_item = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#process_declarative_item.
vhdlListener.prototype.exitProcess_declarative_item = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#process_declarative_part.
vhdlListener.prototype.enterProcess_declarative_part = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#process_declarative_part.
vhdlListener.prototype.exitProcess_declarative_part = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#process_statement.
vhdlListener.prototype.enterProcess_statement = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#process_statement.
vhdlListener.prototype.exitProcess_statement = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#process_statement_part.
vhdlListener.prototype.enterProcess_statement_part = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#process_statement_part.
vhdlListener.prototype.exitProcess_statement_part = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#qualified_expression.
vhdlListener.prototype.enterQualified_expression = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#qualified_expression.
vhdlListener.prototype.exitQualified_expression = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#quantity_declaration.
vhdlListener.prototype.enterQuantity_declaration = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#quantity_declaration.
vhdlListener.prototype.exitQuantity_declaration = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#quantity_list.
vhdlListener.prototype.enterQuantity_list = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#quantity_list.
vhdlListener.prototype.exitQuantity_list = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#quantity_specification.
vhdlListener.prototype.enterQuantity_specification = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#quantity_specification.
vhdlListener.prototype.exitQuantity_specification = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#range_decl.
vhdlListener.prototype.enterRange_decl = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#range_decl.
vhdlListener.prototype.exitRange_decl = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#explicit_range.
vhdlListener.prototype.enterExplicit_range = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#explicit_range.
vhdlListener.prototype.exitExplicit_range = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#range_constraint.
vhdlListener.prototype.enterRange_constraint = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#range_constraint.
vhdlListener.prototype.exitRange_constraint = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#record_nature_definition.
vhdlListener.prototype.enterRecord_nature_definition = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#record_nature_definition.
vhdlListener.prototype.exitRecord_nature_definition = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#record_type_definition.
vhdlListener.prototype.enterRecord_type_definition = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#record_type_definition.
vhdlListener.prototype.exitRecord_type_definition = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#relation.
vhdlListener.prototype.enterRelation = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#relation.
vhdlListener.prototype.exitRelation = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#relational_operator.
vhdlListener.prototype.enterRelational_operator = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#relational_operator.
vhdlListener.prototype.exitRelational_operator = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#report_statement.
vhdlListener.prototype.enterReport_statement = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#report_statement.
vhdlListener.prototype.exitReport_statement = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#return_statement.
vhdlListener.prototype.enterReturn_statement = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#return_statement.
vhdlListener.prototype.exitReturn_statement = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#scalar_nature_definition.
vhdlListener.prototype.enterScalar_nature_definition = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#scalar_nature_definition.
vhdlListener.prototype.exitScalar_nature_definition = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#scalar_type_definition.
vhdlListener.prototype.enterScalar_type_definition = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#scalar_type_definition.
vhdlListener.prototype.exitScalar_type_definition = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#secondary_unit.
vhdlListener.prototype.enterSecondary_unit = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#secondary_unit.
vhdlListener.prototype.exitSecondary_unit = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#secondary_unit_declaration.
vhdlListener.prototype.enterSecondary_unit_declaration = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#secondary_unit_declaration.
vhdlListener.prototype.exitSecondary_unit_declaration = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#selected_signal_assignment.
vhdlListener.prototype.enterSelected_signal_assignment = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#selected_signal_assignment.
vhdlListener.prototype.exitSelected_signal_assignment = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#selected_waveforms.
vhdlListener.prototype.enterSelected_waveforms = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#selected_waveforms.
vhdlListener.prototype.exitSelected_waveforms = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#sensitivity_clause.
vhdlListener.prototype.enterSensitivity_clause = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#sensitivity_clause.
vhdlListener.prototype.exitSensitivity_clause = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#sensitivity_list.
vhdlListener.prototype.enterSensitivity_list = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#sensitivity_list.
vhdlListener.prototype.exitSensitivity_list = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#sequence_of_statements.
vhdlListener.prototype.enterSequence_of_statements = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#sequence_of_statements.
vhdlListener.prototype.exitSequence_of_statements = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#sequential_statement.
vhdlListener.prototype.enterSequential_statement = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#sequential_statement.
vhdlListener.prototype.exitSequential_statement = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#shift_expression.
vhdlListener.prototype.enterShift_expression = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#shift_expression.
vhdlListener.prototype.exitShift_expression = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#shift_operator.
vhdlListener.prototype.enterShift_operator = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#shift_operator.
vhdlListener.prototype.exitShift_operator = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#signal_assignment_statement.
vhdlListener.prototype.enterSignal_assignment_statement = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#signal_assignment_statement.
vhdlListener.prototype.exitSignal_assignment_statement = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#signal_declaration.
vhdlListener.prototype.enterSignal_declaration = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#signal_declaration.
vhdlListener.prototype.exitSignal_declaration = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#signal_kind.
vhdlListener.prototype.enterSignal_kind = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#signal_kind.
vhdlListener.prototype.exitSignal_kind = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#signal_list.
vhdlListener.prototype.enterSignal_list = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#signal_list.
vhdlListener.prototype.exitSignal_list = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#signature.
vhdlListener.prototype.enterSignature = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#signature.
vhdlListener.prototype.exitSignature = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#simple_expression.
vhdlListener.prototype.enterSimple_expression = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#simple_expression.
vhdlListener.prototype.exitSimple_expression = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#simple_simultaneous_statement.
vhdlListener.prototype.enterSimple_simultaneous_statement = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#simple_simultaneous_statement.
vhdlListener.prototype.exitSimple_simultaneous_statement = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#simultaneous_alternative.
vhdlListener.prototype.enterSimultaneous_alternative = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#simultaneous_alternative.
vhdlListener.prototype.exitSimultaneous_alternative = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#simultaneous_case_statement.
vhdlListener.prototype.enterSimultaneous_case_statement = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#simultaneous_case_statement.
vhdlListener.prototype.exitSimultaneous_case_statement = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#simultaneous_if_statement.
vhdlListener.prototype.enterSimultaneous_if_statement = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#simultaneous_if_statement.
vhdlListener.prototype.exitSimultaneous_if_statement = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#simultaneous_procedural_statement.
vhdlListener.prototype.enterSimultaneous_procedural_statement = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#simultaneous_procedural_statement.
vhdlListener.prototype.exitSimultaneous_procedural_statement = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#simultaneous_statement.
vhdlListener.prototype.enterSimultaneous_statement = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#simultaneous_statement.
vhdlListener.prototype.exitSimultaneous_statement = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#simultaneous_statement_part.
vhdlListener.prototype.enterSimultaneous_statement_part = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#simultaneous_statement_part.
vhdlListener.prototype.exitSimultaneous_statement_part = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#source_aspect.
vhdlListener.prototype.enterSource_aspect = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#source_aspect.
vhdlListener.prototype.exitSource_aspect = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#source_quantity_declaration.
vhdlListener.prototype.enterSource_quantity_declaration = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#source_quantity_declaration.
vhdlListener.prototype.exitSource_quantity_declaration = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#step_limit_specification.
vhdlListener.prototype.enterStep_limit_specification = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#step_limit_specification.
vhdlListener.prototype.exitStep_limit_specification = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#subnature_declaration.
vhdlListener.prototype.enterSubnature_declaration = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#subnature_declaration.
vhdlListener.prototype.exitSubnature_declaration = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#subnature_indication.
vhdlListener.prototype.enterSubnature_indication = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#subnature_indication.
vhdlListener.prototype.exitSubnature_indication = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#subprogram_body.
vhdlListener.prototype.enterSubprogram_body = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#subprogram_body.
vhdlListener.prototype.exitSubprogram_body = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#subprogram_declaration.
vhdlListener.prototype.enterSubprogram_declaration = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#subprogram_declaration.
vhdlListener.prototype.exitSubprogram_declaration = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#subprogram_declarative_item.
vhdlListener.prototype.enterSubprogram_declarative_item = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#subprogram_declarative_item.
vhdlListener.prototype.exitSubprogram_declarative_item = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#subprogram_declarative_part.
vhdlListener.prototype.enterSubprogram_declarative_part = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#subprogram_declarative_part.
vhdlListener.prototype.exitSubprogram_declarative_part = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#subprogram_kind.
vhdlListener.prototype.enterSubprogram_kind = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#subprogram_kind.
vhdlListener.prototype.exitSubprogram_kind = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#subprogram_specification.
vhdlListener.prototype.enterSubprogram_specification = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#subprogram_specification.
vhdlListener.prototype.exitSubprogram_specification = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#procedure_specification.
vhdlListener.prototype.enterProcedure_specification = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#procedure_specification.
vhdlListener.prototype.exitProcedure_specification = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#function_specification.
vhdlListener.prototype.enterFunction_specification = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#function_specification.
vhdlListener.prototype.exitFunction_specification = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#subprogram_statement_part.
vhdlListener.prototype.enterSubprogram_statement_part = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#subprogram_statement_part.
vhdlListener.prototype.exitSubprogram_statement_part = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#subtype_declaration.
vhdlListener.prototype.enterSubtype_declaration = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#subtype_declaration.
vhdlListener.prototype.exitSubtype_declaration = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#subtype_indication.
vhdlListener.prototype.enterSubtype_indication = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#subtype_indication.
vhdlListener.prototype.exitSubtype_indication = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#suffix.
vhdlListener.prototype.enterSuffix = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#suffix.
vhdlListener.prototype.exitSuffix = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#target.
vhdlListener.prototype.enterTarget = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#target.
vhdlListener.prototype.exitTarget = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#term.
vhdlListener.prototype.enterTerm = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#term.
vhdlListener.prototype.exitTerm = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#terminal_aspect.
vhdlListener.prototype.enterTerminal_aspect = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#terminal_aspect.
vhdlListener.prototype.exitTerminal_aspect = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#terminal_declaration.
vhdlListener.prototype.enterTerminal_declaration = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#terminal_declaration.
vhdlListener.prototype.exitTerminal_declaration = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#through_aspect.
vhdlListener.prototype.enterThrough_aspect = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#through_aspect.
vhdlListener.prototype.exitThrough_aspect = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#timeout_clause.
vhdlListener.prototype.enterTimeout_clause = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#timeout_clause.
vhdlListener.prototype.exitTimeout_clause = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#tolerance_aspect.
vhdlListener.prototype.enterTolerance_aspect = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#tolerance_aspect.
vhdlListener.prototype.exitTolerance_aspect = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#type_declaration.
vhdlListener.prototype.enterType_declaration = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#type_declaration.
vhdlListener.prototype.exitType_declaration = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#type_definition.
vhdlListener.prototype.enterType_definition = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#type_definition.
vhdlListener.prototype.exitType_definition = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#unconstrained_array_definition.
vhdlListener.prototype.enterUnconstrained_array_definition = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#unconstrained_array_definition.
vhdlListener.prototype.exitUnconstrained_array_definition = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#unconstrained_nature_definition.
vhdlListener.prototype.enterUnconstrained_nature_definition = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#unconstrained_nature_definition.
vhdlListener.prototype.exitUnconstrained_nature_definition = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#use_clause.
vhdlListener.prototype.enterUse_clause = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#use_clause.
vhdlListener.prototype.exitUse_clause = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#variable_assignment_statement.
vhdlListener.prototype.enterVariable_assignment_statement = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#variable_assignment_statement.
vhdlListener.prototype.exitVariable_assignment_statement = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#variable_declaration.
vhdlListener.prototype.enterVariable_declaration = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#variable_declaration.
vhdlListener.prototype.exitVariable_declaration = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#wait_statement.
vhdlListener.prototype.enterWait_statement = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#wait_statement.
vhdlListener.prototype.exitWait_statement = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#waveform.
vhdlListener.prototype.enterWaveform = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#waveform.
vhdlListener.prototype.exitWaveform = function(ctx) {
};


// Enter a parse tree produced by vhdlParser#waveform_element.
vhdlListener.prototype.enterWaveform_element = function(ctx) {
};

// Exit a parse tree produced by vhdlParser#waveform_element.
vhdlListener.prototype.exitWaveform_element = function(ctx) {
};



exports.vhdlListener = vhdlListener;
