// HDLfilesys/operation/prjs.js
"use strict";

const files  = require("./files");
const vscode = require("vscode");

function addDevice(opeParam) {
    let propertyParam = files.pullJsonInfo(`${opeParam.rootPath}/property.json`);
    vscode.window.showInputBox({
        password: false,
        ignoreFocusOut: true,
        placeHolder: 'Please input the name of device',
    }).then((Device) => {
        if (propertyParam.properties.Device.enum.indexOf(Device) == -1) {
            propertyParam.properties.Device.enum.push(Device);
            files.pushJsonInfo(`${opeParam.rootPath}/property.json`, propertyParam);
            vscode.window.showInformationMessage(`Add the ${Device} successfully!!!`)
        }
        else {
            vscode.window.showWarningMessage("The device already exists")
        }
    });
}
exports.addDevice = addDevice;

function delDevice(opeParam) {
    let propertyParam = files.pullJsonInfo(`${opeParam.rootPath}/property.json`);
    vscode.window.showQuickPick(propertyParam.properties.Device.enum).then(Device => {
        if (!Device) {
            return;
        }
        for (var index = 0; index < propertyParam.properties.Device.enum.length; index++) {
            if (Device == propertyParam.properties.Device.enum[index]) {
                propertyParam.properties.Device.enum.splice(index, 1);
            }
        }
        files.pushJsonInfo(`${opeParam.rootPath}/property.json`, propertyParam);
        vscode.window.showInformationMessage(`Delete the ${Device} successfully!!!`)
    });
}
exports.delDevice = delDevice;

function overwriteInitProperty(opeParam) {
    let initPropertyPath = `${opeParam.rootPath}/prjInitParam.json`
    const options = {
        preview: false,
        viewColumn: vscode.ViewColumn.Active
    };
    vscode.window.showTextDocument(vscode.Uri.file(initPropertyPath), options);
}
exports.overwriteInitProperty = overwriteInitProperty;

function generatePropertyFile(opeParam) {
    if (!opeParam.prjInfo) {
        vscode.window.showWarningMessage("property file already exists !!!");
        return null;
    }
    vscode.window.showQuickPick(['.vscode', 'root'], { placeHolder: "Where you want to generate?" }).then(function (select) {
        // "Where you want to generate?"
        if (select == ".vscode") {
            opeParam.propertyPath = `${opeParam.workspacePath}/.vscode/property.json`;
        } else if (select == "root") {
            opeParam.propertyPath = `${opeParam.workspacePath}/property.json`;
        }
        files.pushJsonInfo(opeParam.propertyPath, files.pullJsonInfo(opeParam.prjInitParam));
    });
}
exports.generatePropertyFile = generatePropertyFile;
