// HDLparser/src/main/vhdl/vhdlParser.js
const antlr4 = require("antlr4/index");

const vhdlLexer = require("../../vhdl/vhdlLexer");
const vhdlParser = require("../../vhdl/vhdlParser");
const vhdlListener = require("../../vhdl/vhdlListener");

class vhdlErrorListener extends antlr4.error.ErrorListener {
    constructor() {
        super();

        this.diags = [];
        return this;
    }
    syntaxError(offendingSymbol, postion, character, index, msg) {
        let diag = {};
        diag["severity"] = 1;
        diag["range"] = {
            start: { line: postion.line - 1, character: character },
            end: { line: postion.line - 1, character: character }
        };
        diag["message"] = `${index} : ${msg}`;
        diag["source"] = "Digital-IDE";
        this.diags.push(diag);
    }
}

class fastVisitor extends vhdlListener.vhdlListener {
    constructor() {
        super();

        this.modules = [];
        this.params = [];
        this.ports = [];
        this.instances = [];

        this.curModule = null;
        this.errors = [];
    }

    /************ module ************* */
    exitEntity_declaration(ctx) {
        let name = ctx.identifier();
        if (!name) {
            return null;
        }
        name = name[0].getText();
        let [...ports] = this.ports;
        let [...params] = this.params;
        this.curModule = {
            "includes"   : [],
            "languageId" : "vhdl",
            'moduleName' : name,
            'modulePath' : null,
            'ports'      : ports,
            'params'     : params,
            "instances"  : null,
            "start" : {
                line : ctx.start.line - 1,
                character : ctx.start.column,
            },
            "stop" : {
                line : ctx.stop.line - 1,
                character : ctx.stop.column,
            }
        };
        this.ports = [];
        this.params = [];
        this.modules.push(this.curModule);
    }

    enterEntity_header(ctx) {
        this.process_generic_info(ctx);
        this.process_port_info(ctx);
    }

    /************ params ************* */
    process_generic_info(ctx) {
        let params = [];

        // 123 : interface_constant_declaration
        this.child_recursive(ctx, 123, params);
        for (let i = 0; i < params.length; i++) {
            const param = params[i];
            let names = param.identifier_list();
            if (!names) {
                continue;
            }
            names = names.identifier();

            let expression = [];
            // 96 : expression
            this.child_recursive(param, 96, expression);
            for (let i = 0; i < names.length; i++) {
                const name = names[i];
                let item = {
                    "paramName"  : name.getText(),
                    "paramInit"  : expression.length ? expression[0].getText() : null,
                }
                this.params.push(item); 
            }
        }
    }

    /************ ports ************* */
    process_port_info(ctx) {
        let ports = [];

        // 131 : interface_port_declaration
        this.child_recursive(ctx, 131, ports);
        for (let i = 0; i < ports.length; i++) {
            const port = ports[i];
            let names = port.identifier_list();
            if (!names) {
                continue;
            }
            names = names.identifier();

            let width = [];
            this.child_recursive(port, 74, width);
            for (let i = 0; i < names.length; i++) {
                const name = names[i];
                let item = {
                    "portName" : name.getText(),
                    "portWidth": width.length ? width[0].getText() : null,
                }
                this.ports.push(item);
            }
        }
    }

    /************ instantiation ************* */
    enterComponent_instantiation_statement(ctx) {
        let name = ctx.label_colon();
        if (!name) {
            return null;
        }
        name = name.getText().replace(':', '');

        let module = ctx.instantiated_unit();
        if (!module) {
            return null;
        }
        module = module.getText().replace(':', '');

        let curInstance = {
            "name": name,
            "type": "instance",
            "instparams": this.process_generic_info(ctx),
            "instports" : this.process_port_range(ctx),
            "instName": name,         // name
            "instModule": module,   // module
            "instModPath": "",
        };

        curInstance.start = {
            line : ctx.start.line - 1,
            character : ctx.start.column,
        };
        curInstance.stop = {
            line : ctx.stop.line - 1,
            character : ctx.stop.column,
        };

        this.instances.push(curInstance);
    }

    process_generic_range(ctx) {
        let params = ctx.generic_map_aspect();
        if (!params) {
            return null;
        }

        return {
            start : {
                line : params.start.line - 1,
                character : params.start.column,
            },
            stop : {
                line : params.stop.line - 1,
                character : params.stop.column,
            }
        }
    }

    process_port_range(ctx) {
        let ports = ctx.port_map_aspect();
        if (!ports) {
            return null;
        }

        return {
            start : {
                line : ports.start.line - 1,
                character : ports.start.column,
            },
            stop : {
                line : ports.stop.line - 1,
                character : ports.stop.column,
            }
        }
    }
    
    /**
     * @state finish-test
     * @descriptionCn 从当前语法树中递归找到所有满足的规则
     * @param {Object}  ctx   语法树
     * @param {Array}   index 所需要满足的规则序号
     */
    child_recursive(ctx, index, match) {
        if (ctx.ruleIndex == index) {
            match.push(ctx);
        }

        if (!('children' in ctx)) {
            return null;
        }

        
        let children = ctx.children;
        if (!children) {
            return null;
        }

        for (let i = 0; i < children.length; i++) {
            const child = children[i];
            this.child_recursive(child, index, match);
        }
    }
}

class fsmVisitor extends vhdlListener.vhdlListener {
    constructor () {
        super();

        this.isAlways = false;
        this.state_name = null;
    }

    
}

class visitor extends vhdlListener.vhdlListener {
    constructor(option) {
        super();
        this.option = option;
        
        this.curModule = null;
        this.curParent = null;

        this.parents = [];
    }

    enterEntity_declaration(ctx) {
        this.process_identifier(ctx, "entity");
    }

    enterComponent_declaration(ctx) {
        this.process_identifier(ctx, "component");
    }

    enterComponent_instantiation_statement(ctx) {
        let name = ctx.label_colon();
        if (!name) {
            return null;
        }
        name = name.getText().replace(':', '');
        let item = {
            'name' : name,
            'type' : "instance",
            'parent' : null,
            "start" : {
                line : ctx.start.line - 1,
                character : ctx.start.column,
            },
            "end" : {
                line : ctx.stop.line - 1,
                character : ctx.stop.column,
            }
        };
        if (this.option.symbol) {
            this.option.symbol.setSymbolInfo(item);
        }
    }
    
    enterInterface_constant_declaration(ctx) {
        this.process_identifier_list(ctx, "constant");
    }

    enterInterface_variable_declaration(ctx) {
        this.process_identifier_list(ctx, "variable");
    }

    enterInterface_port_declaration(ctx) {
        this.process_identifier_list(ctx, "port");
    }

    enterInterface_signal_declaration(ctx) {
        this.process_identifier_list(ctx, "signal");
    }

    enterInterface_quantity_declaration(ctx) {
        this.process_identifier_list(ctx, "quantity");
    }

    enterSignal_declaration(ctx) {
        this.process_identifier_list(ctx, "signal")
    }

    enterConstant_declaration(ctx) {
        this.process_identifier_list(ctx, "constant")
    }

    enterVariable_declaration(ctx) {
        this.process_identifier_list(ctx, "variable")
    }

    process_identifier_list(ctx, type) {
        let names = ctx.identifier_list();
        if (!names) {
            return null;
        }
        names = names.identifier();
        for (let i = 0; i < names.length; i++) {
            const name = names[i];
            let item = {
                'name' : name.getText(),
                'type' : type,
                'parent' : null,
                "start" : {
                    line : ctx.start.line - 1,
                    character : ctx.start.column,
                },
                "end" : {
                    line : ctx.stop.line - 1,
                    character : ctx.stop.column,
                }
            };
            if (this.option.symbol) {
                this.option.symbol.setSymbolInfo(item);
            }
        }
    }

    process_identifier(ctx, type) {
        let name = ctx.identifier();
        if (!name) {
            return null;
        }
        name = name[0].getText();
        let item = {
            'name' : name,
            'type' : type,
            'parent' : null,
            "start" : {
                line : ctx.start.line - 1,
                character : ctx.start.column,
            },
            "end" : {
                line : ctx.stop.line - 1,
                character : ctx.stop.column,
            }
        };
        if (this.option.symbol) {
            this.option.symbol.setSymbolInfo(item);
        }
    }
}

function parse(code) {
    const inputStream = new antlr4.InputStream(code);
    const lexer = new vhdlLexer.vhdlLexer(inputStream);

    lexer.removeErrorListeners();
    const errorListener = new vhdlErrorListener();
    lexer.addErrorListener(errorListener);

    const tokenStream = new antlr4.CommonTokenStream(lexer);
    const parser = new vhdlParser.vhdlParser(tokenStream);
    parser.removeErrorListeners();
    parser.addErrorListener(errorListener);

    const ast = parser.parse();
    const errors = errorListener.diags;

    return {
        stream : tokenStream, 
        ast : ast, 
        err : errors 
    };
}

function walk(ast, visit) {
    antlr4.tree.ParseTreeWalker.DEFAULT.walk(visit, ast);
    return visit;
}

class parser {
    lint(code) {
        const parseResult = parse(code);
        return parseResult.err;
    }

    /**
     * @state unfinshed - untest
     * @descriptionCn 获取文件的详细属性
     * @param {Object} option   获取时需要操作的选项 
     * @param {Array}  HDLparam 在快解析模式下模块的详细属性存放的数组
     * @returns {Array}  HDLparam 
     */
    getFileParam(option, HDLparam) {
        try {
            // option是必须的
            if (!option) {
                return HDLparam;
            }
    
            const parseResult = parse(option.text);
            console.log(parseResult.err);
    
            // 仅定义为true时运行，null，false时均全运行
            if (option.isFast) { 
                var fastVisit = new fastVisitor();
                const walkResult = walk(parseResult.ast, fastVisit);
                
                for (let index = 0; index < walkResult.modules.length; index++) {
                    const module = walkResult.modules[index];
                    module.instances = walkResult.instances;
                    module.modulePath = option.path;
                    HDLparam.push(module);
                }
    
                return HDLparam;
            }
    
            var visit = new visitor(option);
            walk(parseResult.ast, visit);
        }
        catch (error) {
            console.log(error);
            return HDLparam;
        }
    }

    getFSMParam() {
        try {
            const parseResult = parse(option.text);
            var visit = new fsmVisitor();
            const walkResult = walk(parseResult.ast, visit);
        }
        catch (error) {
            console.log(error);
            return null;
        }
    }
}
module.exports = parser;
