// HDLparser/src/verilog/new/vlogParserListener.js
// Generated from src/verilog/new/vlogParser.g4 by ANTLR 4.7.2
// jshint ignore: start
var antlr4 = require('antlr4/index');

// This class defines a complete listener for a parse tree produced by vlogParser.
function vlogParserListener() {
	antlr4.tree.ParseTreeListener.call(this);
	return this;
}

vlogParserListener.prototype = Object.create(antlr4.tree.ParseTreeListener.prototype);
vlogParserListener.prototype.constructor = vlogParserListener;

// Enter a parse tree produced by vlogParser#parse.
vlogParserListener.prototype.enterParse = function(ctx) {
};

// Exit a parse tree produced by vlogParser#parse.
vlogParserListener.prototype.exitParse = function(ctx) {
};


// Enter a parse tree produced by vlogParser#library_description.
vlogParserListener.prototype.enterLibrary_description = function(ctx) {
};

// Exit a parse tree produced by vlogParser#library_description.
vlogParserListener.prototype.exitLibrary_description = function(ctx) {
};


// Enter a parse tree produced by vlogParser#library_declaration.
vlogParserListener.prototype.enterLibrary_declaration = function(ctx) {
};

// Exit a parse tree produced by vlogParser#library_declaration.
vlogParserListener.prototype.exitLibrary_declaration = function(ctx) {
};


// Enter a parse tree produced by vlogParser#include_directive.
vlogParserListener.prototype.enterInclude_directive = function(ctx) {
};

// Exit a parse tree produced by vlogParser#include_directive.
vlogParserListener.prototype.exitInclude_directive = function(ctx) {
};


// Enter a parse tree produced by vlogParser#file_path_spec.
vlogParserListener.prototype.enterFile_path_spec = function(ctx) {
};

// Exit a parse tree produced by vlogParser#file_path_spec.
vlogParserListener.prototype.exitFile_path_spec = function(ctx) {
};


// Enter a parse tree produced by vlogParser#description.
vlogParserListener.prototype.enterDescription = function(ctx) {
};

// Exit a parse tree produced by vlogParser#description.
vlogParserListener.prototype.exitDescription = function(ctx) {
};


// Enter a parse tree produced by vlogParser#module_declaration.
vlogParserListener.prototype.enterModule_declaration = function(ctx) {
};

// Exit a parse tree produced by vlogParser#module_declaration.
vlogParserListener.prototype.exitModule_declaration = function(ctx) {
};


// Enter a parse tree produced by vlogParser#module_keyword.
vlogParserListener.prototype.enterModule_keyword = function(ctx) {
};

// Exit a parse tree produced by vlogParser#module_keyword.
vlogParserListener.prototype.exitModule_keyword = function(ctx) {
};


// Enter a parse tree produced by vlogParser#module_parameter_port_list.
vlogParserListener.prototype.enterModule_parameter_port_list = function(ctx) {
};

// Exit a parse tree produced by vlogParser#module_parameter_port_list.
vlogParserListener.prototype.exitModule_parameter_port_list = function(ctx) {
};


// Enter a parse tree produced by vlogParser#list_of_ports.
vlogParserListener.prototype.enterList_of_ports = function(ctx) {
};

// Exit a parse tree produced by vlogParser#list_of_ports.
vlogParserListener.prototype.exitList_of_ports = function(ctx) {
};


// Enter a parse tree produced by vlogParser#list_of_port_declarations.
vlogParserListener.prototype.enterList_of_port_declarations = function(ctx) {
};

// Exit a parse tree produced by vlogParser#list_of_port_declarations.
vlogParserListener.prototype.exitList_of_port_declarations = function(ctx) {
};


// Enter a parse tree produced by vlogParser#port.
vlogParserListener.prototype.enterPort = function(ctx) {
};

// Exit a parse tree produced by vlogParser#port.
vlogParserListener.prototype.exitPort = function(ctx) {
};


// Enter a parse tree produced by vlogParser#port_expression.
vlogParserListener.prototype.enterPort_expression = function(ctx) {
};

// Exit a parse tree produced by vlogParser#port_expression.
vlogParserListener.prototype.exitPort_expression = function(ctx) {
};


// Enter a parse tree produced by vlogParser#port_reference.
vlogParserListener.prototype.enterPort_reference = function(ctx) {
};

// Exit a parse tree produced by vlogParser#port_reference.
vlogParserListener.prototype.exitPort_reference = function(ctx) {
};


// Enter a parse tree produced by vlogParser#port_declaration.
vlogParserListener.prototype.enterPort_declaration = function(ctx) {
};

// Exit a parse tree produced by vlogParser#port_declaration.
vlogParserListener.prototype.exitPort_declaration = function(ctx) {
};


// Enter a parse tree produced by vlogParser#module_item.
vlogParserListener.prototype.enterModule_item = function(ctx) {
};

// Exit a parse tree produced by vlogParser#module_item.
vlogParserListener.prototype.exitModule_item = function(ctx) {
};


// Enter a parse tree produced by vlogParser#module_or_generate_item.
vlogParserListener.prototype.enterModule_or_generate_item = function(ctx) {
};

// Exit a parse tree produced by vlogParser#module_or_generate_item.
vlogParserListener.prototype.exitModule_or_generate_item = function(ctx) {
};


// Enter a parse tree produced by vlogParser#module_or_generate_item_declaration.
vlogParserListener.prototype.enterModule_or_generate_item_declaration = function(ctx) {
};

// Exit a parse tree produced by vlogParser#module_or_generate_item_declaration.
vlogParserListener.prototype.exitModule_or_generate_item_declaration = function(ctx) {
};


// Enter a parse tree produced by vlogParser#non_port_module_item.
vlogParserListener.prototype.enterNon_port_module_item = function(ctx) {
};

// Exit a parse tree produced by vlogParser#non_port_module_item.
vlogParserListener.prototype.exitNon_port_module_item = function(ctx) {
};


// Enter a parse tree produced by vlogParser#parameter_override.
vlogParserListener.prototype.enterParameter_override = function(ctx) {
};

// Exit a parse tree produced by vlogParser#parameter_override.
vlogParserListener.prototype.exitParameter_override = function(ctx) {
};


// Enter a parse tree produced by vlogParser#config_declaration.
vlogParserListener.prototype.enterConfig_declaration = function(ctx) {
};

// Exit a parse tree produced by vlogParser#config_declaration.
vlogParserListener.prototype.exitConfig_declaration = function(ctx) {
};


// Enter a parse tree produced by vlogParser#design_statement.
vlogParserListener.prototype.enterDesign_statement = function(ctx) {
};

// Exit a parse tree produced by vlogParser#design_statement.
vlogParserListener.prototype.exitDesign_statement = function(ctx) {
};


// Enter a parse tree produced by vlogParser#config_rule_statement.
vlogParserListener.prototype.enterConfig_rule_statement = function(ctx) {
};

// Exit a parse tree produced by vlogParser#config_rule_statement.
vlogParserListener.prototype.exitConfig_rule_statement = function(ctx) {
};


// Enter a parse tree produced by vlogParser#default_clause.
vlogParserListener.prototype.enterDefault_clause = function(ctx) {
};

// Exit a parse tree produced by vlogParser#default_clause.
vlogParserListener.prototype.exitDefault_clause = function(ctx) {
};


// Enter a parse tree produced by vlogParser#inst_clause.
vlogParserListener.prototype.enterInst_clause = function(ctx) {
};

// Exit a parse tree produced by vlogParser#inst_clause.
vlogParserListener.prototype.exitInst_clause = function(ctx) {
};


// Enter a parse tree produced by vlogParser#inst_name.
vlogParserListener.prototype.enterInst_name = function(ctx) {
};

// Exit a parse tree produced by vlogParser#inst_name.
vlogParserListener.prototype.exitInst_name = function(ctx) {
};


// Enter a parse tree produced by vlogParser#cell_clause.
vlogParserListener.prototype.enterCell_clause = function(ctx) {
};

// Exit a parse tree produced by vlogParser#cell_clause.
vlogParserListener.prototype.exitCell_clause = function(ctx) {
};


// Enter a parse tree produced by vlogParser#liblist_clause.
vlogParserListener.prototype.enterLiblist_clause = function(ctx) {
};

// Exit a parse tree produced by vlogParser#liblist_clause.
vlogParserListener.prototype.exitLiblist_clause = function(ctx) {
};


// Enter a parse tree produced by vlogParser#use_clause.
vlogParserListener.prototype.enterUse_clause = function(ctx) {
};

// Exit a parse tree produced by vlogParser#use_clause.
vlogParserListener.prototype.exitUse_clause = function(ctx) {
};


// Enter a parse tree produced by vlogParser#local_parameter_declaration.
vlogParserListener.prototype.enterLocal_parameter_declaration = function(ctx) {
};

// Exit a parse tree produced by vlogParser#local_parameter_declaration.
vlogParserListener.prototype.exitLocal_parameter_declaration = function(ctx) {
};


// Enter a parse tree produced by vlogParser#parameter_declaration.
vlogParserListener.prototype.enterParameter_declaration = function(ctx) {
};

// Exit a parse tree produced by vlogParser#parameter_declaration.
vlogParserListener.prototype.exitParameter_declaration = function(ctx) {
};


// Enter a parse tree produced by vlogParser#specparam_declaration.
vlogParserListener.prototype.enterSpecparam_declaration = function(ctx) {
};

// Exit a parse tree produced by vlogParser#specparam_declaration.
vlogParserListener.prototype.exitSpecparam_declaration = function(ctx) {
};


// Enter a parse tree produced by vlogParser#parameter_type.
vlogParserListener.prototype.enterParameter_type = function(ctx) {
};

// Exit a parse tree produced by vlogParser#parameter_type.
vlogParserListener.prototype.exitParameter_type = function(ctx) {
};


// Enter a parse tree produced by vlogParser#inout_declaration.
vlogParserListener.prototype.enterInout_declaration = function(ctx) {
};

// Exit a parse tree produced by vlogParser#inout_declaration.
vlogParserListener.prototype.exitInout_declaration = function(ctx) {
};


// Enter a parse tree produced by vlogParser#input_declaration.
vlogParserListener.prototype.enterInput_declaration = function(ctx) {
};

// Exit a parse tree produced by vlogParser#input_declaration.
vlogParserListener.prototype.exitInput_declaration = function(ctx) {
};


// Enter a parse tree produced by vlogParser#output_declaration.
vlogParserListener.prototype.enterOutput_declaration = function(ctx) {
};

// Exit a parse tree produced by vlogParser#output_declaration.
vlogParserListener.prototype.exitOutput_declaration = function(ctx) {
};


// Enter a parse tree produced by vlogParser#event_declaration.
vlogParserListener.prototype.enterEvent_declaration = function(ctx) {
};

// Exit a parse tree produced by vlogParser#event_declaration.
vlogParserListener.prototype.exitEvent_declaration = function(ctx) {
};


// Enter a parse tree produced by vlogParser#integer_declaration.
vlogParserListener.prototype.enterInteger_declaration = function(ctx) {
};

// Exit a parse tree produced by vlogParser#integer_declaration.
vlogParserListener.prototype.exitInteger_declaration = function(ctx) {
};


// Enter a parse tree produced by vlogParser#net_declaration.
vlogParserListener.prototype.enterNet_declaration = function(ctx) {
};

// Exit a parse tree produced by vlogParser#net_declaration.
vlogParserListener.prototype.exitNet_declaration = function(ctx) {
};


// Enter a parse tree produced by vlogParser#real_declaration.
vlogParserListener.prototype.enterReal_declaration = function(ctx) {
};

// Exit a parse tree produced by vlogParser#real_declaration.
vlogParserListener.prototype.exitReal_declaration = function(ctx) {
};


// Enter a parse tree produced by vlogParser#realtime_declaration.
vlogParserListener.prototype.enterRealtime_declaration = function(ctx) {
};

// Exit a parse tree produced by vlogParser#realtime_declaration.
vlogParserListener.prototype.exitRealtime_declaration = function(ctx) {
};


// Enter a parse tree produced by vlogParser#reg_declaration.
vlogParserListener.prototype.enterReg_declaration = function(ctx) {
};

// Exit a parse tree produced by vlogParser#reg_declaration.
vlogParserListener.prototype.exitReg_declaration = function(ctx) {
};


// Enter a parse tree produced by vlogParser#time_declaration.
vlogParserListener.prototype.enterTime_declaration = function(ctx) {
};

// Exit a parse tree produced by vlogParser#time_declaration.
vlogParserListener.prototype.exitTime_declaration = function(ctx) {
};


// Enter a parse tree produced by vlogParser#net_type.
vlogParserListener.prototype.enterNet_type = function(ctx) {
};

// Exit a parse tree produced by vlogParser#net_type.
vlogParserListener.prototype.exitNet_type = function(ctx) {
};


// Enter a parse tree produced by vlogParser#output_variable_type.
vlogParserListener.prototype.enterOutput_variable_type = function(ctx) {
};

// Exit a parse tree produced by vlogParser#output_variable_type.
vlogParserListener.prototype.exitOutput_variable_type = function(ctx) {
};


// Enter a parse tree produced by vlogParser#real_type.
vlogParserListener.prototype.enterReal_type = function(ctx) {
};

// Exit a parse tree produced by vlogParser#real_type.
vlogParserListener.prototype.exitReal_type = function(ctx) {
};


// Enter a parse tree produced by vlogParser#variable_type.
vlogParserListener.prototype.enterVariable_type = function(ctx) {
};

// Exit a parse tree produced by vlogParser#variable_type.
vlogParserListener.prototype.exitVariable_type = function(ctx) {
};


// Enter a parse tree produced by vlogParser#drive_strength.
vlogParserListener.prototype.enterDrive_strength = function(ctx) {
};

// Exit a parse tree produced by vlogParser#drive_strength.
vlogParserListener.prototype.exitDrive_strength = function(ctx) {
};


// Enter a parse tree produced by vlogParser#strength0.
vlogParserListener.prototype.enterStrength0 = function(ctx) {
};

// Exit a parse tree produced by vlogParser#strength0.
vlogParserListener.prototype.exitStrength0 = function(ctx) {
};


// Enter a parse tree produced by vlogParser#strength1.
vlogParserListener.prototype.enterStrength1 = function(ctx) {
};

// Exit a parse tree produced by vlogParser#strength1.
vlogParserListener.prototype.exitStrength1 = function(ctx) {
};


// Enter a parse tree produced by vlogParser#charge_strength.
vlogParserListener.prototype.enterCharge_strength = function(ctx) {
};

// Exit a parse tree produced by vlogParser#charge_strength.
vlogParserListener.prototype.exitCharge_strength = function(ctx) {
};


// Enter a parse tree produced by vlogParser#delay3.
vlogParserListener.prototype.enterDelay3 = function(ctx) {
};

// Exit a parse tree produced by vlogParser#delay3.
vlogParserListener.prototype.exitDelay3 = function(ctx) {
};


// Enter a parse tree produced by vlogParser#delay2.
vlogParserListener.prototype.enterDelay2 = function(ctx) {
};

// Exit a parse tree produced by vlogParser#delay2.
vlogParserListener.prototype.exitDelay2 = function(ctx) {
};


// Enter a parse tree produced by vlogParser#delay_value.
vlogParserListener.prototype.enterDelay_value = function(ctx) {
};

// Exit a parse tree produced by vlogParser#delay_value.
vlogParserListener.prototype.exitDelay_value = function(ctx) {
};


// Enter a parse tree produced by vlogParser#list_of_defparam_assignments.
vlogParserListener.prototype.enterList_of_defparam_assignments = function(ctx) {
};

// Exit a parse tree produced by vlogParser#list_of_defparam_assignments.
vlogParserListener.prototype.exitList_of_defparam_assignments = function(ctx) {
};


// Enter a parse tree produced by vlogParser#list_of_event_identifiers.
vlogParserListener.prototype.enterList_of_event_identifiers = function(ctx) {
};

// Exit a parse tree produced by vlogParser#list_of_event_identifiers.
vlogParserListener.prototype.exitList_of_event_identifiers = function(ctx) {
};


// Enter a parse tree produced by vlogParser#list_of_net_decl_assignments.
vlogParserListener.prototype.enterList_of_net_decl_assignments = function(ctx) {
};

// Exit a parse tree produced by vlogParser#list_of_net_decl_assignments.
vlogParserListener.prototype.exitList_of_net_decl_assignments = function(ctx) {
};


// Enter a parse tree produced by vlogParser#list_of_net_identifiers.
vlogParserListener.prototype.enterList_of_net_identifiers = function(ctx) {
};

// Exit a parse tree produced by vlogParser#list_of_net_identifiers.
vlogParserListener.prototype.exitList_of_net_identifiers = function(ctx) {
};


// Enter a parse tree produced by vlogParser#list_of_param_assignments.
vlogParserListener.prototype.enterList_of_param_assignments = function(ctx) {
};

// Exit a parse tree produced by vlogParser#list_of_param_assignments.
vlogParserListener.prototype.exitList_of_param_assignments = function(ctx) {
};


// Enter a parse tree produced by vlogParser#list_of_port_identifiers.
vlogParserListener.prototype.enterList_of_port_identifiers = function(ctx) {
};

// Exit a parse tree produced by vlogParser#list_of_port_identifiers.
vlogParserListener.prototype.exitList_of_port_identifiers = function(ctx) {
};


// Enter a parse tree produced by vlogParser#list_of_real_identifiers.
vlogParserListener.prototype.enterList_of_real_identifiers = function(ctx) {
};

// Exit a parse tree produced by vlogParser#list_of_real_identifiers.
vlogParserListener.prototype.exitList_of_real_identifiers = function(ctx) {
};


// Enter a parse tree produced by vlogParser#list_of_specparam_assignments.
vlogParserListener.prototype.enterList_of_specparam_assignments = function(ctx) {
};

// Exit a parse tree produced by vlogParser#list_of_specparam_assignments.
vlogParserListener.prototype.exitList_of_specparam_assignments = function(ctx) {
};


// Enter a parse tree produced by vlogParser#list_of_variable_identifiers.
vlogParserListener.prototype.enterList_of_variable_identifiers = function(ctx) {
};

// Exit a parse tree produced by vlogParser#list_of_variable_identifiers.
vlogParserListener.prototype.exitList_of_variable_identifiers = function(ctx) {
};


// Enter a parse tree produced by vlogParser#list_of_variable_port_identifiers.
vlogParserListener.prototype.enterList_of_variable_port_identifiers = function(ctx) {
};

// Exit a parse tree produced by vlogParser#list_of_variable_port_identifiers.
vlogParserListener.prototype.exitList_of_variable_port_identifiers = function(ctx) {
};


// Enter a parse tree produced by vlogParser#defparam_assignment.
vlogParserListener.prototype.enterDefparam_assignment = function(ctx) {
};

// Exit a parse tree produced by vlogParser#defparam_assignment.
vlogParserListener.prototype.exitDefparam_assignment = function(ctx) {
};


// Enter a parse tree produced by vlogParser#net_decl_assignment.
vlogParserListener.prototype.enterNet_decl_assignment = function(ctx) {
};

// Exit a parse tree produced by vlogParser#net_decl_assignment.
vlogParserListener.prototype.exitNet_decl_assignment = function(ctx) {
};


// Enter a parse tree produced by vlogParser#param_assignment.
vlogParserListener.prototype.enterParam_assignment = function(ctx) {
};

// Exit a parse tree produced by vlogParser#param_assignment.
vlogParserListener.prototype.exitParam_assignment = function(ctx) {
};


// Enter a parse tree produced by vlogParser#specparam_assignment.
vlogParserListener.prototype.enterSpecparam_assignment = function(ctx) {
};

// Exit a parse tree produced by vlogParser#specparam_assignment.
vlogParserListener.prototype.exitSpecparam_assignment = function(ctx) {
};


// Enter a parse tree produced by vlogParser#pulse_control_specparam.
vlogParserListener.prototype.enterPulse_control_specparam = function(ctx) {
};

// Exit a parse tree produced by vlogParser#pulse_control_specparam.
vlogParserListener.prototype.exitPulse_control_specparam = function(ctx) {
};


// Enter a parse tree produced by vlogParser#error_limit_value.
vlogParserListener.prototype.enterError_limit_value = function(ctx) {
};

// Exit a parse tree produced by vlogParser#error_limit_value.
vlogParserListener.prototype.exitError_limit_value = function(ctx) {
};


// Enter a parse tree produced by vlogParser#reject_limit_value.
vlogParserListener.prototype.enterReject_limit_value = function(ctx) {
};

// Exit a parse tree produced by vlogParser#reject_limit_value.
vlogParserListener.prototype.exitReject_limit_value = function(ctx) {
};


// Enter a parse tree produced by vlogParser#limit_value.
vlogParserListener.prototype.enterLimit_value = function(ctx) {
};

// Exit a parse tree produced by vlogParser#limit_value.
vlogParserListener.prototype.exitLimit_value = function(ctx) {
};


// Enter a parse tree produced by vlogParser#dimension.
vlogParserListener.prototype.enterDimension = function(ctx) {
};

// Exit a parse tree produced by vlogParser#dimension.
vlogParserListener.prototype.exitDimension = function(ctx) {
};


// Enter a parse tree produced by vlogParser#range_.
vlogParserListener.prototype.enterRange_ = function(ctx) {
};

// Exit a parse tree produced by vlogParser#range_.
vlogParserListener.prototype.exitRange_ = function(ctx) {
};


// Enter a parse tree produced by vlogParser#function_declaration.
vlogParserListener.prototype.enterFunction_declaration = function(ctx) {
};

// Exit a parse tree produced by vlogParser#function_declaration.
vlogParserListener.prototype.exitFunction_declaration = function(ctx) {
};


// Enter a parse tree produced by vlogParser#function_item_declaration.
vlogParserListener.prototype.enterFunction_item_declaration = function(ctx) {
};

// Exit a parse tree produced by vlogParser#function_item_declaration.
vlogParserListener.prototype.exitFunction_item_declaration = function(ctx) {
};


// Enter a parse tree produced by vlogParser#function_port_list.
vlogParserListener.prototype.enterFunction_port_list = function(ctx) {
};

// Exit a parse tree produced by vlogParser#function_port_list.
vlogParserListener.prototype.exitFunction_port_list = function(ctx) {
};


// Enter a parse tree produced by vlogParser#function_range_or_type.
vlogParserListener.prototype.enterFunction_range_or_type = function(ctx) {
};

// Exit a parse tree produced by vlogParser#function_range_or_type.
vlogParserListener.prototype.exitFunction_range_or_type = function(ctx) {
};


// Enter a parse tree produced by vlogParser#task_declaration.
vlogParserListener.prototype.enterTask_declaration = function(ctx) {
};

// Exit a parse tree produced by vlogParser#task_declaration.
vlogParserListener.prototype.exitTask_declaration = function(ctx) {
};


// Enter a parse tree produced by vlogParser#task_item_declaration.
vlogParserListener.prototype.enterTask_item_declaration = function(ctx) {
};

// Exit a parse tree produced by vlogParser#task_item_declaration.
vlogParserListener.prototype.exitTask_item_declaration = function(ctx) {
};


// Enter a parse tree produced by vlogParser#task_port_list.
vlogParserListener.prototype.enterTask_port_list = function(ctx) {
};

// Exit a parse tree produced by vlogParser#task_port_list.
vlogParserListener.prototype.exitTask_port_list = function(ctx) {
};


// Enter a parse tree produced by vlogParser#task_port_item.
vlogParserListener.prototype.enterTask_port_item = function(ctx) {
};

// Exit a parse tree produced by vlogParser#task_port_item.
vlogParserListener.prototype.exitTask_port_item = function(ctx) {
};


// Enter a parse tree produced by vlogParser#tf_input_declaration.
vlogParserListener.prototype.enterTf_input_declaration = function(ctx) {
};

// Exit a parse tree produced by vlogParser#tf_input_declaration.
vlogParserListener.prototype.exitTf_input_declaration = function(ctx) {
};


// Enter a parse tree produced by vlogParser#tf_output_declaration.
vlogParserListener.prototype.enterTf_output_declaration = function(ctx) {
};

// Exit a parse tree produced by vlogParser#tf_output_declaration.
vlogParserListener.prototype.exitTf_output_declaration = function(ctx) {
};


// Enter a parse tree produced by vlogParser#tf_inout_declaration.
vlogParserListener.prototype.enterTf_inout_declaration = function(ctx) {
};

// Exit a parse tree produced by vlogParser#tf_inout_declaration.
vlogParserListener.prototype.exitTf_inout_declaration = function(ctx) {
};


// Enter a parse tree produced by vlogParser#task_port_type.
vlogParserListener.prototype.enterTask_port_type = function(ctx) {
};

// Exit a parse tree produced by vlogParser#task_port_type.
vlogParserListener.prototype.exitTask_port_type = function(ctx) {
};


// Enter a parse tree produced by vlogParser#block_item_declaration.
vlogParserListener.prototype.enterBlock_item_declaration = function(ctx) {
};

// Exit a parse tree produced by vlogParser#block_item_declaration.
vlogParserListener.prototype.exitBlock_item_declaration = function(ctx) {
};


// Enter a parse tree produced by vlogParser#list_of_block_variable_identifiers.
vlogParserListener.prototype.enterList_of_block_variable_identifiers = function(ctx) {
};

// Exit a parse tree produced by vlogParser#list_of_block_variable_identifiers.
vlogParserListener.prototype.exitList_of_block_variable_identifiers = function(ctx) {
};


// Enter a parse tree produced by vlogParser#list_of_block_real_identifiers.
vlogParserListener.prototype.enterList_of_block_real_identifiers = function(ctx) {
};

// Exit a parse tree produced by vlogParser#list_of_block_real_identifiers.
vlogParserListener.prototype.exitList_of_block_real_identifiers = function(ctx) {
};


// Enter a parse tree produced by vlogParser#block_variable_type.
vlogParserListener.prototype.enterBlock_variable_type = function(ctx) {
};

// Exit a parse tree produced by vlogParser#block_variable_type.
vlogParserListener.prototype.exitBlock_variable_type = function(ctx) {
};


// Enter a parse tree produced by vlogParser#block_real_type.
vlogParserListener.prototype.enterBlock_real_type = function(ctx) {
};

// Exit a parse tree produced by vlogParser#block_real_type.
vlogParserListener.prototype.exitBlock_real_type = function(ctx) {
};


// Enter a parse tree produced by vlogParser#gate_instantiation.
vlogParserListener.prototype.enterGate_instantiation = function(ctx) {
};

// Exit a parse tree produced by vlogParser#gate_instantiation.
vlogParserListener.prototype.exitGate_instantiation = function(ctx) {
};


// Enter a parse tree produced by vlogParser#cmos_switch_instance.
vlogParserListener.prototype.enterCmos_switch_instance = function(ctx) {
};

// Exit a parse tree produced by vlogParser#cmos_switch_instance.
vlogParserListener.prototype.exitCmos_switch_instance = function(ctx) {
};


// Enter a parse tree produced by vlogParser#enable_gate_instance.
vlogParserListener.prototype.enterEnable_gate_instance = function(ctx) {
};

// Exit a parse tree produced by vlogParser#enable_gate_instance.
vlogParserListener.prototype.exitEnable_gate_instance = function(ctx) {
};


// Enter a parse tree produced by vlogParser#mos_switch_instance.
vlogParserListener.prototype.enterMos_switch_instance = function(ctx) {
};

// Exit a parse tree produced by vlogParser#mos_switch_instance.
vlogParserListener.prototype.exitMos_switch_instance = function(ctx) {
};


// Enter a parse tree produced by vlogParser#n_input_gate_instance.
vlogParserListener.prototype.enterN_input_gate_instance = function(ctx) {
};

// Exit a parse tree produced by vlogParser#n_input_gate_instance.
vlogParserListener.prototype.exitN_input_gate_instance = function(ctx) {
};


// Enter a parse tree produced by vlogParser#n_output_gate_instance.
vlogParserListener.prototype.enterN_output_gate_instance = function(ctx) {
};

// Exit a parse tree produced by vlogParser#n_output_gate_instance.
vlogParserListener.prototype.exitN_output_gate_instance = function(ctx) {
};


// Enter a parse tree produced by vlogParser#pass_switch_instance.
vlogParserListener.prototype.enterPass_switch_instance = function(ctx) {
};

// Exit a parse tree produced by vlogParser#pass_switch_instance.
vlogParserListener.prototype.exitPass_switch_instance = function(ctx) {
};


// Enter a parse tree produced by vlogParser#pass_enable_switch_instance.
vlogParserListener.prototype.enterPass_enable_switch_instance = function(ctx) {
};

// Exit a parse tree produced by vlogParser#pass_enable_switch_instance.
vlogParserListener.prototype.exitPass_enable_switch_instance = function(ctx) {
};


// Enter a parse tree produced by vlogParser#pull_gate_instance.
vlogParserListener.prototype.enterPull_gate_instance = function(ctx) {
};

// Exit a parse tree produced by vlogParser#pull_gate_instance.
vlogParserListener.prototype.exitPull_gate_instance = function(ctx) {
};


// Enter a parse tree produced by vlogParser#name_of_gate_instance.
vlogParserListener.prototype.enterName_of_gate_instance = function(ctx) {
};

// Exit a parse tree produced by vlogParser#name_of_gate_instance.
vlogParserListener.prototype.exitName_of_gate_instance = function(ctx) {
};


// Enter a parse tree produced by vlogParser#pulldown_strength.
vlogParserListener.prototype.enterPulldown_strength = function(ctx) {
};

// Exit a parse tree produced by vlogParser#pulldown_strength.
vlogParserListener.prototype.exitPulldown_strength = function(ctx) {
};


// Enter a parse tree produced by vlogParser#pullup_strength.
vlogParserListener.prototype.enterPullup_strength = function(ctx) {
};

// Exit a parse tree produced by vlogParser#pullup_strength.
vlogParserListener.prototype.exitPullup_strength = function(ctx) {
};


// Enter a parse tree produced by vlogParser#enable_terminal.
vlogParserListener.prototype.enterEnable_terminal = function(ctx) {
};

// Exit a parse tree produced by vlogParser#enable_terminal.
vlogParserListener.prototype.exitEnable_terminal = function(ctx) {
};


// Enter a parse tree produced by vlogParser#inout_terminal.
vlogParserListener.prototype.enterInout_terminal = function(ctx) {
};

// Exit a parse tree produced by vlogParser#inout_terminal.
vlogParserListener.prototype.exitInout_terminal = function(ctx) {
};


// Enter a parse tree produced by vlogParser#input_terminal.
vlogParserListener.prototype.enterInput_terminal = function(ctx) {
};

// Exit a parse tree produced by vlogParser#input_terminal.
vlogParserListener.prototype.exitInput_terminal = function(ctx) {
};


// Enter a parse tree produced by vlogParser#ncontrol_terminal.
vlogParserListener.prototype.enterNcontrol_terminal = function(ctx) {
};

// Exit a parse tree produced by vlogParser#ncontrol_terminal.
vlogParserListener.prototype.exitNcontrol_terminal = function(ctx) {
};


// Enter a parse tree produced by vlogParser#output_terminal.
vlogParserListener.prototype.enterOutput_terminal = function(ctx) {
};

// Exit a parse tree produced by vlogParser#output_terminal.
vlogParserListener.prototype.exitOutput_terminal = function(ctx) {
};


// Enter a parse tree produced by vlogParser#pcontrol_terminal.
vlogParserListener.prototype.enterPcontrol_terminal = function(ctx) {
};

// Exit a parse tree produced by vlogParser#pcontrol_terminal.
vlogParserListener.prototype.exitPcontrol_terminal = function(ctx) {
};


// Enter a parse tree produced by vlogParser#cmos_switchtype.
vlogParserListener.prototype.enterCmos_switchtype = function(ctx) {
};

// Exit a parse tree produced by vlogParser#cmos_switchtype.
vlogParserListener.prototype.exitCmos_switchtype = function(ctx) {
};


// Enter a parse tree produced by vlogParser#enable_gatetype.
vlogParserListener.prototype.enterEnable_gatetype = function(ctx) {
};

// Exit a parse tree produced by vlogParser#enable_gatetype.
vlogParserListener.prototype.exitEnable_gatetype = function(ctx) {
};


// Enter a parse tree produced by vlogParser#mos_switchtype.
vlogParserListener.prototype.enterMos_switchtype = function(ctx) {
};

// Exit a parse tree produced by vlogParser#mos_switchtype.
vlogParserListener.prototype.exitMos_switchtype = function(ctx) {
};


// Enter a parse tree produced by vlogParser#n_input_gatetype.
vlogParserListener.prototype.enterN_input_gatetype = function(ctx) {
};

// Exit a parse tree produced by vlogParser#n_input_gatetype.
vlogParserListener.prototype.exitN_input_gatetype = function(ctx) {
};


// Enter a parse tree produced by vlogParser#n_output_gatetype.
vlogParserListener.prototype.enterN_output_gatetype = function(ctx) {
};

// Exit a parse tree produced by vlogParser#n_output_gatetype.
vlogParserListener.prototype.exitN_output_gatetype = function(ctx) {
};


// Enter a parse tree produced by vlogParser#pass_en_switchtype.
vlogParserListener.prototype.enterPass_en_switchtype = function(ctx) {
};

// Exit a parse tree produced by vlogParser#pass_en_switchtype.
vlogParserListener.prototype.exitPass_en_switchtype = function(ctx) {
};


// Enter a parse tree produced by vlogParser#pass_switchtype.
vlogParserListener.prototype.enterPass_switchtype = function(ctx) {
};

// Exit a parse tree produced by vlogParser#pass_switchtype.
vlogParserListener.prototype.exitPass_switchtype = function(ctx) {
};


// Enter a parse tree produced by vlogParser#module_instantiation.
vlogParserListener.prototype.enterModule_instantiation = function(ctx) {
};

// Exit a parse tree produced by vlogParser#module_instantiation.
vlogParserListener.prototype.exitModule_instantiation = function(ctx) {
};


// Enter a parse tree produced by vlogParser#parameter_value_assignment.
vlogParserListener.prototype.enterParameter_value_assignment = function(ctx) {
};

// Exit a parse tree produced by vlogParser#parameter_value_assignment.
vlogParserListener.prototype.exitParameter_value_assignment = function(ctx) {
};


// Enter a parse tree produced by vlogParser#list_of_parameter_assignments.
vlogParserListener.prototype.enterList_of_parameter_assignments = function(ctx) {
};

// Exit a parse tree produced by vlogParser#list_of_parameter_assignments.
vlogParserListener.prototype.exitList_of_parameter_assignments = function(ctx) {
};


// Enter a parse tree produced by vlogParser#ordered_parameter_assignment.
vlogParserListener.prototype.enterOrdered_parameter_assignment = function(ctx) {
};

// Exit a parse tree produced by vlogParser#ordered_parameter_assignment.
vlogParserListener.prototype.exitOrdered_parameter_assignment = function(ctx) {
};


// Enter a parse tree produced by vlogParser#named_parameter_assignment.
vlogParserListener.prototype.enterNamed_parameter_assignment = function(ctx) {
};

// Exit a parse tree produced by vlogParser#named_parameter_assignment.
vlogParserListener.prototype.exitNamed_parameter_assignment = function(ctx) {
};


// Enter a parse tree produced by vlogParser#module_instance.
vlogParserListener.prototype.enterModule_instance = function(ctx) {
};

// Exit a parse tree produced by vlogParser#module_instance.
vlogParserListener.prototype.exitModule_instance = function(ctx) {
};


// Enter a parse tree produced by vlogParser#name_of_module_instance.
vlogParserListener.prototype.enterName_of_module_instance = function(ctx) {
};

// Exit a parse tree produced by vlogParser#name_of_module_instance.
vlogParserListener.prototype.exitName_of_module_instance = function(ctx) {
};


// Enter a parse tree produced by vlogParser#list_of_port_connections.
vlogParserListener.prototype.enterList_of_port_connections = function(ctx) {
};

// Exit a parse tree produced by vlogParser#list_of_port_connections.
vlogParserListener.prototype.exitList_of_port_connections = function(ctx) {
};


// Enter a parse tree produced by vlogParser#ordered_port_connection.
vlogParserListener.prototype.enterOrdered_port_connection = function(ctx) {
};

// Exit a parse tree produced by vlogParser#ordered_port_connection.
vlogParserListener.prototype.exitOrdered_port_connection = function(ctx) {
};


// Enter a parse tree produced by vlogParser#named_port_connection.
vlogParserListener.prototype.enterNamed_port_connection = function(ctx) {
};

// Exit a parse tree produced by vlogParser#named_port_connection.
vlogParserListener.prototype.exitNamed_port_connection = function(ctx) {
};


// Enter a parse tree produced by vlogParser#generate_region.
vlogParserListener.prototype.enterGenerate_region = function(ctx) {
};

// Exit a parse tree produced by vlogParser#generate_region.
vlogParserListener.prototype.exitGenerate_region = function(ctx) {
};


// Enter a parse tree produced by vlogParser#genvar_declaration.
vlogParserListener.prototype.enterGenvar_declaration = function(ctx) {
};

// Exit a parse tree produced by vlogParser#genvar_declaration.
vlogParserListener.prototype.exitGenvar_declaration = function(ctx) {
};


// Enter a parse tree produced by vlogParser#list_of_genvar_identifiers.
vlogParserListener.prototype.enterList_of_genvar_identifiers = function(ctx) {
};

// Exit a parse tree produced by vlogParser#list_of_genvar_identifiers.
vlogParserListener.prototype.exitList_of_genvar_identifiers = function(ctx) {
};


// Enter a parse tree produced by vlogParser#loop_generate_construct.
vlogParserListener.prototype.enterLoop_generate_construct = function(ctx) {
};

// Exit a parse tree produced by vlogParser#loop_generate_construct.
vlogParserListener.prototype.exitLoop_generate_construct = function(ctx) {
};


// Enter a parse tree produced by vlogParser#genvar_initialization.
vlogParserListener.prototype.enterGenvar_initialization = function(ctx) {
};

// Exit a parse tree produced by vlogParser#genvar_initialization.
vlogParserListener.prototype.exitGenvar_initialization = function(ctx) {
};


// Enter a parse tree produced by vlogParser#genvar_expression.
vlogParserListener.prototype.enterGenvar_expression = function(ctx) {
};

// Exit a parse tree produced by vlogParser#genvar_expression.
vlogParserListener.prototype.exitGenvar_expression = function(ctx) {
};


// Enter a parse tree produced by vlogParser#genvar_iteration.
vlogParserListener.prototype.enterGenvar_iteration = function(ctx) {
};

// Exit a parse tree produced by vlogParser#genvar_iteration.
vlogParserListener.prototype.exitGenvar_iteration = function(ctx) {
};


// Enter a parse tree produced by vlogParser#genvar_primary.
vlogParserListener.prototype.enterGenvar_primary = function(ctx) {
};

// Exit a parse tree produced by vlogParser#genvar_primary.
vlogParserListener.prototype.exitGenvar_primary = function(ctx) {
};


// Enter a parse tree produced by vlogParser#conditional_generate_construct.
vlogParserListener.prototype.enterConditional_generate_construct = function(ctx) {
};

// Exit a parse tree produced by vlogParser#conditional_generate_construct.
vlogParserListener.prototype.exitConditional_generate_construct = function(ctx) {
};


// Enter a parse tree produced by vlogParser#if_generate_construct.
vlogParserListener.prototype.enterIf_generate_construct = function(ctx) {
};

// Exit a parse tree produced by vlogParser#if_generate_construct.
vlogParserListener.prototype.exitIf_generate_construct = function(ctx) {
};


// Enter a parse tree produced by vlogParser#case_generate_construct.
vlogParserListener.prototype.enterCase_generate_construct = function(ctx) {
};

// Exit a parse tree produced by vlogParser#case_generate_construct.
vlogParserListener.prototype.exitCase_generate_construct = function(ctx) {
};


// Enter a parse tree produced by vlogParser#case_generate_item.
vlogParserListener.prototype.enterCase_generate_item = function(ctx) {
};

// Exit a parse tree produced by vlogParser#case_generate_item.
vlogParserListener.prototype.exitCase_generate_item = function(ctx) {
};


// Enter a parse tree produced by vlogParser#generate_block.
vlogParserListener.prototype.enterGenerate_block = function(ctx) {
};

// Exit a parse tree produced by vlogParser#generate_block.
vlogParserListener.prototype.exitGenerate_block = function(ctx) {
};


// Enter a parse tree produced by vlogParser#generate_block_or_null.
vlogParserListener.prototype.enterGenerate_block_or_null = function(ctx) {
};

// Exit a parse tree produced by vlogParser#generate_block_or_null.
vlogParserListener.prototype.exitGenerate_block_or_null = function(ctx) {
};


// Enter a parse tree produced by vlogParser#udp_declaration.
vlogParserListener.prototype.enterUdp_declaration = function(ctx) {
};

// Exit a parse tree produced by vlogParser#udp_declaration.
vlogParserListener.prototype.exitUdp_declaration = function(ctx) {
};


// Enter a parse tree produced by vlogParser#udp_port_list.
vlogParserListener.prototype.enterUdp_port_list = function(ctx) {
};

// Exit a parse tree produced by vlogParser#udp_port_list.
vlogParserListener.prototype.exitUdp_port_list = function(ctx) {
};


// Enter a parse tree produced by vlogParser#udp_declaration_port_list.
vlogParserListener.prototype.enterUdp_declaration_port_list = function(ctx) {
};

// Exit a parse tree produced by vlogParser#udp_declaration_port_list.
vlogParserListener.prototype.exitUdp_declaration_port_list = function(ctx) {
};


// Enter a parse tree produced by vlogParser#udp_port_declaration.
vlogParserListener.prototype.enterUdp_port_declaration = function(ctx) {
};

// Exit a parse tree produced by vlogParser#udp_port_declaration.
vlogParserListener.prototype.exitUdp_port_declaration = function(ctx) {
};


// Enter a parse tree produced by vlogParser#udp_output_declaration.
vlogParserListener.prototype.enterUdp_output_declaration = function(ctx) {
};

// Exit a parse tree produced by vlogParser#udp_output_declaration.
vlogParserListener.prototype.exitUdp_output_declaration = function(ctx) {
};


// Enter a parse tree produced by vlogParser#udp_input_declaration.
vlogParserListener.prototype.enterUdp_input_declaration = function(ctx) {
};

// Exit a parse tree produced by vlogParser#udp_input_declaration.
vlogParserListener.prototype.exitUdp_input_declaration = function(ctx) {
};


// Enter a parse tree produced by vlogParser#udp_reg_declaration.
vlogParserListener.prototype.enterUdp_reg_declaration = function(ctx) {
};

// Exit a parse tree produced by vlogParser#udp_reg_declaration.
vlogParserListener.prototype.exitUdp_reg_declaration = function(ctx) {
};


// Enter a parse tree produced by vlogParser#udp_body.
vlogParserListener.prototype.enterUdp_body = function(ctx) {
};

// Exit a parse tree produced by vlogParser#udp_body.
vlogParserListener.prototype.exitUdp_body = function(ctx) {
};


// Enter a parse tree produced by vlogParser#combinational_body.
vlogParserListener.prototype.enterCombinational_body = function(ctx) {
};

// Exit a parse tree produced by vlogParser#combinational_body.
vlogParserListener.prototype.exitCombinational_body = function(ctx) {
};


// Enter a parse tree produced by vlogParser#combinational_entry.
vlogParserListener.prototype.enterCombinational_entry = function(ctx) {
};

// Exit a parse tree produced by vlogParser#combinational_entry.
vlogParserListener.prototype.exitCombinational_entry = function(ctx) {
};


// Enter a parse tree produced by vlogParser#sequential_body.
vlogParserListener.prototype.enterSequential_body = function(ctx) {
};

// Exit a parse tree produced by vlogParser#sequential_body.
vlogParserListener.prototype.exitSequential_body = function(ctx) {
};


// Enter a parse tree produced by vlogParser#udp_initial_statement.
vlogParserListener.prototype.enterUdp_initial_statement = function(ctx) {
};

// Exit a parse tree produced by vlogParser#udp_initial_statement.
vlogParserListener.prototype.exitUdp_initial_statement = function(ctx) {
};


// Enter a parse tree produced by vlogParser#init_val.
vlogParserListener.prototype.enterInit_val = function(ctx) {
};

// Exit a parse tree produced by vlogParser#init_val.
vlogParserListener.prototype.exitInit_val = function(ctx) {
};


// Enter a parse tree produced by vlogParser#sequential_entry.
vlogParserListener.prototype.enterSequential_entry = function(ctx) {
};

// Exit a parse tree produced by vlogParser#sequential_entry.
vlogParserListener.prototype.exitSequential_entry = function(ctx) {
};


// Enter a parse tree produced by vlogParser#seq_input_list.
vlogParserListener.prototype.enterSeq_input_list = function(ctx) {
};

// Exit a parse tree produced by vlogParser#seq_input_list.
vlogParserListener.prototype.exitSeq_input_list = function(ctx) {
};


// Enter a parse tree produced by vlogParser#level_input_list.
vlogParserListener.prototype.enterLevel_input_list = function(ctx) {
};

// Exit a parse tree produced by vlogParser#level_input_list.
vlogParserListener.prototype.exitLevel_input_list = function(ctx) {
};


// Enter a parse tree produced by vlogParser#edge_input_list.
vlogParserListener.prototype.enterEdge_input_list = function(ctx) {
};

// Exit a parse tree produced by vlogParser#edge_input_list.
vlogParserListener.prototype.exitEdge_input_list = function(ctx) {
};


// Enter a parse tree produced by vlogParser#edge_indicator.
vlogParserListener.prototype.enterEdge_indicator = function(ctx) {
};

// Exit a parse tree produced by vlogParser#edge_indicator.
vlogParserListener.prototype.exitEdge_indicator = function(ctx) {
};


// Enter a parse tree produced by vlogParser#current_state.
vlogParserListener.prototype.enterCurrent_state = function(ctx) {
};

// Exit a parse tree produced by vlogParser#current_state.
vlogParserListener.prototype.exitCurrent_state = function(ctx) {
};


// Enter a parse tree produced by vlogParser#next_state.
vlogParserListener.prototype.enterNext_state = function(ctx) {
};

// Exit a parse tree produced by vlogParser#next_state.
vlogParserListener.prototype.exitNext_state = function(ctx) {
};


// Enter a parse tree produced by vlogParser#output_symbol.
vlogParserListener.prototype.enterOutput_symbol = function(ctx) {
};

// Exit a parse tree produced by vlogParser#output_symbol.
vlogParserListener.prototype.exitOutput_symbol = function(ctx) {
};


// Enter a parse tree produced by vlogParser#level_symbol.
vlogParserListener.prototype.enterLevel_symbol = function(ctx) {
};

// Exit a parse tree produced by vlogParser#level_symbol.
vlogParserListener.prototype.exitLevel_symbol = function(ctx) {
};


// Enter a parse tree produced by vlogParser#edge_symbol.
vlogParserListener.prototype.enterEdge_symbol = function(ctx) {
};

// Exit a parse tree produced by vlogParser#edge_symbol.
vlogParserListener.prototype.exitEdge_symbol = function(ctx) {
};


// Enter a parse tree produced by vlogParser#udp_instantiation.
vlogParserListener.prototype.enterUdp_instantiation = function(ctx) {
};

// Exit a parse tree produced by vlogParser#udp_instantiation.
vlogParserListener.prototype.exitUdp_instantiation = function(ctx) {
};


// Enter a parse tree produced by vlogParser#udp_instance.
vlogParserListener.prototype.enterUdp_instance = function(ctx) {
};

// Exit a parse tree produced by vlogParser#udp_instance.
vlogParserListener.prototype.exitUdp_instance = function(ctx) {
};


// Enter a parse tree produced by vlogParser#name_of_udp_instance.
vlogParserListener.prototype.enterName_of_udp_instance = function(ctx) {
};

// Exit a parse tree produced by vlogParser#name_of_udp_instance.
vlogParserListener.prototype.exitName_of_udp_instance = function(ctx) {
};


// Enter a parse tree produced by vlogParser#continuous_assign.
vlogParserListener.prototype.enterContinuous_assign = function(ctx) {
};

// Exit a parse tree produced by vlogParser#continuous_assign.
vlogParserListener.prototype.exitContinuous_assign = function(ctx) {
};


// Enter a parse tree produced by vlogParser#list_of_net_assignments.
vlogParserListener.prototype.enterList_of_net_assignments = function(ctx) {
};

// Exit a parse tree produced by vlogParser#list_of_net_assignments.
vlogParserListener.prototype.exitList_of_net_assignments = function(ctx) {
};


// Enter a parse tree produced by vlogParser#net_assignment.
vlogParserListener.prototype.enterNet_assignment = function(ctx) {
};

// Exit a parse tree produced by vlogParser#net_assignment.
vlogParserListener.prototype.exitNet_assignment = function(ctx) {
};


// Enter a parse tree produced by vlogParser#initial_construct.
vlogParserListener.prototype.enterInitial_construct = function(ctx) {
};

// Exit a parse tree produced by vlogParser#initial_construct.
vlogParserListener.prototype.exitInitial_construct = function(ctx) {
};


// Enter a parse tree produced by vlogParser#always_construct.
vlogParserListener.prototype.enterAlways_construct = function(ctx) {
};

// Exit a parse tree produced by vlogParser#always_construct.
vlogParserListener.prototype.exitAlways_construct = function(ctx) {
};


// Enter a parse tree produced by vlogParser#blocking_assignment.
vlogParserListener.prototype.enterBlocking_assignment = function(ctx) {
};

// Exit a parse tree produced by vlogParser#blocking_assignment.
vlogParserListener.prototype.exitBlocking_assignment = function(ctx) {
};


// Enter a parse tree produced by vlogParser#nonblocking_assignment.
vlogParserListener.prototype.enterNonblocking_assignment = function(ctx) {
};

// Exit a parse tree produced by vlogParser#nonblocking_assignment.
vlogParserListener.prototype.exitNonblocking_assignment = function(ctx) {
};


// Enter a parse tree produced by vlogParser#procedural_continuous_assignments.
vlogParserListener.prototype.enterProcedural_continuous_assignments = function(ctx) {
};

// Exit a parse tree produced by vlogParser#procedural_continuous_assignments.
vlogParserListener.prototype.exitProcedural_continuous_assignments = function(ctx) {
};


// Enter a parse tree produced by vlogParser#variable_assignment.
vlogParserListener.prototype.enterVariable_assignment = function(ctx) {
};

// Exit a parse tree produced by vlogParser#variable_assignment.
vlogParserListener.prototype.exitVariable_assignment = function(ctx) {
};


// Enter a parse tree produced by vlogParser#par_block.
vlogParserListener.prototype.enterPar_block = function(ctx) {
};

// Exit a parse tree produced by vlogParser#par_block.
vlogParserListener.prototype.exitPar_block = function(ctx) {
};


// Enter a parse tree produced by vlogParser#seq_block.
vlogParserListener.prototype.enterSeq_block = function(ctx) {
};

// Exit a parse tree produced by vlogParser#seq_block.
vlogParserListener.prototype.exitSeq_block = function(ctx) {
};


// Enter a parse tree produced by vlogParser#statement.
vlogParserListener.prototype.enterStatement = function(ctx) {
};

// Exit a parse tree produced by vlogParser#statement.
vlogParserListener.prototype.exitStatement = function(ctx) {
};


// Enter a parse tree produced by vlogParser#statement_or_null.
vlogParserListener.prototype.enterStatement_or_null = function(ctx) {
};

// Exit a parse tree produced by vlogParser#statement_or_null.
vlogParserListener.prototype.exitStatement_or_null = function(ctx) {
};


// Enter a parse tree produced by vlogParser#function_statement.
vlogParserListener.prototype.enterFunction_statement = function(ctx) {
};

// Exit a parse tree produced by vlogParser#function_statement.
vlogParserListener.prototype.exitFunction_statement = function(ctx) {
};


// Enter a parse tree produced by vlogParser#delay_control.
vlogParserListener.prototype.enterDelay_control = function(ctx) {
};

// Exit a parse tree produced by vlogParser#delay_control.
vlogParserListener.prototype.exitDelay_control = function(ctx) {
};


// Enter a parse tree produced by vlogParser#delay_or_event_control.
vlogParserListener.prototype.enterDelay_or_event_control = function(ctx) {
};

// Exit a parse tree produced by vlogParser#delay_or_event_control.
vlogParserListener.prototype.exitDelay_or_event_control = function(ctx) {
};


// Enter a parse tree produced by vlogParser#disable_statement.
vlogParserListener.prototype.enterDisable_statement = function(ctx) {
};

// Exit a parse tree produced by vlogParser#disable_statement.
vlogParserListener.prototype.exitDisable_statement = function(ctx) {
};


// Enter a parse tree produced by vlogParser#event_control.
vlogParserListener.prototype.enterEvent_control = function(ctx) {
};

// Exit a parse tree produced by vlogParser#event_control.
vlogParserListener.prototype.exitEvent_control = function(ctx) {
};


// Enter a parse tree produced by vlogParser#event_trigger.
vlogParserListener.prototype.enterEvent_trigger = function(ctx) {
};

// Exit a parse tree produced by vlogParser#event_trigger.
vlogParserListener.prototype.exitEvent_trigger = function(ctx) {
};


// Enter a parse tree produced by vlogParser#event_expression.
vlogParserListener.prototype.enterEvent_expression = function(ctx) {
};

// Exit a parse tree produced by vlogParser#event_expression.
vlogParserListener.prototype.exitEvent_expression = function(ctx) {
};


// Enter a parse tree produced by vlogParser#procedural_timing_control.
vlogParserListener.prototype.enterProcedural_timing_control = function(ctx) {
};

// Exit a parse tree produced by vlogParser#procedural_timing_control.
vlogParserListener.prototype.exitProcedural_timing_control = function(ctx) {
};


// Enter a parse tree produced by vlogParser#procedural_timing_control_statement.
vlogParserListener.prototype.enterProcedural_timing_control_statement = function(ctx) {
};

// Exit a parse tree produced by vlogParser#procedural_timing_control_statement.
vlogParserListener.prototype.exitProcedural_timing_control_statement = function(ctx) {
};


// Enter a parse tree produced by vlogParser#wait_statement.
vlogParserListener.prototype.enterWait_statement = function(ctx) {
};

// Exit a parse tree produced by vlogParser#wait_statement.
vlogParserListener.prototype.exitWait_statement = function(ctx) {
};


// Enter a parse tree produced by vlogParser#conditional_statement.
vlogParserListener.prototype.enterConditional_statement = function(ctx) {
};

// Exit a parse tree produced by vlogParser#conditional_statement.
vlogParserListener.prototype.exitConditional_statement = function(ctx) {
};


// Enter a parse tree produced by vlogParser#case_statement.
vlogParserListener.prototype.enterCase_statement = function(ctx) {
};

// Exit a parse tree produced by vlogParser#case_statement.
vlogParserListener.prototype.exitCase_statement = function(ctx) {
};


// Enter a parse tree produced by vlogParser#case_item.
vlogParserListener.prototype.enterCase_item = function(ctx) {
};

// Exit a parse tree produced by vlogParser#case_item.
vlogParserListener.prototype.exitCase_item = function(ctx) {
};


// Enter a parse tree produced by vlogParser#loop_statement.
vlogParserListener.prototype.enterLoop_statement = function(ctx) {
};

// Exit a parse tree produced by vlogParser#loop_statement.
vlogParserListener.prototype.exitLoop_statement = function(ctx) {
};


// Enter a parse tree produced by vlogParser#system_task_enable.
vlogParserListener.prototype.enterSystem_task_enable = function(ctx) {
};

// Exit a parse tree produced by vlogParser#system_task_enable.
vlogParserListener.prototype.exitSystem_task_enable = function(ctx) {
};


// Enter a parse tree produced by vlogParser#task_enable.
vlogParserListener.prototype.enterTask_enable = function(ctx) {
};

// Exit a parse tree produced by vlogParser#task_enable.
vlogParserListener.prototype.exitTask_enable = function(ctx) {
};


// Enter a parse tree produced by vlogParser#specify_block.
vlogParserListener.prototype.enterSpecify_block = function(ctx) {
};

// Exit a parse tree produced by vlogParser#specify_block.
vlogParserListener.prototype.exitSpecify_block = function(ctx) {
};


// Enter a parse tree produced by vlogParser#specify_item.
vlogParserListener.prototype.enterSpecify_item = function(ctx) {
};

// Exit a parse tree produced by vlogParser#specify_item.
vlogParserListener.prototype.exitSpecify_item = function(ctx) {
};


// Enter a parse tree produced by vlogParser#pulsestyle_declaration.
vlogParserListener.prototype.enterPulsestyle_declaration = function(ctx) {
};

// Exit a parse tree produced by vlogParser#pulsestyle_declaration.
vlogParserListener.prototype.exitPulsestyle_declaration = function(ctx) {
};


// Enter a parse tree produced by vlogParser#showcancelled_declaration.
vlogParserListener.prototype.enterShowcancelled_declaration = function(ctx) {
};

// Exit a parse tree produced by vlogParser#showcancelled_declaration.
vlogParserListener.prototype.exitShowcancelled_declaration = function(ctx) {
};


// Enter a parse tree produced by vlogParser#path_declaration.
vlogParserListener.prototype.enterPath_declaration = function(ctx) {
};

// Exit a parse tree produced by vlogParser#path_declaration.
vlogParserListener.prototype.exitPath_declaration = function(ctx) {
};


// Enter a parse tree produced by vlogParser#simple_path_declaration.
vlogParserListener.prototype.enterSimple_path_declaration = function(ctx) {
};

// Exit a parse tree produced by vlogParser#simple_path_declaration.
vlogParserListener.prototype.exitSimple_path_declaration = function(ctx) {
};


// Enter a parse tree produced by vlogParser#parallel_path_description.
vlogParserListener.prototype.enterParallel_path_description = function(ctx) {
};

// Exit a parse tree produced by vlogParser#parallel_path_description.
vlogParserListener.prototype.exitParallel_path_description = function(ctx) {
};


// Enter a parse tree produced by vlogParser#full_path_description.
vlogParserListener.prototype.enterFull_path_description = function(ctx) {
};

// Exit a parse tree produced by vlogParser#full_path_description.
vlogParserListener.prototype.exitFull_path_description = function(ctx) {
};


// Enter a parse tree produced by vlogParser#list_of_path_inputs.
vlogParserListener.prototype.enterList_of_path_inputs = function(ctx) {
};

// Exit a parse tree produced by vlogParser#list_of_path_inputs.
vlogParserListener.prototype.exitList_of_path_inputs = function(ctx) {
};


// Enter a parse tree produced by vlogParser#list_of_path_outputs.
vlogParserListener.prototype.enterList_of_path_outputs = function(ctx) {
};

// Exit a parse tree produced by vlogParser#list_of_path_outputs.
vlogParserListener.prototype.exitList_of_path_outputs = function(ctx) {
};


// Enter a parse tree produced by vlogParser#specify_input_terminal_descriptor.
vlogParserListener.prototype.enterSpecify_input_terminal_descriptor = function(ctx) {
};

// Exit a parse tree produced by vlogParser#specify_input_terminal_descriptor.
vlogParserListener.prototype.exitSpecify_input_terminal_descriptor = function(ctx) {
};


// Enter a parse tree produced by vlogParser#specify_output_terminal_descriptor.
vlogParserListener.prototype.enterSpecify_output_terminal_descriptor = function(ctx) {
};

// Exit a parse tree produced by vlogParser#specify_output_terminal_descriptor.
vlogParserListener.prototype.exitSpecify_output_terminal_descriptor = function(ctx) {
};


// Enter a parse tree produced by vlogParser#input_identifier.
vlogParserListener.prototype.enterInput_identifier = function(ctx) {
};

// Exit a parse tree produced by vlogParser#input_identifier.
vlogParserListener.prototype.exitInput_identifier = function(ctx) {
};


// Enter a parse tree produced by vlogParser#output_identifier.
vlogParserListener.prototype.enterOutput_identifier = function(ctx) {
};

// Exit a parse tree produced by vlogParser#output_identifier.
vlogParserListener.prototype.exitOutput_identifier = function(ctx) {
};


// Enter a parse tree produced by vlogParser#path_delay_value.
vlogParserListener.prototype.enterPath_delay_value = function(ctx) {
};

// Exit a parse tree produced by vlogParser#path_delay_value.
vlogParserListener.prototype.exitPath_delay_value = function(ctx) {
};


// Enter a parse tree produced by vlogParser#list_of_path_delay_expressions.
vlogParserListener.prototype.enterList_of_path_delay_expressions = function(ctx) {
};

// Exit a parse tree produced by vlogParser#list_of_path_delay_expressions.
vlogParserListener.prototype.exitList_of_path_delay_expressions = function(ctx) {
};


// Enter a parse tree produced by vlogParser#t_path_delay_expression.
vlogParserListener.prototype.enterT_path_delay_expression = function(ctx) {
};

// Exit a parse tree produced by vlogParser#t_path_delay_expression.
vlogParserListener.prototype.exitT_path_delay_expression = function(ctx) {
};


// Enter a parse tree produced by vlogParser#trise_path_delay_expression.
vlogParserListener.prototype.enterTrise_path_delay_expression = function(ctx) {
};

// Exit a parse tree produced by vlogParser#trise_path_delay_expression.
vlogParserListener.prototype.exitTrise_path_delay_expression = function(ctx) {
};


// Enter a parse tree produced by vlogParser#tfall_path_delay_expression.
vlogParserListener.prototype.enterTfall_path_delay_expression = function(ctx) {
};

// Exit a parse tree produced by vlogParser#tfall_path_delay_expression.
vlogParserListener.prototype.exitTfall_path_delay_expression = function(ctx) {
};


// Enter a parse tree produced by vlogParser#tz_path_delay_expression.
vlogParserListener.prototype.enterTz_path_delay_expression = function(ctx) {
};

// Exit a parse tree produced by vlogParser#tz_path_delay_expression.
vlogParserListener.prototype.exitTz_path_delay_expression = function(ctx) {
};


// Enter a parse tree produced by vlogParser#t01_path_delay_expression.
vlogParserListener.prototype.enterT01_path_delay_expression = function(ctx) {
};

// Exit a parse tree produced by vlogParser#t01_path_delay_expression.
vlogParserListener.prototype.exitT01_path_delay_expression = function(ctx) {
};


// Enter a parse tree produced by vlogParser#t10_path_delay_expression.
vlogParserListener.prototype.enterT10_path_delay_expression = function(ctx) {
};

// Exit a parse tree produced by vlogParser#t10_path_delay_expression.
vlogParserListener.prototype.exitT10_path_delay_expression = function(ctx) {
};


// Enter a parse tree produced by vlogParser#t0z_path_delay_expression.
vlogParserListener.prototype.enterT0z_path_delay_expression = function(ctx) {
};

// Exit a parse tree produced by vlogParser#t0z_path_delay_expression.
vlogParserListener.prototype.exitT0z_path_delay_expression = function(ctx) {
};


// Enter a parse tree produced by vlogParser#tz1_path_delay_expression.
vlogParserListener.prototype.enterTz1_path_delay_expression = function(ctx) {
};

// Exit a parse tree produced by vlogParser#tz1_path_delay_expression.
vlogParserListener.prototype.exitTz1_path_delay_expression = function(ctx) {
};


// Enter a parse tree produced by vlogParser#t1z_path_delay_expression.
vlogParserListener.prototype.enterT1z_path_delay_expression = function(ctx) {
};

// Exit a parse tree produced by vlogParser#t1z_path_delay_expression.
vlogParserListener.prototype.exitT1z_path_delay_expression = function(ctx) {
};


// Enter a parse tree produced by vlogParser#tz0_path_delay_expression.
vlogParserListener.prototype.enterTz0_path_delay_expression = function(ctx) {
};

// Exit a parse tree produced by vlogParser#tz0_path_delay_expression.
vlogParserListener.prototype.exitTz0_path_delay_expression = function(ctx) {
};


// Enter a parse tree produced by vlogParser#t0x_path_delay_expression.
vlogParserListener.prototype.enterT0x_path_delay_expression = function(ctx) {
};

// Exit a parse tree produced by vlogParser#t0x_path_delay_expression.
vlogParserListener.prototype.exitT0x_path_delay_expression = function(ctx) {
};


// Enter a parse tree produced by vlogParser#tx1_path_delay_expression.
vlogParserListener.prototype.enterTx1_path_delay_expression = function(ctx) {
};

// Exit a parse tree produced by vlogParser#tx1_path_delay_expression.
vlogParserListener.prototype.exitTx1_path_delay_expression = function(ctx) {
};


// Enter a parse tree produced by vlogParser#t1x_path_delay_expression.
vlogParserListener.prototype.enterT1x_path_delay_expression = function(ctx) {
};

// Exit a parse tree produced by vlogParser#t1x_path_delay_expression.
vlogParserListener.prototype.exitT1x_path_delay_expression = function(ctx) {
};


// Enter a parse tree produced by vlogParser#tx0_path_delay_expression.
vlogParserListener.prototype.enterTx0_path_delay_expression = function(ctx) {
};

// Exit a parse tree produced by vlogParser#tx0_path_delay_expression.
vlogParserListener.prototype.exitTx0_path_delay_expression = function(ctx) {
};


// Enter a parse tree produced by vlogParser#txz_path_delay_expression.
vlogParserListener.prototype.enterTxz_path_delay_expression = function(ctx) {
};

// Exit a parse tree produced by vlogParser#txz_path_delay_expression.
vlogParserListener.prototype.exitTxz_path_delay_expression = function(ctx) {
};


// Enter a parse tree produced by vlogParser#tzx_path_delay_expression.
vlogParserListener.prototype.enterTzx_path_delay_expression = function(ctx) {
};

// Exit a parse tree produced by vlogParser#tzx_path_delay_expression.
vlogParserListener.prototype.exitTzx_path_delay_expression = function(ctx) {
};


// Enter a parse tree produced by vlogParser#path_delay_expression.
vlogParserListener.prototype.enterPath_delay_expression = function(ctx) {
};

// Exit a parse tree produced by vlogParser#path_delay_expression.
vlogParserListener.prototype.exitPath_delay_expression = function(ctx) {
};


// Enter a parse tree produced by vlogParser#edge_sensitive_path_declaration.
vlogParserListener.prototype.enterEdge_sensitive_path_declaration = function(ctx) {
};

// Exit a parse tree produced by vlogParser#edge_sensitive_path_declaration.
vlogParserListener.prototype.exitEdge_sensitive_path_declaration = function(ctx) {
};


// Enter a parse tree produced by vlogParser#parallel_edge_sensitive_path_description.
vlogParserListener.prototype.enterParallel_edge_sensitive_path_description = function(ctx) {
};

// Exit a parse tree produced by vlogParser#parallel_edge_sensitive_path_description.
vlogParserListener.prototype.exitParallel_edge_sensitive_path_description = function(ctx) {
};


// Enter a parse tree produced by vlogParser#full_edge_sensitive_path_description.
vlogParserListener.prototype.enterFull_edge_sensitive_path_description = function(ctx) {
};

// Exit a parse tree produced by vlogParser#full_edge_sensitive_path_description.
vlogParserListener.prototype.exitFull_edge_sensitive_path_description = function(ctx) {
};


// Enter a parse tree produced by vlogParser#data_source_expression.
vlogParserListener.prototype.enterData_source_expression = function(ctx) {
};

// Exit a parse tree produced by vlogParser#data_source_expression.
vlogParserListener.prototype.exitData_source_expression = function(ctx) {
};


// Enter a parse tree produced by vlogParser#edge_identifier.
vlogParserListener.prototype.enterEdge_identifier = function(ctx) {
};

// Exit a parse tree produced by vlogParser#edge_identifier.
vlogParserListener.prototype.exitEdge_identifier = function(ctx) {
};


// Enter a parse tree produced by vlogParser#state_dependent_path_declaration.
vlogParserListener.prototype.enterState_dependent_path_declaration = function(ctx) {
};

// Exit a parse tree produced by vlogParser#state_dependent_path_declaration.
vlogParserListener.prototype.exitState_dependent_path_declaration = function(ctx) {
};


// Enter a parse tree produced by vlogParser#polarity_operator.
vlogParserListener.prototype.enterPolarity_operator = function(ctx) {
};

// Exit a parse tree produced by vlogParser#polarity_operator.
vlogParserListener.prototype.exitPolarity_operator = function(ctx) {
};


// Enter a parse tree produced by vlogParser#system_timing_check.
vlogParserListener.prototype.enterSystem_timing_check = function(ctx) {
};

// Exit a parse tree produced by vlogParser#system_timing_check.
vlogParserListener.prototype.exitSystem_timing_check = function(ctx) {
};


// Enter a parse tree produced by vlogParser#setup_timing_check.
vlogParserListener.prototype.enterSetup_timing_check = function(ctx) {
};

// Exit a parse tree produced by vlogParser#setup_timing_check.
vlogParserListener.prototype.exitSetup_timing_check = function(ctx) {
};


// Enter a parse tree produced by vlogParser#hold_timing_check.
vlogParserListener.prototype.enterHold_timing_check = function(ctx) {
};

// Exit a parse tree produced by vlogParser#hold_timing_check.
vlogParserListener.prototype.exitHold_timing_check = function(ctx) {
};


// Enter a parse tree produced by vlogParser#setuphold_timing_check.
vlogParserListener.prototype.enterSetuphold_timing_check = function(ctx) {
};

// Exit a parse tree produced by vlogParser#setuphold_timing_check.
vlogParserListener.prototype.exitSetuphold_timing_check = function(ctx) {
};


// Enter a parse tree produced by vlogParser#recovery_timing_check.
vlogParserListener.prototype.enterRecovery_timing_check = function(ctx) {
};

// Exit a parse tree produced by vlogParser#recovery_timing_check.
vlogParserListener.prototype.exitRecovery_timing_check = function(ctx) {
};


// Enter a parse tree produced by vlogParser#removal_timing_check.
vlogParserListener.prototype.enterRemoval_timing_check = function(ctx) {
};

// Exit a parse tree produced by vlogParser#removal_timing_check.
vlogParserListener.prototype.exitRemoval_timing_check = function(ctx) {
};


// Enter a parse tree produced by vlogParser#recrem_timing_check.
vlogParserListener.prototype.enterRecrem_timing_check = function(ctx) {
};

// Exit a parse tree produced by vlogParser#recrem_timing_check.
vlogParserListener.prototype.exitRecrem_timing_check = function(ctx) {
};


// Enter a parse tree produced by vlogParser#skew_timing_check.
vlogParserListener.prototype.enterSkew_timing_check = function(ctx) {
};

// Exit a parse tree produced by vlogParser#skew_timing_check.
vlogParserListener.prototype.exitSkew_timing_check = function(ctx) {
};


// Enter a parse tree produced by vlogParser#timeskew_timing_check.
vlogParserListener.prototype.enterTimeskew_timing_check = function(ctx) {
};

// Exit a parse tree produced by vlogParser#timeskew_timing_check.
vlogParserListener.prototype.exitTimeskew_timing_check = function(ctx) {
};


// Enter a parse tree produced by vlogParser#fullskew_timing_check.
vlogParserListener.prototype.enterFullskew_timing_check = function(ctx) {
};

// Exit a parse tree produced by vlogParser#fullskew_timing_check.
vlogParserListener.prototype.exitFullskew_timing_check = function(ctx) {
};


// Enter a parse tree produced by vlogParser#period_timing_check.
vlogParserListener.prototype.enterPeriod_timing_check = function(ctx) {
};

// Exit a parse tree produced by vlogParser#period_timing_check.
vlogParserListener.prototype.exitPeriod_timing_check = function(ctx) {
};


// Enter a parse tree produced by vlogParser#width_timing_check.
vlogParserListener.prototype.enterWidth_timing_check = function(ctx) {
};

// Exit a parse tree produced by vlogParser#width_timing_check.
vlogParserListener.prototype.exitWidth_timing_check = function(ctx) {
};


// Enter a parse tree produced by vlogParser#nochange_timing_check.
vlogParserListener.prototype.enterNochange_timing_check = function(ctx) {
};

// Exit a parse tree produced by vlogParser#nochange_timing_check.
vlogParserListener.prototype.exitNochange_timing_check = function(ctx) {
};


// Enter a parse tree produced by vlogParser#checktime_condition.
vlogParserListener.prototype.enterChecktime_condition = function(ctx) {
};

// Exit a parse tree produced by vlogParser#checktime_condition.
vlogParserListener.prototype.exitChecktime_condition = function(ctx) {
};


// Enter a parse tree produced by vlogParser#controlled_reference_event.
vlogParserListener.prototype.enterControlled_reference_event = function(ctx) {
};

// Exit a parse tree produced by vlogParser#controlled_reference_event.
vlogParserListener.prototype.exitControlled_reference_event = function(ctx) {
};


// Enter a parse tree produced by vlogParser#data_event.
vlogParserListener.prototype.enterData_event = function(ctx) {
};

// Exit a parse tree produced by vlogParser#data_event.
vlogParserListener.prototype.exitData_event = function(ctx) {
};


// Enter a parse tree produced by vlogParser#delayed_data.
vlogParserListener.prototype.enterDelayed_data = function(ctx) {
};

// Exit a parse tree produced by vlogParser#delayed_data.
vlogParserListener.prototype.exitDelayed_data = function(ctx) {
};


// Enter a parse tree produced by vlogParser#delayed_reference.
vlogParserListener.prototype.enterDelayed_reference = function(ctx) {
};

// Exit a parse tree produced by vlogParser#delayed_reference.
vlogParserListener.prototype.exitDelayed_reference = function(ctx) {
};


// Enter a parse tree produced by vlogParser#end_edge_offset.
vlogParserListener.prototype.enterEnd_edge_offset = function(ctx) {
};

// Exit a parse tree produced by vlogParser#end_edge_offset.
vlogParserListener.prototype.exitEnd_edge_offset = function(ctx) {
};


// Enter a parse tree produced by vlogParser#event_based_flag.
vlogParserListener.prototype.enterEvent_based_flag = function(ctx) {
};

// Exit a parse tree produced by vlogParser#event_based_flag.
vlogParserListener.prototype.exitEvent_based_flag = function(ctx) {
};


// Enter a parse tree produced by vlogParser#notifier.
vlogParserListener.prototype.enterNotifier = function(ctx) {
};

// Exit a parse tree produced by vlogParser#notifier.
vlogParserListener.prototype.exitNotifier = function(ctx) {
};


// Enter a parse tree produced by vlogParser#reference_event.
vlogParserListener.prototype.enterReference_event = function(ctx) {
};

// Exit a parse tree produced by vlogParser#reference_event.
vlogParserListener.prototype.exitReference_event = function(ctx) {
};


// Enter a parse tree produced by vlogParser#remain_active_flag.
vlogParserListener.prototype.enterRemain_active_flag = function(ctx) {
};

// Exit a parse tree produced by vlogParser#remain_active_flag.
vlogParserListener.prototype.exitRemain_active_flag = function(ctx) {
};


// Enter a parse tree produced by vlogParser#stamptime_condition.
vlogParserListener.prototype.enterStamptime_condition = function(ctx) {
};

// Exit a parse tree produced by vlogParser#stamptime_condition.
vlogParserListener.prototype.exitStamptime_condition = function(ctx) {
};


// Enter a parse tree produced by vlogParser#start_edge_offset.
vlogParserListener.prototype.enterStart_edge_offset = function(ctx) {
};

// Exit a parse tree produced by vlogParser#start_edge_offset.
vlogParserListener.prototype.exitStart_edge_offset = function(ctx) {
};


// Enter a parse tree produced by vlogParser#threshold.
vlogParserListener.prototype.enterThreshold = function(ctx) {
};

// Exit a parse tree produced by vlogParser#threshold.
vlogParserListener.prototype.exitThreshold = function(ctx) {
};


// Enter a parse tree produced by vlogParser#timing_check_limit.
vlogParserListener.prototype.enterTiming_check_limit = function(ctx) {
};

// Exit a parse tree produced by vlogParser#timing_check_limit.
vlogParserListener.prototype.exitTiming_check_limit = function(ctx) {
};


// Enter a parse tree produced by vlogParser#timing_check_event.
vlogParserListener.prototype.enterTiming_check_event = function(ctx) {
};

// Exit a parse tree produced by vlogParser#timing_check_event.
vlogParserListener.prototype.exitTiming_check_event = function(ctx) {
};


// Enter a parse tree produced by vlogParser#controlled_timing_check_event.
vlogParserListener.prototype.enterControlled_timing_check_event = function(ctx) {
};

// Exit a parse tree produced by vlogParser#controlled_timing_check_event.
vlogParserListener.prototype.exitControlled_timing_check_event = function(ctx) {
};


// Enter a parse tree produced by vlogParser#timing_check_event_control.
vlogParserListener.prototype.enterTiming_check_event_control = function(ctx) {
};

// Exit a parse tree produced by vlogParser#timing_check_event_control.
vlogParserListener.prototype.exitTiming_check_event_control = function(ctx) {
};


// Enter a parse tree produced by vlogParser#specify_terminal_descriptor.
vlogParserListener.prototype.enterSpecify_terminal_descriptor = function(ctx) {
};

// Exit a parse tree produced by vlogParser#specify_terminal_descriptor.
vlogParserListener.prototype.exitSpecify_terminal_descriptor = function(ctx) {
};


// Enter a parse tree produced by vlogParser#edge_control_specifier.
vlogParserListener.prototype.enterEdge_control_specifier = function(ctx) {
};

// Exit a parse tree produced by vlogParser#edge_control_specifier.
vlogParserListener.prototype.exitEdge_control_specifier = function(ctx) {
};


// Enter a parse tree produced by vlogParser#edge_descriptor.
vlogParserListener.prototype.enterEdge_descriptor = function(ctx) {
};

// Exit a parse tree produced by vlogParser#edge_descriptor.
vlogParserListener.prototype.exitEdge_descriptor = function(ctx) {
};


// Enter a parse tree produced by vlogParser#timing_check_condition.
vlogParserListener.prototype.enterTiming_check_condition = function(ctx) {
};

// Exit a parse tree produced by vlogParser#timing_check_condition.
vlogParserListener.prototype.exitTiming_check_condition = function(ctx) {
};


// Enter a parse tree produced by vlogParser#scalar_timing_check_condition.
vlogParserListener.prototype.enterScalar_timing_check_condition = function(ctx) {
};

// Exit a parse tree produced by vlogParser#scalar_timing_check_condition.
vlogParserListener.prototype.exitScalar_timing_check_condition = function(ctx) {
};


// Enter a parse tree produced by vlogParser#scalar_constant.
vlogParserListener.prototype.enterScalar_constant = function(ctx) {
};

// Exit a parse tree produced by vlogParser#scalar_constant.
vlogParserListener.prototype.exitScalar_constant = function(ctx) {
};


// Enter a parse tree produced by vlogParser#concatenation.
vlogParserListener.prototype.enterConcatenation = function(ctx) {
};

// Exit a parse tree produced by vlogParser#concatenation.
vlogParserListener.prototype.exitConcatenation = function(ctx) {
};


// Enter a parse tree produced by vlogParser#constant_concatenation.
vlogParserListener.prototype.enterConstant_concatenation = function(ctx) {
};

// Exit a parse tree produced by vlogParser#constant_concatenation.
vlogParserListener.prototype.exitConstant_concatenation = function(ctx) {
};


// Enter a parse tree produced by vlogParser#constant_multiple_concatenation.
vlogParserListener.prototype.enterConstant_multiple_concatenation = function(ctx) {
};

// Exit a parse tree produced by vlogParser#constant_multiple_concatenation.
vlogParserListener.prototype.exitConstant_multiple_concatenation = function(ctx) {
};


// Enter a parse tree produced by vlogParser#module_path_concatenation.
vlogParserListener.prototype.enterModule_path_concatenation = function(ctx) {
};

// Exit a parse tree produced by vlogParser#module_path_concatenation.
vlogParserListener.prototype.exitModule_path_concatenation = function(ctx) {
};


// Enter a parse tree produced by vlogParser#module_path_multiple_concatenation.
vlogParserListener.prototype.enterModule_path_multiple_concatenation = function(ctx) {
};

// Exit a parse tree produced by vlogParser#module_path_multiple_concatenation.
vlogParserListener.prototype.exitModule_path_multiple_concatenation = function(ctx) {
};


// Enter a parse tree produced by vlogParser#multiple_concatenation.
vlogParserListener.prototype.enterMultiple_concatenation = function(ctx) {
};

// Exit a parse tree produced by vlogParser#multiple_concatenation.
vlogParserListener.prototype.exitMultiple_concatenation = function(ctx) {
};


// Enter a parse tree produced by vlogParser#constant_function_call.
vlogParserListener.prototype.enterConstant_function_call = function(ctx) {
};

// Exit a parse tree produced by vlogParser#constant_function_call.
vlogParserListener.prototype.exitConstant_function_call = function(ctx) {
};


// Enter a parse tree produced by vlogParser#constant_system_function_call.
vlogParserListener.prototype.enterConstant_system_function_call = function(ctx) {
};

// Exit a parse tree produced by vlogParser#constant_system_function_call.
vlogParserListener.prototype.exitConstant_system_function_call = function(ctx) {
};


// Enter a parse tree produced by vlogParser#function_call.
vlogParserListener.prototype.enterFunction_call = function(ctx) {
};

// Exit a parse tree produced by vlogParser#function_call.
vlogParserListener.prototype.exitFunction_call = function(ctx) {
};


// Enter a parse tree produced by vlogParser#system_function_call.
vlogParserListener.prototype.enterSystem_function_call = function(ctx) {
};

// Exit a parse tree produced by vlogParser#system_function_call.
vlogParserListener.prototype.exitSystem_function_call = function(ctx) {
};


// Enter a parse tree produced by vlogParser#base_expression.
vlogParserListener.prototype.enterBase_expression = function(ctx) {
};

// Exit a parse tree produced by vlogParser#base_expression.
vlogParserListener.prototype.exitBase_expression = function(ctx) {
};


// Enter a parse tree produced by vlogParser#constant_base_expression.
vlogParserListener.prototype.enterConstant_base_expression = function(ctx) {
};

// Exit a parse tree produced by vlogParser#constant_base_expression.
vlogParserListener.prototype.exitConstant_base_expression = function(ctx) {
};


// Enter a parse tree produced by vlogParser#constant_expression.
vlogParserListener.prototype.enterConstant_expression = function(ctx) {
};

// Exit a parse tree produced by vlogParser#constant_expression.
vlogParserListener.prototype.exitConstant_expression = function(ctx) {
};


// Enter a parse tree produced by vlogParser#constant_mintypmax_expression.
vlogParserListener.prototype.enterConstant_mintypmax_expression = function(ctx) {
};

// Exit a parse tree produced by vlogParser#constant_mintypmax_expression.
vlogParserListener.prototype.exitConstant_mintypmax_expression = function(ctx) {
};


// Enter a parse tree produced by vlogParser#constant_range_expression.
vlogParserListener.prototype.enterConstant_range_expression = function(ctx) {
};

// Exit a parse tree produced by vlogParser#constant_range_expression.
vlogParserListener.prototype.exitConstant_range_expression = function(ctx) {
};


// Enter a parse tree produced by vlogParser#dimension_constant_expression.
vlogParserListener.prototype.enterDimension_constant_expression = function(ctx) {
};

// Exit a parse tree produced by vlogParser#dimension_constant_expression.
vlogParserListener.prototype.exitDimension_constant_expression = function(ctx) {
};


// Enter a parse tree produced by vlogParser#expression.
vlogParserListener.prototype.enterExpression = function(ctx) {
};

// Exit a parse tree produced by vlogParser#expression.
vlogParserListener.prototype.exitExpression = function(ctx) {
};


// Enter a parse tree produced by vlogParser#lsb_constant_expression.
vlogParserListener.prototype.enterLsb_constant_expression = function(ctx) {
};

// Exit a parse tree produced by vlogParser#lsb_constant_expression.
vlogParserListener.prototype.exitLsb_constant_expression = function(ctx) {
};


// Enter a parse tree produced by vlogParser#mintypmax_expression.
vlogParserListener.prototype.enterMintypmax_expression = function(ctx) {
};

// Exit a parse tree produced by vlogParser#mintypmax_expression.
vlogParserListener.prototype.exitMintypmax_expression = function(ctx) {
};


// Enter a parse tree produced by vlogParser#module_path_expression.
vlogParserListener.prototype.enterModule_path_expression = function(ctx) {
};

// Exit a parse tree produced by vlogParser#module_path_expression.
vlogParserListener.prototype.exitModule_path_expression = function(ctx) {
};


// Enter a parse tree produced by vlogParser#module_path_mintypmax_expression.
vlogParserListener.prototype.enterModule_path_mintypmax_expression = function(ctx) {
};

// Exit a parse tree produced by vlogParser#module_path_mintypmax_expression.
vlogParserListener.prototype.exitModule_path_mintypmax_expression = function(ctx) {
};


// Enter a parse tree produced by vlogParser#msb_constant_expression.
vlogParserListener.prototype.enterMsb_constant_expression = function(ctx) {
};

// Exit a parse tree produced by vlogParser#msb_constant_expression.
vlogParserListener.prototype.exitMsb_constant_expression = function(ctx) {
};


// Enter a parse tree produced by vlogParser#range_expression.
vlogParserListener.prototype.enterRange_expression = function(ctx) {
};

// Exit a parse tree produced by vlogParser#range_expression.
vlogParserListener.prototype.exitRange_expression = function(ctx) {
};


// Enter a parse tree produced by vlogParser#width_constant_expression.
vlogParserListener.prototype.enterWidth_constant_expression = function(ctx) {
};

// Exit a parse tree produced by vlogParser#width_constant_expression.
vlogParserListener.prototype.exitWidth_constant_expression = function(ctx) {
};


// Enter a parse tree produced by vlogParser#constant_primary.
vlogParserListener.prototype.enterConstant_primary = function(ctx) {
};

// Exit a parse tree produced by vlogParser#constant_primary.
vlogParserListener.prototype.exitConstant_primary = function(ctx) {
};


// Enter a parse tree produced by vlogParser#module_path_primary.
vlogParserListener.prototype.enterModule_path_primary = function(ctx) {
};

// Exit a parse tree produced by vlogParser#module_path_primary.
vlogParserListener.prototype.exitModule_path_primary = function(ctx) {
};


// Enter a parse tree produced by vlogParser#primary.
vlogParserListener.prototype.enterPrimary = function(ctx) {
};

// Exit a parse tree produced by vlogParser#primary.
vlogParserListener.prototype.exitPrimary = function(ctx) {
};


// Enter a parse tree produced by vlogParser#net_lvalue.
vlogParserListener.prototype.enterNet_lvalue = function(ctx) {
};

// Exit a parse tree produced by vlogParser#net_lvalue.
vlogParserListener.prototype.exitNet_lvalue = function(ctx) {
};


// Enter a parse tree produced by vlogParser#variable_lvalue.
vlogParserListener.prototype.enterVariable_lvalue = function(ctx) {
};

// Exit a parse tree produced by vlogParser#variable_lvalue.
vlogParserListener.prototype.exitVariable_lvalue = function(ctx) {
};


// Enter a parse tree produced by vlogParser#unary_operator.
vlogParserListener.prototype.enterUnary_operator = function(ctx) {
};

// Exit a parse tree produced by vlogParser#unary_operator.
vlogParserListener.prototype.exitUnary_operator = function(ctx) {
};


// Enter a parse tree produced by vlogParser#binary_operator.
vlogParserListener.prototype.enterBinary_operator = function(ctx) {
};

// Exit a parse tree produced by vlogParser#binary_operator.
vlogParserListener.prototype.exitBinary_operator = function(ctx) {
};


// Enter a parse tree produced by vlogParser#unary_module_path_operator.
vlogParserListener.prototype.enterUnary_module_path_operator = function(ctx) {
};

// Exit a parse tree produced by vlogParser#unary_module_path_operator.
vlogParserListener.prototype.exitUnary_module_path_operator = function(ctx) {
};


// Enter a parse tree produced by vlogParser#binary_module_path_operator.
vlogParserListener.prototype.enterBinary_module_path_operator = function(ctx) {
};

// Exit a parse tree produced by vlogParser#binary_module_path_operator.
vlogParserListener.prototype.exitBinary_module_path_operator = function(ctx) {
};


// Enter a parse tree produced by vlogParser#number.
vlogParserListener.prototype.enterNumber = function(ctx) {
};

// Exit a parse tree produced by vlogParser#number.
vlogParserListener.prototype.exitNumber = function(ctx) {
};


// Enter a parse tree produced by vlogParser#real_number.
vlogParserListener.prototype.enterReal_number = function(ctx) {
};

// Exit a parse tree produced by vlogParser#real_number.
vlogParserListener.prototype.exitReal_number = function(ctx) {
};


// Enter a parse tree produced by vlogParser#decimal_number.
vlogParserListener.prototype.enterDecimal_number = function(ctx) {
};

// Exit a parse tree produced by vlogParser#decimal_number.
vlogParserListener.prototype.exitDecimal_number = function(ctx) {
};


// Enter a parse tree produced by vlogParser#binary_number.
vlogParserListener.prototype.enterBinary_number = function(ctx) {
};

// Exit a parse tree produced by vlogParser#binary_number.
vlogParserListener.prototype.exitBinary_number = function(ctx) {
};


// Enter a parse tree produced by vlogParser#octal_number.
vlogParserListener.prototype.enterOctal_number = function(ctx) {
};

// Exit a parse tree produced by vlogParser#octal_number.
vlogParserListener.prototype.exitOctal_number = function(ctx) {
};


// Enter a parse tree produced by vlogParser#hex_number.
vlogParserListener.prototype.enterHex_number = function(ctx) {
};

// Exit a parse tree produced by vlogParser#hex_number.
vlogParserListener.prototype.exitHex_number = function(ctx) {
};


// Enter a parse tree produced by vlogParser#unsigned_number.
vlogParserListener.prototype.enterUnsigned_number = function(ctx) {
};

// Exit a parse tree produced by vlogParser#unsigned_number.
vlogParserListener.prototype.exitUnsigned_number = function(ctx) {
};


// Enter a parse tree produced by vlogParser#string_.
vlogParserListener.prototype.enterString_ = function(ctx) {
};

// Exit a parse tree produced by vlogParser#string_.
vlogParserListener.prototype.exitString_ = function(ctx) {
};


// Enter a parse tree produced by vlogParser#attribute_instance.
vlogParserListener.prototype.enterAttribute_instance = function(ctx) {
};

// Exit a parse tree produced by vlogParser#attribute_instance.
vlogParserListener.prototype.exitAttribute_instance = function(ctx) {
};


// Enter a parse tree produced by vlogParser#attr_spec.
vlogParserListener.prototype.enterAttr_spec = function(ctx) {
};

// Exit a parse tree produced by vlogParser#attr_spec.
vlogParserListener.prototype.exitAttr_spec = function(ctx) {
};


// Enter a parse tree produced by vlogParser#attr_name.
vlogParserListener.prototype.enterAttr_name = function(ctx) {
};

// Exit a parse tree produced by vlogParser#attr_name.
vlogParserListener.prototype.exitAttr_name = function(ctx) {
};


// Enter a parse tree produced by vlogParser#block_identifier.
vlogParserListener.prototype.enterBlock_identifier = function(ctx) {
};

// Exit a parse tree produced by vlogParser#block_identifier.
vlogParserListener.prototype.exitBlock_identifier = function(ctx) {
};


// Enter a parse tree produced by vlogParser#cell_identifier.
vlogParserListener.prototype.enterCell_identifier = function(ctx) {
};

// Exit a parse tree produced by vlogParser#cell_identifier.
vlogParserListener.prototype.exitCell_identifier = function(ctx) {
};


// Enter a parse tree produced by vlogParser#config_identifier.
vlogParserListener.prototype.enterConfig_identifier = function(ctx) {
};

// Exit a parse tree produced by vlogParser#config_identifier.
vlogParserListener.prototype.exitConfig_identifier = function(ctx) {
};


// Enter a parse tree produced by vlogParser#escaped_identifier.
vlogParserListener.prototype.enterEscaped_identifier = function(ctx) {
};

// Exit a parse tree produced by vlogParser#escaped_identifier.
vlogParserListener.prototype.exitEscaped_identifier = function(ctx) {
};


// Enter a parse tree produced by vlogParser#event_identifier.
vlogParserListener.prototype.enterEvent_identifier = function(ctx) {
};

// Exit a parse tree produced by vlogParser#event_identifier.
vlogParserListener.prototype.exitEvent_identifier = function(ctx) {
};


// Enter a parse tree produced by vlogParser#function_identifier.
vlogParserListener.prototype.enterFunction_identifier = function(ctx) {
};

// Exit a parse tree produced by vlogParser#function_identifier.
vlogParserListener.prototype.exitFunction_identifier = function(ctx) {
};


// Enter a parse tree produced by vlogParser#gate_instance_identifier.
vlogParserListener.prototype.enterGate_instance_identifier = function(ctx) {
};

// Exit a parse tree produced by vlogParser#gate_instance_identifier.
vlogParserListener.prototype.exitGate_instance_identifier = function(ctx) {
};


// Enter a parse tree produced by vlogParser#generate_block_identifier.
vlogParserListener.prototype.enterGenerate_block_identifier = function(ctx) {
};

// Exit a parse tree produced by vlogParser#generate_block_identifier.
vlogParserListener.prototype.exitGenerate_block_identifier = function(ctx) {
};


// Enter a parse tree produced by vlogParser#genvar_identifier.
vlogParserListener.prototype.enterGenvar_identifier = function(ctx) {
};

// Exit a parse tree produced by vlogParser#genvar_identifier.
vlogParserListener.prototype.exitGenvar_identifier = function(ctx) {
};


// Enter a parse tree produced by vlogParser#hierarchical_block_identifier.
vlogParserListener.prototype.enterHierarchical_block_identifier = function(ctx) {
};

// Exit a parse tree produced by vlogParser#hierarchical_block_identifier.
vlogParserListener.prototype.exitHierarchical_block_identifier = function(ctx) {
};


// Enter a parse tree produced by vlogParser#hierarchical_event_identifier.
vlogParserListener.prototype.enterHierarchical_event_identifier = function(ctx) {
};

// Exit a parse tree produced by vlogParser#hierarchical_event_identifier.
vlogParserListener.prototype.exitHierarchical_event_identifier = function(ctx) {
};


// Enter a parse tree produced by vlogParser#hierarchical_function_identifier.
vlogParserListener.prototype.enterHierarchical_function_identifier = function(ctx) {
};

// Exit a parse tree produced by vlogParser#hierarchical_function_identifier.
vlogParserListener.prototype.exitHierarchical_function_identifier = function(ctx) {
};


// Enter a parse tree produced by vlogParser#hierarchical_identifier.
vlogParserListener.prototype.enterHierarchical_identifier = function(ctx) {
};

// Exit a parse tree produced by vlogParser#hierarchical_identifier.
vlogParserListener.prototype.exitHierarchical_identifier = function(ctx) {
};


// Enter a parse tree produced by vlogParser#hierarchical_net_identifier.
vlogParserListener.prototype.enterHierarchical_net_identifier = function(ctx) {
};

// Exit a parse tree produced by vlogParser#hierarchical_net_identifier.
vlogParserListener.prototype.exitHierarchical_net_identifier = function(ctx) {
};


// Enter a parse tree produced by vlogParser#hierarchical_parameter_identifier.
vlogParserListener.prototype.enterHierarchical_parameter_identifier = function(ctx) {
};

// Exit a parse tree produced by vlogParser#hierarchical_parameter_identifier.
vlogParserListener.prototype.exitHierarchical_parameter_identifier = function(ctx) {
};


// Enter a parse tree produced by vlogParser#hierarchical_variable_identifier.
vlogParserListener.prototype.enterHierarchical_variable_identifier = function(ctx) {
};

// Exit a parse tree produced by vlogParser#hierarchical_variable_identifier.
vlogParserListener.prototype.exitHierarchical_variable_identifier = function(ctx) {
};


// Enter a parse tree produced by vlogParser#hierarchical_task_identifier.
vlogParserListener.prototype.enterHierarchical_task_identifier = function(ctx) {
};

// Exit a parse tree produced by vlogParser#hierarchical_task_identifier.
vlogParserListener.prototype.exitHierarchical_task_identifier = function(ctx) {
};


// Enter a parse tree produced by vlogParser#identifier.
vlogParserListener.prototype.enterIdentifier = function(ctx) {
};

// Exit a parse tree produced by vlogParser#identifier.
vlogParserListener.prototype.exitIdentifier = function(ctx) {
};


// Enter a parse tree produced by vlogParser#macro_usage_identifier.
vlogParserListener.prototype.enterMacro_usage_identifier = function(ctx) {
};

// Exit a parse tree produced by vlogParser#macro_usage_identifier.
vlogParserListener.prototype.exitMacro_usage_identifier = function(ctx) {
};


// Enter a parse tree produced by vlogParser#inout_port_identifier.
vlogParserListener.prototype.enterInout_port_identifier = function(ctx) {
};

// Exit a parse tree produced by vlogParser#inout_port_identifier.
vlogParserListener.prototype.exitInout_port_identifier = function(ctx) {
};


// Enter a parse tree produced by vlogParser#input_port_identifier.
vlogParserListener.prototype.enterInput_port_identifier = function(ctx) {
};

// Exit a parse tree produced by vlogParser#input_port_identifier.
vlogParserListener.prototype.exitInput_port_identifier = function(ctx) {
};


// Enter a parse tree produced by vlogParser#instance_identifier.
vlogParserListener.prototype.enterInstance_identifier = function(ctx) {
};

// Exit a parse tree produced by vlogParser#instance_identifier.
vlogParserListener.prototype.exitInstance_identifier = function(ctx) {
};


// Enter a parse tree produced by vlogParser#library_identifier.
vlogParserListener.prototype.enterLibrary_identifier = function(ctx) {
};

// Exit a parse tree produced by vlogParser#library_identifier.
vlogParserListener.prototype.exitLibrary_identifier = function(ctx) {
};


// Enter a parse tree produced by vlogParser#module_identifier.
vlogParserListener.prototype.enterModule_identifier = function(ctx) {
};

// Exit a parse tree produced by vlogParser#module_identifier.
vlogParserListener.prototype.exitModule_identifier = function(ctx) {
};


// Enter a parse tree produced by vlogParser#module_instance_identifier.
vlogParserListener.prototype.enterModule_instance_identifier = function(ctx) {
};

// Exit a parse tree produced by vlogParser#module_instance_identifier.
vlogParserListener.prototype.exitModule_instance_identifier = function(ctx) {
};


// Enter a parse tree produced by vlogParser#net_identifier.
vlogParserListener.prototype.enterNet_identifier = function(ctx) {
};

// Exit a parse tree produced by vlogParser#net_identifier.
vlogParserListener.prototype.exitNet_identifier = function(ctx) {
};


// Enter a parse tree produced by vlogParser#output_port_identifier.
vlogParserListener.prototype.enterOutput_port_identifier = function(ctx) {
};

// Exit a parse tree produced by vlogParser#output_port_identifier.
vlogParserListener.prototype.exitOutput_port_identifier = function(ctx) {
};


// Enter a parse tree produced by vlogParser#parameter_identifier.
vlogParserListener.prototype.enterParameter_identifier = function(ctx) {
};

// Exit a parse tree produced by vlogParser#parameter_identifier.
vlogParserListener.prototype.exitParameter_identifier = function(ctx) {
};


// Enter a parse tree produced by vlogParser#port_identifier.
vlogParserListener.prototype.enterPort_identifier = function(ctx) {
};

// Exit a parse tree produced by vlogParser#port_identifier.
vlogParserListener.prototype.exitPort_identifier = function(ctx) {
};


// Enter a parse tree produced by vlogParser#real_identifier.
vlogParserListener.prototype.enterReal_identifier = function(ctx) {
};

// Exit a parse tree produced by vlogParser#real_identifier.
vlogParserListener.prototype.exitReal_identifier = function(ctx) {
};


// Enter a parse tree produced by vlogParser#simple_identifier.
vlogParserListener.prototype.enterSimple_identifier = function(ctx) {
};

// Exit a parse tree produced by vlogParser#simple_identifier.
vlogParserListener.prototype.exitSimple_identifier = function(ctx) {
};


// Enter a parse tree produced by vlogParser#specparam_identifier.
vlogParserListener.prototype.enterSpecparam_identifier = function(ctx) {
};

// Exit a parse tree produced by vlogParser#specparam_identifier.
vlogParserListener.prototype.exitSpecparam_identifier = function(ctx) {
};


// Enter a parse tree produced by vlogParser#system_function_identifier.
vlogParserListener.prototype.enterSystem_function_identifier = function(ctx) {
};

// Exit a parse tree produced by vlogParser#system_function_identifier.
vlogParserListener.prototype.exitSystem_function_identifier = function(ctx) {
};


// Enter a parse tree produced by vlogParser#system_task_identifier.
vlogParserListener.prototype.enterSystem_task_identifier = function(ctx) {
};

// Exit a parse tree produced by vlogParser#system_task_identifier.
vlogParserListener.prototype.exitSystem_task_identifier = function(ctx) {
};


// Enter a parse tree produced by vlogParser#task_identifier.
vlogParserListener.prototype.enterTask_identifier = function(ctx) {
};

// Exit a parse tree produced by vlogParser#task_identifier.
vlogParserListener.prototype.exitTask_identifier = function(ctx) {
};


// Enter a parse tree produced by vlogParser#terminal_identifier.
vlogParserListener.prototype.enterTerminal_identifier = function(ctx) {
};

// Exit a parse tree produced by vlogParser#terminal_identifier.
vlogParserListener.prototype.exitTerminal_identifier = function(ctx) {
};


// Enter a parse tree produced by vlogParser#text_macro_identifier.
vlogParserListener.prototype.enterText_macro_identifier = function(ctx) {
};

// Exit a parse tree produced by vlogParser#text_macro_identifier.
vlogParserListener.prototype.exitText_macro_identifier = function(ctx) {
};


// Enter a parse tree produced by vlogParser#topmodule_identifier.
vlogParserListener.prototype.enterTopmodule_identifier = function(ctx) {
};

// Exit a parse tree produced by vlogParser#topmodule_identifier.
vlogParserListener.prototype.exitTopmodule_identifier = function(ctx) {
};


// Enter a parse tree produced by vlogParser#udp_identifier.
vlogParserListener.prototype.enterUdp_identifier = function(ctx) {
};

// Exit a parse tree produced by vlogParser#udp_identifier.
vlogParserListener.prototype.exitUdp_identifier = function(ctx) {
};


// Enter a parse tree produced by vlogParser#udp_instance_identifier.
vlogParserListener.prototype.enterUdp_instance_identifier = function(ctx) {
};

// Exit a parse tree produced by vlogParser#udp_instance_identifier.
vlogParserListener.prototype.exitUdp_instance_identifier = function(ctx) {
};


// Enter a parse tree produced by vlogParser#variable_identifier.
vlogParserListener.prototype.enterVariable_identifier = function(ctx) {
};

// Exit a parse tree produced by vlogParser#variable_identifier.
vlogParserListener.prototype.exitVariable_identifier = function(ctx) {
};



exports.vlogParserListener = vlogParserListener;
