// HDLtool/tool/fsm/fsm_designer.js
"use strict";

export function get_hdl_code(language, reset_signal, reset_state, state_name, clock_name, complete) {
    let fsm = complete.fsm;
    let type = "one_process";
    let fsm_hdl = "";
    if (language === "vhdl") {
        let fsm_vhdl = new Vhdl_code();
        fsm_hdl = fsm_vhdl.get_vhdl_code(type, reset_signal, reset_state, state_name, clock_name, fsm);
    }
    else if (language === "verilog") {
        let fsm_vhdl = new Verilog_code();
        fsm_hdl = fsm_vhdl.get_verilog_code(type, reset_signal, reset_state, state_name, clock_name, fsm);
    }
    return fsm_hdl;
}

////////////////////////////////////////////////////////////////////////////////
// VHDL
////////////////////////////////////////////////////////////////////////////////
class Vhdl_code {
    constructor() {
        this.indent = "  ";
        this.indent1 = "  ";
        this.indent2 = this.indent1 + this.indent1;
        this.indent3 = this.indent2 + this.indent;
        this.indent4 = this.indent3 + this.indent;
    }


    get_vhdl_code(type, reset_signal, reset_state,
        state_name, clock_name, fsm) {
        let fsm_hdl = "";
        if (type === "one_process") {
            fsm_hdl = this.get_vhdl_one_process(reset_signal, reset_state, state_name, clock_name, fsm);
        }
        return fsm_hdl;
    }

    get_vhdl_one_process(reset_signal, reset_state, state_name, clock_name, fsm) {
        let fsm_hdl = this.example_state_declaration(state_name, fsm);
        fsm_hdl += this.open_process(clock_name);
        let extra_indent = '';
        if (reset_signal !== '' && reset_state !== '') {
            fsm_hdl += this.open_reset(reset_signal, reset_state, state_name);
            extra_indent = this.indent1;
        }
        fsm_hdl += this.open_case(state_name, extra_indent);
        for (let i = 0; i < fsm.length; ++i) {
            if (fsm[i].transitions.length !== 0 || fsm[i].outputs.length !== 0) {
                fsm_hdl += this.open_when(fsm[i].name, extra_indent);
                fsm_hdl += this.add_outputs(fsm[i].outputs, extra_indent);
                if (fsm[i].transitions.length !== 0) {
                    fsm_hdl += this.generate_if_case(fsm[i].transitions, state_name, extra_indent);
                }
            }
        }
        fsm_hdl += this.close_case(extra_indent);
        if (reset_signal !== '' && reset_state !== '') {
            fsm_hdl += this.close_reset();
        }
        fsm_hdl += this.close_process();
        return fsm_hdl;
    }

    example_state_declaration(state_name, fsm) {
        let str = '';

        str += `-- type t_${state_name} is (`;
        for (let i = 0; i < fsm.length - 1; ++i) {
            str += `${fsm[i].name}, `;
        }
        str += `${fsm[fsm.length - 1].name});\n`;
        str += `-- signal ${state_name} : type type_${state_name};\n\n`;
        return str;
    }

    open_reset(reset_signal, reset_state, state_name) {
        let str = "";
        str += this.indent1 + `if (${reset_signal} = '1') then\n`;
        str += this.indent2 + `${state_name} <= ${reset_state};\n`;
        str += this.indent1 + `else\n`;
        return str;
    }

    close_reset() {
        let str = "";
        str += this.indent1 + 'end if;\n';
        return str;
    }

    open_when(state_name, extra_indent) {
        let str = "";
        str += extra_indent + this.indent2 + `when ${state_name} =>\n`;
        return str;
    }

    add_outputs(outputs, extra_indent) {
        let str = "";
        for (let i = 0; i < outputs.length; ++i) {
            let output = outputs[i];
            output = output.trim();
            if (output[output.length - 1] !== ';') {
                output += ';';
            }
            str += extra_indent + this.indent3 + `${output}\n`;
        }
        return str;
    }

    generate_if_case(transitions, state_name, extra_indent) {
        let str = "";
        for (let i = 0; i < transitions.length; ++i) {
            if (i === 0) {
                str += extra_indent + this.indent3 + `if (${transitions[i].condition}) then\n`;
                str += extra_indent + this.indent4 + `${state_name} <= ${transitions[i].destination};\n`;
            }
            else {
                str += extra_indent + this.indent3 + `elsif (${transitions[i].condition}) then\n`;
                str += extra_indent + this.indent4 + `${state_name} <= ${transitions[i].destination};\n`;
            }
        }
        str += extra_indent + this.indent3 + "end if;\n";
        return str;
    }

    open_process(clock_name) {
        let str = "";
        str += `process(${clock_name}) begin\n`;
        return str;
    }

    close_process() {
        let str = "";
        str += "end process;\n";
        return str;
    }

    open_case(state_name, extra_indent) {
        let str = "";
        str += extra_indent + this.indent + `case ${state_name} is\n`;
        return str;
    }
    close_case(extra_indent) {
        let str = "";
        str += extra_indent + this.indent + "end case;\n";
        return str;
    }
}

////////////////////////////////////////////////////////////////////////////////
// Verilog
////////////////////////////////////////////////////////////////////////////////
class Verilog_code {
    constructor() {
        this.indent1 = "  ";
        this.indent2 = this.indent1 + this.indent1;
        this.indent3 = this.indent2 + this.indent1;
        this.indent4 = this.indent3 + this.indent1;
        this.indent5 = this.indent4 + this.indent1;
        this.indent6 = this.indent5 + this.indent1;
    }


    get_verilog_code(type, reset_signal, reset_state,
        state_name, clock_name, fsm) {
        let fsm_hdl = "";
        if (type === "one_process") {
            fsm_hdl = this.get_verilog_one_process(reset_signal, reset_state, state_name, clock_name, fsm);

        }
        return fsm_hdl;
    }

    get_verilog_one_process(reset_signal, reset_state, state_name, clock_name, fsm) {
        let fsm_hdl = this.example_state_declaration(state_name, fsm);
        let extra_indent = '';
        fsm_hdl += this.open_process(clock_name);
        if (reset_signal !== '' && reset_state !== '') {
            fsm_hdl += this.open_reset(reset_signal, reset_state, state_name);
            extra_indent = this.indent1;
        }
        fsm_hdl += this.open_case(extra_indent);
        for (let i = 0; i < fsm.length; ++i) {
            if (fsm[i].transitions.length !== 0 || fsm[i].outputs.length !== 0) {
                fsm_hdl += this.open_when(fsm[i].name, extra_indent);
                fsm_hdl += this.add_outputs(fsm[i].outputs, extra_indent);
                if (fsm[i].transitions.length !== 0) {
                    fsm_hdl += this.generate_if_case(fsm[i].transitions, state_name, extra_indent);
                }
                fsm_hdl += this.close_when(extra_indent);
            }
        }
        fsm_hdl += this.close_case(extra_indent);
        if (reset_signal !== '' && reset_state !== '') {
            fsm_hdl += this.close_reset();
        }
        fsm_hdl += this.close_process();
        return fsm_hdl;
    }

    open_reset(reset_signal, reset_state, state_name) {
        let str = '';
        str += this.indent1 + `if (${reset_signal}) begin\n`;
        str += this.indent2 + `${state_name} = ${reset_state};\n`;
        str += this.indent1 + 'end\n';
        str += this.indent1 + 'else begin\n';
        return str;
    }

    close_reset() {
        let str = "";
        str += this.indent1 + 'end\n';
        return str;
    }

    example_state_declaration(state_name, fsm) {
        let str = '';
        str += `//  reg [${fsm.length}-1:0] ${state_name};\n`;
        str += '//  localparam \n';
        for (let i = 0; i < fsm.length - 1; ++i) {
            str += '//  ' + this.indent1 + `${fsm[i].name} = ${i},\n`;
        }
        str += '//  ' + this.indent1 + `${fsm[fsm.length - 1].name} = ${fsm.length - 1};\n\n`;
        return str;
    }

    open_when(state_name, extra_indent) {
        let str = "";
        str += extra_indent + this.indent2 + `${state_name}:\n`;
        str += extra_indent + this.indent2 + `begin\n`;
        return str;
    }

    close_when(extra_indent) {
        let str = "";
        str += extra_indent + this.indent2 + `end\n`;
        return str;
    }

    add_outputs(outputs, extra_indent) {
        let str = "";
        for (let i = 0; i < outputs.length; ++i) {
            let output = outputs[i];
            output = output.trim();
            if (output[output.length - 1] !== ';') {
                output += ';';
            }
            str += extra_indent + this.indent3 + `${output}\n`;
        }
        return str;
    }

    generate_if_case(transitions, state_name, extra_indent) {
        let str = "";
        for (let i = 0; i < transitions.length; ++i) {
            if (i === 0) {
                str += extra_indent + this.indent3 + `if (${transitions[i].condition}) begin\n`;
                str += extra_indent + this.indent4 + `${state_name} = ${transitions[i].destination};\n`;
                str += extra_indent + this.indent3 + `end\n`;
            }
            else {
                str += extra_indent + this.indent3 + `else if (${transitions[i].condition}) begin\n`;
                str += extra_indent + this.indent4 + `${state_name} = ${transitions[i].destination};\n`;
                str += extra_indent + this.indent3 + `end\n`;
            }
        }
        return str;
    }

    open_process(clock_name) {
        let str = "";
        str += `always @(posedge ${clock_name}) begin\n`;
        return str;
    }

    close_process() {
        let str = "";
        str += "end\n";
        return str;
    }

    open_case(extra_indent) {
        let str = "";
        str += extra_indent + this.indent1 + "case(state)\n";
        return str;
    }
    close_case(extra_indent) {
        let str = "";
        str += extra_indent + this.indent1 + "endcase\n";
        return str;
    }
}
