// HDLtool/sim/Instance.js
"use strict";

const vscode = require("vscode");
const parser = require("HDLparser");

function instantiateModuleInteract(indexer, workspacePath) {
    var doc = vscode.window.activeTextEditor.document;
    parser.simParser.selectModule(indexer.HDLparam, workspacePath).then(selectModule => {
        if (selectModule != null) {
            if (doc.languageId == "systemverilog" || doc.languageId == "verilog") {
                let inst = instantiateVlogModule(selectModule);
                selectInsert(inst);
            } else if (doc.languageId == "vhdl") {
                let inst = instantiateVhdlModule(selectModule);
                selectInsert(inst);
            }
        }
    })
}
exports.instantiateModuleInteract = instantiateModuleInteract;

function selectInsert(content) {
    const editor = vscode.window.activeTextEditor;
    if (editor === undefined) {
        return;
    }
    const selections = editor.selections;
    editor.edit((editBuilder) => {
        selections.forEach((selection) => {
            editBuilder.insert(selection.active, content);
        });
    });
}
exports.selectInsert = selectInsert;

function instantiateVlogModule(module) {
    // module 2001 style
    let moduleName = module.moduleName;

    let initPortMax_len = 0;
    let initPort = '';
    for (let i = 0; i < module.port.output.length; i++) {
        if (module.port.output[i].portWidth.length > initPortMax_len) {
            initPortMax_len = module.port.output[i].portWidth.length;
        }
    }
    for (let i = 0; i < module.port.output.length; i++) {
        let element = module.port.output[i].portWidth;
        let padding = initPortMax_len - element.length + 1;
        element = element.replace('(','[').replace(')',']').replace(/downto/mgi,':') + ' '.repeat(padding);
        initPort += `wire ${element}\t${module.port.output[i].portName};\n`;
    }

    let paramString = ``;
    if (module.param.length > 0) {
        paramString = `#(\n${instantiateVlogParam(module.param)})\n`;
    }
    let portString = ``;
    portString = instantiateVlogPort(module.port);

    let instContent = initPort + "\n";
    instContent += moduleName;
    instContent += " ";
    instContent += paramString;
    instContent += `u_${moduleName}(\n`;
    instContent += portString;
    instContent += ');\n';

    return instContent;
}
exports.instantiateVlogModule = instantiateVlogModule;

function instantiateVhdlModule(module) {
    // module 2001 style
    let moduleName = module.moduleName;
    // let space = '';
    // const editor = vscode.window.activeTextEditor;
    // if (editor === undefined) {
    //     return;
    // }
    // const selections = editor.selections;
    // if (selections.length == 1) {
    //     space = ''.repeat(selections[0].active.character);
    // }
    // let initPortMax_len = 0;
    // let initPort = '';
    // for (let i = 0; i < module.port.output.length; i++) {
    //     if (module.port.output[i].portWidth.length > initPortMax_len) {
    //         initPortMax_len = module.port.output[i].portWidth.length;
    //     }
    // }
    // for (let i = 0; i < module.port.output.length; i++) {
    //     let element = module.port.output[i].portWidth;
    //     let padding = initPortMax_len - element.length + 1;
    //     element = element + ' '.repeat(padding);
    //     initPort += `wire ${element}\t${module.port.output[i].portName};\n`;
    // }

    let paramString = ``;
    if (module.param.length > 0) {
        paramString = `generic map(\n${instantiateVhdlParam(module.param)})\n`;
    }
    let portString = ``;
    portString = `port map(\n${instantiateVhdlPort(module.port)})`;

    // let instContent = initPort + "\n";
    let instContent = `${moduleName} : u_${moduleName}\n`;
    instContent += paramString;
    instContent += portString;
    instContent += ';\n';

    return instContent;
}
exports.instantiateVhdlModule = instantiateVhdlModule;

function instantiateVlogPort(ports) {
    let port = '';
    let index = 0;
    let max_len = 0;
    for (index = 0; index < ports.length; index++) {
        if (ports.input[index].portName.length > max_len) {
            max_len = ports[index].portName.length;
        }
    }
    for (index = 0; index < ports.output.length; index++) {
        if (ports.output[index].portName.length > max_len) {
            max_len = ports.output[index].portName.length;
        }
    }
    for (index = 0; index < ports.inout.length; index++) {
        if (ports.inout[index].portName.length > max_len) {
            max_len = ports.inout[index].portName.length;
        }
    }
    // .NAME(NAME)
    port += `\t//input\n`;
    for (let i = 0; i < ports.input.length; i++) {
        let element = ports.input[i].portName;
        let padding = max_len - element.length + 1;
        element = element + ' '.repeat(padding);
        if (ports.output.length == 0 ) {
            port += `\t.${element}\t\t( ${element}\t\t)`;
            if (i !== ports.input.length - 1) {
                port += ',';
            }
            port += '\n';
        } else {
            port += `\t.${element}\t\t( ${element}\t\t),\n`;
        }
    }

    port += `\n\t//output\n`;
    for (let i = 0; i < ports.output.length; i++) {
        let element = ports.output[i].portName;
        let padding = max_len - element.length + 1;
        element = element + ' '.repeat(padding);
        if (ports.inout.length == 0 ) {
            port += `\t.${element}\t\t( ${element}\t\t)`;
            if (i !== ports.output.length - 1) {
                port += ',';
            }
            port += '\n';
        } else {
            port += `\t.${element}\t\t( ${element}\t\t),\n`;
        }
    }

    port += `\n\t//inout\n`;
    for (let i = 0; i < ports.inout.length; i++) {
        let element = ports.inout[i].portName;
        let padding = max_len - element.length + 1;
        element = element + ' '.repeat(padding);
        port += `\t.${element}\t\t( ${element}\t\t)`;
        if (i !== ports.length - 1) {
            port += ',';
        }
        port += '\n';
    }
    return port;
}

function instantiateVlogParam(params) {
    let param = '';
    let nameMax_len = 0;
    let initMax_len = 0;
    for (let i = 0; i < params.length; i++) {
        if (params[i].paramName.length > nameMax_len) {
            nameMax_len = params[i].paramName.length;
        }
    }

    for (let i = 0; i < params.length; i++) {
        params[i].paramInit = params[i].paramInit.replace(',','');
        if (params[i].paramInit.length > initMax_len) {
            initMax_len = params[i].paramInit.length;
        }
    }
    // .NAME  ( INIT  ),
    for (let i = 0; i < params.length; i++) {
        let elementName = params[i].paramName;
        let elementInit = params[i].paramInit;

        let namePadding = nameMax_len - elementName.length + 1;
        let initPadding = initMax_len - elementInit.length + 1;

        elementName = elementName + ' '.repeat(namePadding);
        elementInit = elementInit + ' '.repeat(initPadding);

        param += `\t.${elementName}\t\t( ${elementInit}\t\t)`;
        if (i !== params.length - 1) {
            param += ',';
            param += '\n';
        }
    }
    return param;
}

function instantiateVhdlPort(ports) {
    let port = '';
    let index = 0;
    let max_len = 0;
    for (index = 0; index < ports.length; index++) {
        if (ports.input[index].portName.length > max_len) {
            max_len = ports[index].portName.length;
        }
    }
    for (index = 0; index < ports.output.length; index++) {
        if (ports.output[index].portName.length > max_len) {
            max_len = ports.output[index].portName.length;
        }
    }
    for (index = 0; index < ports.inout.length; index++) {
        if (ports.inout[index].portName.length > max_len) {
            max_len = ports.inout[index].portName.length;
        }
    }
    // NAME => NAME,
    port += `\t-- input\n`;
    for (let i = 0; i < ports.input.length; i++) {
        let element = ports.input[i].portName;
        let padding = max_len - element.length + 1;
        element = element + ' '.repeat(padding);
        if (ports.output.length == 0 ) {
            port += `\t${element} => ${ports.input[i].portName}`;
            if (i !== ports.input.length - 1) {
                port += ',';
            }
            port += '\n';
        } else {
            port += `\t${element} => ${ports.input[i].portName},\n`;
        }
    }

    port += `\n\t-- output\n`;
    for (let i = 0; i < ports.output.length; i++) {
        let element = ports.output[i].portName;
        let padding = max_len - element.length + 1;
        element = element + ' '.repeat(padding);
        if (ports.inout.length == 0 ) {
            port += `\t${element} => ${ports.output[i].portName}`;
            if (i !== ports.output.length - 1) {
                port += ',';
            }
            port += '\n';
        } else {
            port += `\t${element} => ${ports.output[i].portName},\n`;
        }
    }

    port += `\n\t-- inout\n`;
    for (let i = 0; i < ports.inout.length; i++) {
        let element = ports.inout[i].portName;
        let padding = max_len - element.length + 1;
        element = element + ' '.repeat(padding);
        port += `\t${element} => ${ports.inout[i].portName}`;
        if (i !== ports.length - 1) {
            port += ',';
        }
        port += '\n';
    }
    return port;
}

function instantiateVhdlParam(params) {
    let param = '';
    let nameMax_len = 0;
    // let initMax_len = 0;
    for (let i = 0; i < params.length; i++) {
        if (params[i].paramName.length > nameMax_len) {
            nameMax_len = params[i].paramName.length;
        }
    }

    // for (let i = 0; i < params.length; i++) {
    //     params[i].paramInit = params[i].paramInit.replace(',','');
    //     if (params[i].paramInit.length > initMax_len) {
    //         initMax_len = params[i].paramInit.length;
    //     }
    // }
    // NAME => NAME,
    for (let i = 0; i < params.length; i++) {
        let elementName = params[i].paramName;
        let elementInit = params[i].paramInit;

        let namePadding = nameMax_len - elementName.length + 1;
        // let initPadding = initMax_len - elementInit.length + 1;

        elementName = elementName + ' '.repeat(namePadding);
        // elementInit = elementInit + ' '.repeat(initPadding);

        param += `\t${elementName} => ${elementInit}`;
        if (i !== params.length - 1) {
            param += ',';
            param += '\n';
        }
    }
    return param;
}
