// HDLtool/lsp/completion.js
"use strict";

const vscode = require("vscode");
const parser = require("HDLparser");

const ENT_BEGIN = /\s*entity\s+(\w*)\s+is.*/;
const ARCH_BEGIN = /\s*architecture\s+(\w*)\s+of\s+(\w*)\s+is.*/;
const CONF_BEGIN = /\s*configuration\s+(\w*)\s+of\s+(\w*)\s+is.*/;
const SCOPE_END = /\s*end\s+(\w*).*/;

let kwLibrary       = createCompletionKeyword('library');
let kwUse           = createCompletionKeyword('use');
let kwPackage       = createCompletionKeyword('package');
let kwArchitecture  = createCompletionKeyword('architecture');
let kwEntity        = createCompletionKeyword('entity');
let kwConfiguration = createCompletionKeyword('configuration');
let kwIs            = createCompletionKeyword('is');
let kwBegin         = createCompletionKeyword('begin');
let kwEnd           = createCompletionKeyword('end');
let kwMap           = createCompletionKeyword('map');
let kwOf            = createCompletionKeyword('of');
let kwFor           = createCompletionKeyword('for');
let operatorOptions = [
    createCompletionOption('abs'),
    createCompletionOption('and'),
    createCompletionOption('mod'),
    createCompletionOption('nand'),
    createCompletionOption('nor'),
    createCompletionOption('not'),
    createCompletionOption('or'),
    createCompletionOption('rem'),
    createCompletionOption('rol'),
    createCompletionOption('ror'),
    createCompletionOption('sla'),
    createCompletionOption('sll'),
    createCompletionOption('sra'),
    createCompletionOption('srl'),
    createCompletionOption('xnor'),
    createCompletionOption('xor'),
];
let archTypeOptions = [
    createCompletionOption('array'),
    createCompletionOption('type'),
    createCompletionOption('component'),
    createCompletionOption('constant'),
    createCompletionOption('signal'),
    createCompletionOption('subtype'),
    createCompletionOption('variable'),
    createCompletionOption('assert'),
    createCompletionOption('severity'),
    createCompletionOption('report'),
    createCompletionOption('process'),
    createCompletionOption('with'),
    createCompletionOption('select'),
    createCompletionOption('when'),
    createCompletionOption('others'),
    createCompletionOption('block'),
    createCompletionOption('function'),
    createCompletionOption('procedure'),
    createCompletionOption('case'),
    createCompletionOption('else'),
    createCompletionOption('elsif'),
    createCompletionOption('for'),
    createCompletionOption('generate'),
    createCompletionOption('if'),
    createCompletionOption('loop'),
    createCompletionOption('map'),
    createCompletionOption('next'),
    createCompletionOption('others'),
    createCompletionOption('return'),
    createCompletionOption('wait'),
    createCompletionOption('then'),
    createCompletionOption('return'),
    createCompletionOption('when'),
    createCompletionOption('while'),
];
let portTypeOptions = [
    createCompletionOption('in'),
    createCompletionOption('out'),
    createCompletionOption('inout'),
    createCompletionOption('buffer'),
    createCompletionOption('linkage'),
];
let entityOptions = [
    createCompletionOption('generic'),
    createCompletionOption('port'),
];
let scalaTypes = [
    createCompletionKeyword('bit', `The bit data type can only have the value 0 or 1.`),
    createCompletionKeyword('bit_vector', `The bit_vector data type is the vector version of the bit type consisting of two or more bits. Each bit in a bit_vector can only have the value 0 or 1.`),
    createCompletionKeyword('boolean', `True or false`),
    createCompletionKeyword('integer', `32-bit	integers.`),
    createCompletionKeyword('natural', `non	negative integer.`),
    createCompletionKeyword('positive', `positive	integer.`),
    createCompletionKeyword('real', `floating point number.`),
    createCompletionKeyword('time', `Time in fs,	ps,	ns,	us,	ms,	sec, min, hr`),
    createCompletionKeyword('character', ``),
    createCompletionKeyword('string', `String for VHDL.`),
    createCompletionOption('downto'),
    createCompletionOption('std_logic'),
];

(function (VhdlScopeKind) {
    VhdlScopeKind[VhdlScopeKind["Vhdl"] = 0] = "Vhdl";
    VhdlScopeKind[VhdlScopeKind["Entity"] = 1] = "Entity";
    VhdlScopeKind[VhdlScopeKind["Architecture"] = 2] = "Architecture";
    VhdlScopeKind[VhdlScopeKind["Configuration"] = 3] = "Configuration";
})(exports.VhdlScopeKind || (exports.VhdlScopeKind = {}));

function guessScope(doc, cursorLineNum) {
    return new ScopeGuesser(cursorLineNum).guess(doc);
}

class VhdlScope {
    constructor(kind, lineFrom) {
        this.kind = kind;
        this.children = [];
        this.lineFrom = lineFrom;
    }
    addChild(child) {
        this.children.push(child);
        child.parent = this;
    }
}

class ScopeGuesser {
    constructor(cursorLineNum) {
        this.cursorLineNum = cursorLineNum;
    }
    guess(doc) {
        this.enterScope(VhdlScopeKind.Vhdl, 0);
        for (var i = 0; i < doc.lineCount; i++) {
            var line = doc.lineAt(i);
            if (!line.isEmptyOrWhitespace) {
                let lineText = line.text;
                if (lineText.match(/^\s*\-\-/)) {
                    continue;
                }
                else if (lineText.match(ENT_BEGIN)) {
                    this.enterScope(VhdlScopeKind.Entity, i);
                }
                else if (lineText.match(ARCH_BEGIN)) {
                    this.enterScope(VhdlScopeKind.Architecture, i);
                }
                else if (lineText.match(CONF_BEGIN)) {
                    this.enterScope(VhdlScopeKind.Configuration, i);
                }
                else if (lineText.match(SCOPE_END)) {
                    this.exitScope(i);
                }
            }
        }
        this.exitScope(doc.lineCount);
        return this.scopeAtCursor;
    }
    enterScope(kind, lineNum) {
        let newScope = new VhdlScope(kind, lineNum);
        if (this.currentScope) {
            this.currentScope.addChild(newScope);
        }
        this.currentScope = newScope;
    }
    exitScope(lineNum) {
        this.currentScope.lineTo = lineNum;
        if (!this.scopeAtCursor) {
            if (this.currentScope.lineFrom <= this.cursorLineNum
                && this.currentScope.lineTo >= this.cursorLineNum) {
                this.scopeAtCursor = this.currentScope;
            }
        }
        if (this.currentScope.parent) {
            this.currentScope = this.currentScope.parent;
        }
    }
}

function createCompletionKeyword(label, doc) {
    let item = new vscode.CompletionItem(label);
    item.kind = vscode.CompletionItemKind.Keyword;
    if (doc) {
        item.documentation = doc;
    }
    return item;
}
function createCompletionOption(option, doc) {
    let item = new vscode.CompletionItem(option);
    item.kind = vscode.CompletionItemKind.Value;
    item.documentation = doc;
    return item;
}
class vhdlCompletion {
    provideCompletionItems(document, position, token) {
        return new Promise((resolve, reject) => {
            let filename = document.fileName;
            let lineText = document.lineAt(position.line).text;
            if (lineText.match(/^\s*\-\-/)) {
                return resolve([]);
            }
            let inString = false;
            if ((lineText.substring(0, position.character).match(/\"/g) || []).length % 2 === 1) {
                inString = true;
            }
            let suggestions = [];
            let textBeforeCursor = lineText.substring(0, position.character - 1);
            let scope = guessScope(document, position.line);
            //console.log(scope.syntax);
            //console.log(textBeforeCursor);
            switch (scope.kind) {
                case VhdlScopeKind.Vhdl: {
                    suggestions.push(kwArchitecture);
                    suggestions.push(kwBegin);
                    suggestions.push(kwConfiguration);
                    suggestions.push(kwEnd);
                    suggestions.push(kwEntity);
                    suggestions.push(kwIs);
                    suggestions.push(kwPackage);
                    suggestions.push(kwUse);
                    suggestions.push(kwLibrary);
                    break;
                }
                case VhdlScopeKind.Entity: {
                    if (textBeforeCursor.match(/^\s*\w*$/)) {
                        suggestions.push(...entityOptions);
                        suggestions.push(...portTypeOptions);
                        suggestions.push(kwBegin);
                        suggestions.push(kwEnd);
                    }
                    else if (textBeforeCursor.match(/(in|out|inout|buffer|linkage)\s*$/)) {
                        suggestions.push(...scalaTypes);
                    }
                    break;
                }
                case VhdlScopeKind.Architecture: {
                    if (textBeforeCursor.match(/^\s*\w*$/)) {
                        suggestions.push(...archTypeOptions);
                        suggestions.push(kwBegin);
                        suggestions.push(kwEnd);
                        suggestions.push(kwIs);
                        suggestions.push(kwOf);
                    }
                    else if (textBeforeCursor.match(/(in|out|inout|buffer|linkage)\s*$/)) {
                        suggestions.push(...scalaTypes);
                    }
                    else if (textBeforeCursor.match(/(signal|variable|constant|subtype|type|array)\s*\w*:\s*$/)) {
                        suggestions.push(...scalaTypes);
                    }
                    else if (textBeforeCursor.match(/(<=|:=)\s*\w*\s*$/)) {
                        suggestions.push(...operatorOptions);
                    }
                    break;
                }
                case VhdlScopeKind.Configuration: {
                    suggestions.push(kwFor);
                    break;
                }
            }
            return resolve(suggestions);
        });
    }
}
exports.vhdlCompletion = vhdlCompletion;


class vlogCompletion {
    constructor(HDLparam){
        this.HDLparam = HDLparam;
    }
    provideCompletionItems(document, position, token) {
        return new Promise((resolve, reject) => {
            var vlog_instantiation = new RegExp([
                /(?:(?<modifier>virtual|static|automatic|rand|randc|pure virtual)\s+)?/,
                /\b(?<type>[:\w]+)\s*/,
                /(?<params>#\s*\(\s*\.[\w\W]*?\))?\s*/,
                /(\b\w+\s*,\s*)*?/,
                /\b(?<name>\w+)\s*/,
                /(?:(\(\s*\.[\w\W]*?\)))\s*/,
                /\s*(?<end>;)/
            ].map(x => (typeof x === 'string') ? x : x.source).join(''), 'mg');
            let suggestions = [];
            let content = document.getText();
            let completionItemsIndex = this.line_index_to_character_index(content, position);
            // completionItemsIndex = this.positionToIndex(content, position);
            let IllegalRange = parser.simParser.getCommentRange(content,0);
            content = parser.simParser.delComments(content);
            while (1) {
                // 匹配每个元素
                let match = vlog_instantiation.exec(content);

                // 判断元素是否有效--是否为空 或者 是否在非法范围内
                if (match == null) {
                    break;      // 即跳出此类型的匹配，执行下一类型
                }
                else if ((match.index == 0) || parser.simParser.isIllegalRange(match,IllegalRange,0)) {
                    continue;   // 跳过匹配组中的一个
                }
                let paramIndexStart = match.index + match[0].indexOf(this.getMatchContent(match[3]));
                let paramIndexEnd   = paramIndexStart + this.getMatchLength(match[3]);
                let portIndexStart  = match.index + match[0].indexOf(this.getMatchContent(match[6]));
                let portIndexEnd    = portIndexStart  + this.getMatchLength(match[6]);
                if (completionItemsIndex >= paramIndexStart && 
                    completionItemsIndex <= paramIndexEnd) {
                    suggestions = this.getParamInfo(this.HDLparam, match.groups.type);
                }
                if (completionItemsIndex >= portIndexStart && 
                    completionItemsIndex <= portIndexEnd) {
                    suggestions = this.getPortInfo(this.HDLparam, match.groups.type);
                }
            }
            return resolve(suggestions);
        });
    }
    getParamInfo(HDLparam, moduleName){
        let index = 0
        let paramList = [];
        for (index = 0; index < HDLparam.length; index++) {
            const element = HDLparam[index];
            if (element.moduleName == moduleName) {
                break;
            }
        }
        for (let i = 0; i < HDLparam[index].param.length; i++) {
            const element = HDLparam[index].param[i];
            let item = new vscode.CompletionItem(element.paramName);
            item.kind = vscode.CompletionItemKind.Value;
            paramList.push(item);
        }
        return paramList;
    }
    getPortInfo(HDLparam, moduleName){
        let index = 0
        let portList = [];
        for (index = 0; index < HDLparam.length; index++) {
            const element = HDLparam[index];
            if (element.moduleName == moduleName) {
                break;
            }
        }
        for (let i = 0; i < HDLparam[index].port.input.length; i++) {
            const element = HDLparam[index].port.input[i];
            let item = new vscode.CompletionItem(element.portName);
            item.kind = vscode.CompletionItemKind.Interface;
            portList.push(item);
        }
        for (let i = 0; i < HDLparam[index].port.output.length; i++) {
            const element = HDLparam[index].port.output[i];
            let item = new vscode.CompletionItem(element.portName);
            item.kind = vscode.CompletionItemKind.Interface;
            portList.push(item);
        }
        for (let i = 0; i < HDLparam[index].port.inout.length; i++) {
            const element = HDLparam[index].port.inout[i];
            let item = new vscode.CompletionItem(element.portName);
            item.kind = vscode.CompletionItemKind.Interface;
            portList.push(item);
        }
        return portList;
    }
    getMatchLength(match){
        if (match != undefined) {
            return match.length;
        } else {
            return 0;
        }
    }
    getMatchContent(match){
        if (match != undefined) {
            return match;
        } else {
            return '';
        }
    }
    line_index_to_character_index(content, position) {
        let content_split = content.split('\n');
        let character_index = 0;
        for (let i = 0; i < position.line; ++i) {
            character_index += content_split[i].length + 1;
        }
        character_index += position.character;
        return character_index;
    }
}
exports.vlogCompletion = vlogCompletion;
