// HDLtool/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};

const vscode = require("vscode");

const tree  = require("./tree/tree");
const build = require("./build/build");

const lspAlignment  = require("./lsp/alignment");
const lspProvider   = require("./lsp/provider");
const lspCompletion = require("./lsp/completion");

const simulation   = require("./sim/simulation");
const simInstance  = require("./sim/Instance");
const simTestbench = require("./sim/Testbench");

function registerLspServer(context, indexer, HDLparam) {
    
    context.subscriptions.push(vscode.commands.registerCommand('FPGA.alignment', lspAlignment.smartIndent));
    // Configure Provider manager
    const HDL_lsp = [
        { scheme: 'file', language: 'systemverilog' }, 
        { scheme: 'file', language: 'verilog' },
        { scheme: 'file', language: 'vhdl' }
    ];
    
    // VHDL Language sever
    vscode.languages.setLanguageConfiguration(HDL_lsp[2].language, {
        indentationRules: {
            increaseIndentPattern: /^.*(begin|then|loop|is)$/,
            decreaseIndentPattern: /^end\s+\w*$/
        },
        wordPattern: /(-?\d*\.\d\w*)|([^\`\~\!\@\#\%\^\&\*\(\)\-\=\+\[\{\]\}\\\|\;\:\'\"\,\.\<\>\/\?\s]+)/g,
    });
    context.subscriptions.push(vscode.languages.registerCompletionItemProvider(
        HDL_lsp[2], 
        new lspCompletion.vhdlCompletion(), 
        ['.', '\"'])
    );
    context.subscriptions.push(vscode.languages.registerCompletionItemProvider(
        HDL_lsp[1], 
        new lspCompletion.vlogCompletion(HDLparam), 
        '.')
    );

    // Providers
    const hovProvider = new lspProvider.HoverProvider();
    const numProvider = new lspProvider.numberHoverProvider();
    const docProvider = new lspProvider.DocumentSymbolProvider(indexer);
    const symProvider = new lspProvider.WorkspaceSymbolProvider(indexer);
    const defProvider = new lspProvider.DefinitionProvider(symProvider);
    context.subscriptions.push(vscode.languages.registerHoverProvider(HDL_lsp, hovProvider));
    context.subscriptions.push(vscode.languages.registerHoverProvider(HDL_lsp, numProvider));
	context.subscriptions.push(vscode.languages.registerWorkspaceSymbolProvider(symProvider));
    context.subscriptions.push(vscode.languages.registerDefinitionProvider(HDL_lsp, defProvider));
    context.subscriptions.push(vscode.languages.registerDocumentSymbolProvider(HDL_lsp, docProvider));
}
exports.registerLspServer = registerLspServer;

function registerSimServer(context, HDLparam, opeParam){
    context.subscriptions.push(vscode.commands.registerCommand('FPGA.instance', () => {
        simInstance.instantiateModuleInteract(HDLparam);
    }));
    context.subscriptions.push(vscode.commands.registerCommand('FPGA.testbench', () => {
        simTestbench.genInstancetoTbFile(HDLparam, tbFilePath);
    }));
    context.subscriptions.push(vscode.commands.registerCommand('FPGA.simulate', () => {
        
    }));
    context.subscriptions.push(vscode.commands.registerCommand('FPGA.Overwrite_tb', () => {
        simulation.Overwrite_tb(opeParam);
    }));
}
exports.registerSimServer = registerSimServer;

function registerTreeServer(opeParam, HDLparam){
    // // Tree View
    vscode.window.registerTreeDataProvider('TOOL.sdk_tree' , new tree.sdkTreeDataProvider(opeParam));
    vscode.window.registerTreeDataProvider('TOOL.fpga_tree', new tree.fpgaTreeDataProvider(opeParam));
    vscode.window.registerTreeDataProvider('TOOL.Tool_tree', new tree.toolTreeDataProvider(opeParam));
    new tree.FileExplorer(HDLparam);
}
exports.registerTreeServer = registerTreeServer;

function registerBuildServer(context, HDLparam){
    let netlist_viewer_manager = new build.synthNetlist(context);
    context.subscriptions.push(vscode.commands.registerCommand('FPGA.netlist', () => __awaiter(this, void 0, void 0, function* () {
        yield netlist_viewer_manager.open_viewer();
    })), vscode.workspace.onDidOpenTextDocument((e) => netlist_viewer_manager.update_viewer()), vscode.workspace.onDidSaveTextDocument((e) => netlist_viewer_manager.update_viewer()), vscode.window.onDidChangeVisibleTextEditors((e) => netlist_viewer_manager.update_visible_viewer(e)));
}
exports.registerBuildServer = registerBuildServer;
