// HDLlinter/linter/Verilator.js

const Base_linter = require('./base_linter');

class Verilator extends Base_linter {
    constructor() {
        super();
        // SystemVerilog
        this.SVLOGPARAMETERS = {
            'SYNT': "verilator --lint-only -sv -Wall -bbox-sys --bbox-unsup -DGLBL",
            'SYNT_WINDOWS': "verilator.exe --lint-only -sv -Wall -bbox-sys --bbox-unsup -DGLBL"
        };
        // Verilog
        this.VLOGPARAMETERS = {
            'SYNT': "verilator --lint-only -Wall -bbox-sys --bbox-unsup -DGLBL",
            'SYNT_WINDOWS': "verilator.exe --lint-only -Wall -bbox-sys --bbox-unsup -DGLBL"
        };
    }

    // options = {custom_bin:"", custom_arguments:"", custom_path:""}
    async lint_from_file(file, options, parameter) {
        let normalized_file = file.replace(' ', '\\ ');
        let errors = await this._lint(normalized_file, options, parameter);
        return errors;
    }

    async lint_from_code(file, code, options, parameter) {
        let temp_file = await this._create_temp_file_of_code(code);
        let errors = await this._lint(temp_file, options, parameter);
        return errors;
    }

    async _lint(file, options, parameter) {
        let result = await this._exec_linter(file, parameter.SYNT,
            parameter.SYNT_WINDOWS, options);
        let file_split_space = file.split('\\ ')[0];
        let errors_str = result.stderr;
        let errors_str_lines = errors_str.split(/\r?\n/g);
        let errors = [];
        // Parse output lines
        errors_str_lines.forEach((line) => {
            if (line.startsWith('%')) {
                // remove the %
                line = line.substr(1);

                // was it for a submodule
                if (line.search(file_split_space) > 0) {
                    // remove the filename
                    line = line.replace(file_split_space, '');
                    line = line.replace(/\s+/g, ' ').trim();

                    let terms = this.split_terms(line);
                    let severity = this.getSeverity(terms[0]);
                    let message = terms.slice(2).join(' ');
                    let lineNum = parseInt(terms[1].trim()) - 1;

                    if (!isNaN(lineNum)) {
                        let error = {
                            'severity': severity,
                            'description': message,
                            'location': {
                                'file': file.replace('\\ ', ' '),
                                'position': [lineNum, 0]
                            }
                        };
                        errors.push(error);
                    }
                }
            }
        });
        return errors;
    }
    split_terms(line) {
        let terms = line.split(':');
        for (var i = 0; i < terms.length; i++) {
            if (terms[i] === ' ') {
                terms.splice(i, 1);
                i--;
            }
            else {
                terms[i] = terms[i].trim();
            }
        }
        return terms;
    }
    getSeverity(severity_string) {
        let severity = "";
        if (severity_string.startsWith('Error')) {
            severity = "error";
        }
        else if (severity_string.startsWith('Warning')) {
            severity = "warning";
        }
        return severity;
    }
}

module.exports = {
    Verilator: Verilator
};
