// HDLparser/index.js
"use strict";

const simParser  = require("./Parser/simParser");
exports.simParser = simParser;

const vlogParser = require("./Parser/vlogParser");
exports.vlogParser = vlogParser;

const vhdlParser = require("./Parser/vhdlParser");
exports.vhdlParser = vhdlParser;

class index {
    constructor(statusbar, parser, channel, HDLparam) {
        this.building = false;
        this.symbolsCount = 0;
        this.NUM_FILES = 250;
        this.FileExtensions = ["sv", "v", "svh", "vh", "json"];
        this.globPattern = "**/*.{" + this.FileExtensions.join(",") + "}";
        this.exclude = undefined;
        this.forceFastIndexing = false;

        this.HDLparam = HDLparam
        this.statusbar = statusbar;
        this.parser = parser;
        this.outputChannel = channel;

        this.symbols = new Map();
        utils.getOpeParam(opeParam);
        this.settings = vscode.workspace.getConfiguration();
        // if (!settings.get('HDL.Indexing')) {
        //     this.statusbar.text = "HDL: Indexing disabled on boot";
        // }
        // else {
        //     this.build_index().then(() => {
        //         this.updateMostRecentSymbols(undefined);
        //         new tree.FileExplorer(this.parser, this.globPattern, this.HDLparam);
        //     });
        // }
    };
    /**
        Processes one file and updates this.symbols with an entry if symbols exist in the file.

        @param uri uri to the document
        @param total_files total number of files to determine parse-precision
    */
    processFile(uri, total_files = 0, HDLparam) {
        return __awaiter(this, void 0, void 0, function* () {
            return new Promise((resolve) => __awaiter(this, void 0, void 0, function* () {
                resolve(vscode.workspace.openTextDocument(uri).then(doc => {
                    if (total_files >= 1000 * this.parallelProcessing || this.forceFastIndexing) {
                        return this.parser.get_HDLfileparam(doc, null, 0, null, HDLparam);
                    }
                    else if (total_files >= 100 * this.parallelProcessing) {
                        return this.parser.get_HDLfileparam(doc, null, 0, null, HDLparam);
                    }
                    else {
                        return this.parser.get_HDLfileparam(doc, null, 0, null, HDLparam);
                    }
                }));
            })).then((output) => {
                if (output.length > 0) {
                    if (this.symbols.has(uri.fsPath)) {
                        this.symbolsCount += output.length - this.symbols.get(uri.fsPath).length;
                    }
                    else {
                        this.symbolsCount += output.length;
                    }
                    this.symbols.set(uri.fsPath, output);
                    if (total_files == 0) { // If total files is 0, it is being used onChange
                        this.statusbar.text = 'HDL: ' + this.symbolsCount + ' indexed objects';
                    }
                }
            }).catch((error) => {
                // this.outputChannel.appendLine("HDL: Indexing: Unable to process file: " + uri.toString());
                // this.outputChannel.appendLine(error);
                if (this.symbols.has(uri.fsPath)) {
                    this.symbolsCount -= this.symbols.get(uri.fsPath).length;
                    this.symbols.delete(uri.fsPath);
                }
                return undefined;
            });
        });
    }
    /**
        Scans the `workspace` for HDL,
        Looks up all the `symbols` that it exist on the queried files,
        and saves the symbols as `HDLSymbol` objects to `this.symbols`.

        @return status message when indexing is successful or failed with an error.
    */
    build_index() {
        return __awaiter(this, void 0, void 0, function* () {
            var cancelled = false;
            this.building = true;
            this.symbolsCount = 0;
            this.statusbar.text = "HDL: Indexing..";
            const settings = vscode.workspace.getConfiguration();
            this.forceFastIndexing  = settings.get('HDL.forceFastIndexing');
            this.parallelProcessing = settings.get('HDL.parallelProcessing');
            let exclude = settings.get('HDL.excludeIndexing');
            if (exclude == "insert globPattern here") {
                exclude = undefined;
            }
            return yield vscode.window.withProgress({
                location: vscode.ProgressLocation.Notification,
                title: "HDL Indexing...",
                cancellable: true
            }, (_progress, token) => __awaiter(this, void 0, void 0, function* () {
                this.symbols = new Map();
                let uris = yield Promise.resolve(vscode.workspace.findFiles(this.globPattern, exclude, undefined, token));

                for (var filenr = 0; filenr < uris.length; filenr += this.parallelProcessing) {
                    let subset = uris.slice(filenr, filenr + this.parallelProcessing);
                    if (token.isCancellationRequested) {
                        cancelled = true;
                        break;
                    }
                    yield Promise.all(subset.map(uri => {
                        return this.processFile(uri, uris.length, this.HDLparam);
                    }));
                }
            })).then(() => {
                this.building = false;
                this.parser.get_instModulePath(this.HDLparam);
                if (cancelled) {
                    this.statusbar.text = "HDL: Indexing cancelled";
                }
                else {
                    this.statusbar.text = 'HDL: ' + this.symbolsCount + ' indexed objects';
                }
            });
        });
    }
    /**
        Updates `mostRecentSymbols` with the most recently used symbols
        When `mostRecentSymbols` is undefined, add the top `this.NUM_FILES` symbol from `this.symbols`
        When `mostRecentSymbols` is defined, add the symbols in `recentSymbols` one by one to the top of the array

        @param recentSymbols the recent symbols
    */
    updateMostRecentSymbols(recentSymbols) {
        if (this.mostRecentSymbols) {
            if (!recentSymbols) {
                return;
            }
            while (recentSymbols.length > 0) {
                let currentSymbol = recentSymbols.pop();
                //if symbol already exists, remove it
                for (let i = 0; i < this.mostRecentSymbols.length; i++) {
                    let symbol = this.mostRecentSymbols[i];
                    if (symbol == currentSymbol) {
                        this.mostRecentSymbols.splice(i, 1);
                        break;
                    }
                }
                //if the array has reached maximum capacity, remove the last element
                if (this.mostRecentSymbols.length >= this.NUM_FILES) {
                    this.mostRecentSymbols.pop();
                }
                //add the symbol to the top of the array
                this.mostRecentSymbols.unshift(currentSymbol);
            }
        }
        else {
            let maxSymbols = new Array();
            //collect the top symbols in `this.symbols`
            for (var list of this.symbols.values()) {
                if (maxSymbols.length + list.length >= this.NUM_FILES) {
                    let limit = this.NUM_FILES - maxSymbols.length;
                    maxSymbols = maxSymbols.concat(list.slice(-1 * limit));
                    break;
                }
                else {
                    maxSymbols = maxSymbols.concat(list);
                }
            }
            this.mostRecentSymbols = maxSymbols;
        }
    }
}
exports.index = index;
